<!DOCTYPE HTML>
<html lang="zh-CN">


<head>
    <meta charset="utf-8">
    <meta name="keywords" content="lncRNA教程, HL&#39;s blog">
    <meta name="description" content="小知识点：（输入id和gtf。得到gtf文件）# 剩余的transcripts得到gtftbtools也有这个插件
1grep -w -Ff filter1_transcript_ID -w ~/lncRNA_project/06.gffc">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
    <meta name="renderer" content="webkit|ie-stand|ie-comp">
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="format-detection" content="telephone=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <meta name="referrer" content="no-referrer-when-downgrade">
    <!-- Global site tag (gtag.js) - Google Analytics -->


    <title>lncRNA教程 | HL&#39;s blog</title>
    <link rel="icon" type="image/png" href="/favicon.png">
    


    <!-- bg-cover style     -->



<link rel="stylesheet" type="text/css" href="/libs/awesome/css/all.min.css">
<link rel="stylesheet" type="text/css" href="/libs/materialize/materialize.min.css">
<link rel="stylesheet" type="text/css" href="/libs/aos/aos.css">
<link rel="stylesheet" type="text/css" href="/libs/animate/animate.min.css">
<link rel="stylesheet" type="text/css" href="/libs/lightGallery/css/lightgallery.min.css">
<link rel="stylesheet" type="text/css" href="/css/matery.css">
<link rel="stylesheet" type="text/css" href="/css/my.css">
<link rel="stylesheet" type="text/css" href="/css/dark.css" media="none" onload="if(media!='all')media='all'">




    <link rel="stylesheet" href="/libs/tocbot/tocbot.css">
    <link rel="stylesheet" href="/css/post.css">




    
        <link rel="stylesheet" type="text/css" href="/css/reward.css">
    



    <script src="/libs/jquery/jquery-3.6.0.min.js"></script>

<meta name="generator" content="Hexo 5.4.2"></head>


<body>
    <header class="navbar-fixed">
    <nav id="headNav" class="bg-color nav-transparent">
        <div id="navContainer" class="nav-wrapper container">
            <div class="brand-logo">
                <a href="/" class="waves-effect waves-light">
                    
                    <img src="/medias/logo.png" class="logo-img" alt="LOGO">
                    
                    <span class="logo-span">HL&#39;s blog</span>
                </a>
            </div>
            

<a href="#" data-target="mobile-nav" class="sidenav-trigger button-collapse"><i class="fas fa-bars"></i></a>
<ul class="right nav-menu">
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/" class="waves-effect waves-light">
      
      <i class="fas fa-home" style="zoom: 0.6;"></i>
      
      <span>首页</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/tags" class="waves-effect waves-light">
      
      <i class="fas fa-tags" style="zoom: 0.6;"></i>
      
      <span>标签</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/categories" class="waves-effect waves-light">
      
      <i class="fas fa-bookmark" style="zoom: 0.6;"></i>
      
      <span>分类</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/archives" class="waves-effect waves-light">
      
      <i class="fas fa-archive" style="zoom: 0.6;"></i>
      
      <span>归档</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/about" class="waves-effect waves-light">
      
      <i class="fas fa-user-circle" style="zoom: 0.6;"></i>
      
      <span>关于</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/contact" class="waves-effect waves-light">
      
      <i class="fas fa-comments" style="zoom: 0.6;"></i>
      
      <span>留言板</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/friends" class="waves-effect waves-light">
      
      <i class="fas fa-address-book" style="zoom: 0.6;"></i>
      
      <span>友情链接</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="" class="waves-effect waves-light">

      
      <i class="fas fa-tools" style="zoom: 0.6;"></i>
      
      <span>Utilities</span>
      <i class="fas fa-chevron-down" aria-hidden="true" style="zoom: 0.6;"></i>
    </a>
    <ul class="sub-nav menus_item_child ">
      
      <li>
        <a href="/html/%E9%9A%8F%E6%9C%BA%E8%BD%AE%E7%9B%98.html">
          
          <i class="fas fa-random" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>随机轮盘</span>
        </a>
      </li>
      
      <li>
        <a href="/html/%E4%B8%8D%E8%A6%81%E7%86%AC%E5%A4%9C.html">
          
          <i class="fas fa-moon" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>不要熬夜</span>
        </a>
      </li>
      
      <li>
        <a href="/html/%E4%BF%9D%E6%8A%A4%E7%9C%BC%E7%9D%9B.html">
          
          <i class="fas fa-eye" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>保护眼睛</span>
        </a>
      </li>
      
      <li>
        <a href="/html/%E5%A4%AA%E9%98%B3.html">
          
          <i class="fas fa-sun" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>太阳</span>
        </a>
      </li>
      
      <li>
        <a href="/html/%E6%96%87%E5%AD%97%E8%BD%AC%E8%AF%AD%E8%A8%80.html">
          
          <i class="fas fa-voice" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>文字转语言</span>
        </a>
      </li>
      
      <li>
        <a href="/html/tnt%E5%BC%B9%E9%81%93%E8%BE%85%E5%8A%A9.html">
          
          <i class="fas fa-bomb" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>tnt弹道辅助</span>
        </a>
      </li>
      
      <li>
        <a href="/html/%E5%AE%8C%E4%B8%8D%E6%88%90%E7%9A%84%E6%AF%8F%E6%97%A5%E4%BB%BB%E5%8A%A1.html">
          
          <i class="fas fa-tasks" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>完不成的每日任务</span>
        </a>
      </li>
      
    </ul>
    
  </li>
  
  <li>
    <a href="#searchModal" class="modal-trigger waves-effect waves-light">
      <i id="searchIcon" class="fas fa-search" title="搜索" style="zoom: 0.85;"></i>
    </a>
  </li>
  <li>
    <a href="javascript:;" class="waves-effect waves-light" onclick="switchNightMode()" title="深色/浅色模式" >
      <i id="sum-moon-icon" class="fas fa-sun" style="zoom: 0.85;"></i>
    </a>
  </li>
</ul>


<div id="mobile-nav" class="side-nav sidenav">

    <div class="mobile-head bg-color">
        
        <img src="/medias/logo.png" class="logo-img circle responsive-img">
        
        <div class="logo-name">HL&#39;s blog</div>
        <div class="logo-desc">
            
            Never really desperate, only the lost of the soul.
            
        </div>
    </div>

    <ul class="menu-list mobile-menu-list">
        
        <li class="m-nav-item">
	  
		<a href="/" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-home"></i>
			
			首页
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/tags" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-tags"></i>
			
			标签
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/categories" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-bookmark"></i>
			
			分类
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/archives" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-archive"></i>
			
			归档
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/about" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-user-circle"></i>
			
			关于
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/contact" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-comments"></i>
			
			留言板
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/friends" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-address-book"></i>
			
			友情链接
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="javascript:;">
			
				<i class="fa-fw fas fa-tools"></i>
			
			Utilities
			<span class="m-icon"><i class="fas fa-chevron-right"></i></span>
		</a>
            <ul  style="background:  ;" >
              
                <li>

                  <a href="/html/%E9%9A%8F%E6%9C%BA%E8%BD%AE%E7%9B%98.html " style="margin-left:75px">
				  
				   <i class="fa fas fa-random" style="position: absolute;left:50px" ></i>
			      
		          <span>随机轮盘</span>
                  </a>
                </li>
              
                <li>

                  <a href="/html/%E4%B8%8D%E8%A6%81%E7%86%AC%E5%A4%9C.html " style="margin-left:75px">
				  
				   <i class="fa fas fa-moon" style="position: absolute;left:50px" ></i>
			      
		          <span>不要熬夜</span>
                  </a>
                </li>
              
                <li>

                  <a href="/html/%E4%BF%9D%E6%8A%A4%E7%9C%BC%E7%9D%9B.html " style="margin-left:75px">
				  
				   <i class="fa fas fa-eye" style="position: absolute;left:50px" ></i>
			      
		          <span>保护眼睛</span>
                  </a>
                </li>
              
                <li>

                  <a href="/html/%E5%A4%AA%E9%98%B3.html " style="margin-left:75px">
				  
				   <i class="fa fas fa-sun" style="position: absolute;left:50px" ></i>
			      
		          <span>太阳</span>
                  </a>
                </li>
              
                <li>

                  <a href="/html/%E6%96%87%E5%AD%97%E8%BD%AC%E8%AF%AD%E8%A8%80.html " style="margin-left:75px">
				  
				   <i class="fa fas fa-voice" style="position: absolute;left:50px" ></i>
			      
		          <span>文字转语言</span>
                  </a>
                </li>
              
                <li>

                  <a href="/html/tnt%E5%BC%B9%E9%81%93%E8%BE%85%E5%8A%A9.html " style="margin-left:75px">
				  
				   <i class="fa fas fa-bomb" style="position: absolute;left:50px" ></i>
			      
		          <span>tnt弹道辅助</span>
                  </a>
                </li>
              
                <li>

                  <a href="/html/%E5%AE%8C%E4%B8%8D%E6%88%90%E7%9A%84%E6%AF%8F%E6%97%A5%E4%BB%BB%E5%8A%A1.html " style="margin-left:75px">
				  
				   <i class="fa fas fa-tasks" style="position: absolute;left:50px" ></i>
			      
		          <span>完不成的每日任务</span>
                  </a>
                </li>
              
            </ul>
          
        </li>
        
        
        <li><div class="divider"></div></li>
        <li>
            <a href="https://github.com/blinkfox/hexo-theme-matery" class="waves-effect waves-light" target="_blank">
                <i class="fab fa-github-square fa-fw"></i>Fork Me
            </a>
        </li>
        
    </ul>
</div>


        </div>

        
            <style>
    .nav-transparent .github-corner {
        display: none !important;
    }

    .github-corner {
        position: absolute;
        z-index: 10;
        top: 0;
        right: 0;
        border: 0;
        transform: scale(1.1);
    }

    .github-corner svg {
        color: #0f9d58;
        fill: #fff;
        height: 64px;
        width: 64px;
    }

    .github-corner:hover .octo-arm {
        animation: a 0.56s ease-in-out;
    }

    .github-corner .octo-arm {
        animation: none;
    }

    @keyframes a {
        0%,
        to {
            transform: rotate(0);
        }
        20%,
        60% {
            transform: rotate(-25deg);
        }
        40%,
        80% {
            transform: rotate(10deg);
        }
    }
</style>

<a href="https://github.com/blinkfox/hexo-theme-matery" class="github-corner tooltipped hide-on-med-and-down" target="_blank"
   data-tooltip="Fork Me" data-position="left" data-delay="50">
    <svg viewBox="0 0 250 250" aria-hidden="true">
        <path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path>
        <path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2"
              fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path>
        <path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z"
              fill="currentColor" class="octo-body"></path>
    </svg>
</a>
        
    </nav>

</header>

    
<script src="/libs/cryptojs/crypto-js.min.js"></script>
<script>
    (function() {
        let pwd = '';
        if (pwd && pwd.length > 0) {
            if (pwd !== CryptoJS.SHA256(prompt('请输入访问本文章的密码')).toString(CryptoJS.enc.Hex)) {
                alert('密码错误，将返回主页！');
                location.href = '/';
            }
        }
    })();
</script>




<div class="bg-cover pd-header post-cover" style="background-image: url('/medias/featureimages/6.jpg')">
    <div class="container" style="right: 0px;left: 0px;">
        <div class="row">
            <div class="col s12 m12 l12">
                <div class="brand">
                    <h1 class="description center-align post-title">lncRNA教程</h1>
                </div>
            </div>
        </div>
    </div>
</div>




<main class="post-container content">

    
    <div class="row">
    <div id="main-content" class="col s12 m12 l9">
        <!-- 文章内容详情 -->
<div id="artDetail">
    <div class="card">
        <div class="card-content article-info">
            <div class="row tag-cate">
                <div class="col s7">
                    
                          <div class="article-tag">
                            <span class="chip bg-color">无标签</span>
                          </div>
                    
                </div>
                <div class="col s5 right-align">
                    
                </div>
            </div>

            <div class="post-info">
                
                <div class="post-date info-break-policy">
                    <i class="far fa-calendar-minus fa-fw"></i>发布日期:&nbsp;&nbsp;
                    2024-02-06
                </div>
                

                

                

                

                
            </div>
        </div>
        <hr class="clearfix">

        

        

        <div class="card-content article-card-content">
            <div id="articleContent">
                <p>小知识点：（输入id和gtf。得到gtf文件）# 剩余的transcripts得到gtf<br>tbtools也有这个插件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">grep -w -Ff filter1_transcript_ID -w ~/lncRNA_project/06.gffcompare/merged.annotated.gtf &gt; filter1_transcript.gtf</span><br></pre></td></tr></table></figure>


<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/gffr.png" class="" title="经常用的代码">



<h1 id="可能有用的网站"><a href="#可能有用的网站" class="headerlink" title="可能有用的网站"></a>可能有用的网站</h1><p>1，在鸡肌内前脂肪细胞分化过程中使用 RNA 测序分析长非编码 RNA 和 mRNA |公共图书馆一号 (plos.org)<br><a target="_blank" rel="noopener" href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0172389">https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0172389</a></p>
<p>2，如何预测LncRNA靶基因（靶蛋白）？ (sohu.com)lncRNA分析 - 简书 (jianshu.com)这个看着很厉害<br><a target="_blank" rel="noopener" href="https://www.sohu.com/a/110662733_390793">https://www.sohu.com/a/110662733_390793</a></p>
<p><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/b6c40e9612c4">https://www.jianshu.com/p/b6c40e9612c4</a></p>
<p>3，lncRNA分析完全指北 - 知乎 (zhihu.com)简单又详细的<br><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/615591750">https://zhuanlan.zhihu.com/p/615591750</a></p>
<p>4，STEP8:鉴定全新的lncRNA - 简书 (jianshu.com)<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/5b104830751b">https://www.jianshu.com/p/5b104830751b</a></p>
<p>5，文献分享 | 长链非编码RNAs(lncRNAs)分析流程 | 转录组分析流程 - 知乎 (zhihu.com)<br><a target="_blank" rel="noopener" href="https://www.zhihu.com/zvideo/1466176131120066560">https://www.zhihu.com/zvideo/1466176131120066560</a></p>
<p>6，GitHub - Alipe2021/NLncCirSmk：用于 lncRNA、环状 RNA 和新型 mRNA 鉴定的管道。<br><a target="_blank" rel="noopener" href="https://github.com/Alipe2021/NLncCirSmk">https://github.com/Alipe2021/NLncCirSmk</a></p>
<p>7，小杜的生信筆記 - 知乎 (zhihu.com)可以询问他问题<br><a target="_blank" rel="noopener" href="https://www.zhihu.com/people/xiaodu.com">https://www.zhihu.com/people/xiaodu.com</a></p>
<p>8，转录组上游分析教程 | 课程介绍 - 知乎 (zhihu.com)付费教程，可能更全一点<br><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/619830781">https://zhuanlan.zhihu.com/p/619830781</a></p>
<p>9，lncRNA鉴定方法整理 - 知乎 (zhihu.com)整理了很多方法，很有用<br><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/612384333">https://zhuanlan.zhihu.com/p/612384333</a></p>
<p>10，如何快速从基因组中提取基因、转录本、蛋白、启动子、非编码序列？ - 知乎 (zhihu.com)<br><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/439168788">https://zhuanlan.zhihu.com/p/439168788</a></p>
<p>11，如何快速从基因组中提取基因、转录本、蛋白、启动子、非编码序列？ - 知乎 (zhihu.com)<br><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/439168788">https://zhuanlan.zhihu.com/p/439168788</a></p>
<p>12，长非编码RNA和mRNA的综合分析揭示了玉米幼苗根系对盐胁迫响应的调控网络 |BMC 基因组学 |全文 (biomedcentral.com)这个里面有个脚本<br><a target="_blank" rel="noopener" href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-021-08286-7">https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-021-08286-7</a></p>
<p>13，RNA-seq：转录组数据分析处理（上）_super_qun的博客-CSDN博客<br><a target="_blank" rel="noopener" href="https://blog.csdn.net/weixin_44452187/article/details/86646409?spm=1001.2101.3001.6650.2&amp;utm_medium=distribute.pc_relevant.none-task-blog-2~default~CTRLIST~Rate-2-86646409-blog-129909917.235%5Ev38%5Epc_relevant_sort_base1&amp;depth_1-utm_source=distribute.pc_relevant.none-task-blog-2~default~CTRLIST~Rate-2-86646409-blog-129909917.235%5Ev38%5Epc_relevant_sort_base1&amp;utm_relevant_index=5">https://blog.csdn.net/weixin_44452187/article/details/86646409?spm=1001.2101.3001.6650.2&amp;utm_medium=distribute.pc_relevant.none-task-blog-2%7Edefault%7ECTRLIST%7ERate-2-86646409-blog-129909917.235%5Ev38%5Epc_relevant_sort_base1&amp;depth_1-utm_source=distribute.pc_relevant.none-task-blog-2%7Edefault%7ECTRLIST%7ERate-2-86646409-blog-129909917.235%5Ev38%5Epc_relevant_sort_base1&amp;utm_relevant_index=5</a></p>
<p>细致一点<br>14，RNA-seq入门实战（三）：从featureCounts与Salmon输出文件获取counts矩阵 - 简书 (jianshu.com)<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/15ef7df72391">https://www.jianshu.com/p/15ef7df72391</a></p>
<p>15，4个发育时间点的总共12个鸡转录组测序样本的长非编码RNA的鉴定-腾讯云开发者社区-腾讯云 (tencent.com)<br><a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1842686">https://cloud.tencent.com/developer/article/1842686</a></p>
<p>16，在鸡肌内前脂肪细胞分化过程中使用RNA测序分析长非编码RNA和mRNA - PubMed (nih.gov)<br><a target="_blank" rel="noopener" href="https://pubmed.ncbi.nlm.nih.gov/28199418/">https://pubmed.ncbi.nlm.nih.gov/28199418/</a></p>
<p>17，关于stringtie定量基因的时候，最后输出很多MSTRG样式的geneid - 简书 (jianshu.com)<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/1697f2d91eb5">https://www.jianshu.com/p/1697f2d91eb5</a></p>
<p>18，「 bedtools 」提取上游+gene+下游序列 - 简书 (jianshu.com)<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/acb2bcb49882">https://www.jianshu.com/p/acb2bcb49882</a></p>
<p><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/2b6aba1263d1">https://www.jianshu.com/p/2b6aba1263d1</a></p>
<p><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/2b6aba1263d1">https://www.jianshu.com/p/2b6aba1263d1</a><br>19，3.0 转录本比对组装HISAT2+StringTie+cuffcompare - 简书 (jianshu.com)<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/2b6aba1263d1">https://www.jianshu.com/p/2b6aba1263d1</a></p>
<p><a target="_blank" rel="noopener" href="https://www.bioinfo-scrounger.com/archives/284/">https://www.bioinfo-scrounger.com/archives/284/</a></p>
<p>20，转录组的组装Stingtie和Cufflinks | KeepNotes blog (bioinfo-scrounger.com)<br><a target="_blank" rel="noopener" href="https://www.bioinfo-scrounger.com/archives/284/">https://www.bioinfo-scrounger.com/archives/284/</a></p>
<p>21，RNA-seq入门实战（零）：RNA-seq流程前的准备——Linux与R的环境创建 - 知乎 (zhihu.com)<br><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/518132262">https://zhuanlan.zhihu.com/p/518132262</a></p>
<p>22，小L生信学习日记-3丨原始数据质量如何判断？-上 (qq.com)<br>看图<br><a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s?__biz=MzU3ODgxMDYyNw==&amp;mid=2247483687&amp;idx=1&amp;sn=8689b12a0d30f90ae9ebe72182ee2899&amp;chksm=fd6ee74bca196e5de8c17e762fb423c7e0cd24b8e8edb5fe4cf94ab061d166a10257447d3eb5&amp;scene=21#wechat_redirect">https://mp.weixin.qq.com/s?__biz=MzU3ODgxMDYyNw==&amp;mid=2247483687&amp;idx=1&amp;sn=8689b12a0d30f90ae9ebe72182ee2899&amp;chksm=fd6ee74bca196e5de8c17e762fb423c7e0cd24b8e8edb5fe4cf94ab061d166a10257447d3eb5&amp;scene=21#wechat_redirect</a></p>
<p>23，小L生信学习日记-4丨原始数据质量如何判断？-下 - 知乎 (zhihu.com)<br><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/57628300">https://zhuanlan.zhihu.com/p/57628300</a><br>质控的<br>24，lncRNA组装流程的软件介绍之trim-galore (qq.com)<br><a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s?search_click_id=2253394900344533860-1633341597788-628761&amp;sub=&amp;__biz=MzAxMDkxODM1Ng==&amp;mid=2247503527&amp;idx=4&amp;sn=261be2c7ff2a7f80b2e6ab139028d780&amp;chksm=9b4b8e1cac3c070a963f59fcc55095fe0ae37d1eecde3e67f6682779c91d0e907a8fd7eb8f84&amp;scene=3&amp;subscene=10000&amp;clicktime=1633341597&amp;enterid=1633341597&amp;ascene=0&amp;devicetype=android-30&amp;version=28000f35&amp;nettype=cmnet&amp;abtest_cookie=AAACAA==&amp;lang=zh_CN&amp;exportkey=AdXQxZkOzbrJK+UwprsoMEk=&amp;pass_ticket=pvvLsQPcUT6hjrE3KSHbHTaZKL/VtNBVq+LbY9r0hucz/DdbU3NgO9ofB9mtC3fS&amp;wx_header=1">https://mp.weixin.qq.com/s?search_click_id=2253394900344533860-1633341597788-628761&amp;sub=&amp;__biz=MzAxMDkxODM1Ng==&amp;mid=2247503527&amp;idx=4&amp;sn=261be2c7ff2a7f80b2e6ab139028d780&amp;chksm=9b4b8e1cac3c070a963f59fcc55095fe0ae37d1eecde3e67f6682779c91d0e907a8fd7eb8f84&amp;scene=3&amp;subscene=10000&amp;clicktime=1633341597&amp;enterid=1633341597&amp;ascene=0&amp;devicetype=android-30&amp;version=28000f35&amp;nettype=cmnet&amp;abtest_cookie=AAACAA%3D%3D&amp;lang=zh_CN&amp;exportkey=AdXQxZkOzbrJK%2BUwprsoMEk%3D&amp;pass_ticket=pvvLsQPcUT6hjrE3KSHbHTaZKL%2FVtNBVq%2BLbY9r0hucz%2FDdbU3NgO9ofB9mtC3fS&amp;wx_header=1</a></p>
<p>25，基因组注释文件格式 –（一）BED文件格式 - 简书 (jianshu.com)<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/9208c3b89e44">https://www.jianshu.com/p/9208c3b89e44</a></p>
<p>26，一文搞懂NCBI Blast本地数据库（NT/NR等）构建 - 知乎 (zhihu.com)<br><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/296579512">https://zhuanlan.zhihu.com/p/296579512</a></p>
<p>27，转录组测序（RNA-seq）的数据如何分析？_转录组数据怎么分析_砷在氟中的博客-CSDN博客<br>Nr数据库<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/45fdb5cf930a">https://www.jianshu.com/p/45fdb5cf930a</a></p>
<p><a target="_blank" rel="noopener" href="https://blog.csdn.net/weixin_55372631/article/details/129909917#:~:text=%E8%8B%A5%E6%A0%B7%E6%9C%AC%E6%95%B0%E7%9B%AE%E5%A4%A7%E4%BA%8E1,3%E4%B8%AA%E5%90%84%E5%85%B3%E9%94%AE%E5%9F%BA%E5%9B%A0%E3%80%82">https://blog.csdn.net/weixin_55372631/article/details/129909917#:~:text=%E8%8B%A5%E6%A0%B7%E6%9C%AC%E6%95%B0%E7%9B%AE%E5%A4%A7%E4%BA%8E1,3%E4%B8%AA%E5%90%84%E5%85%B3%E9%94%AE%E5%9F%BA%E5%9B%A0%E3%80%82</a></p>
<p>28，在线工具】lncRNA编号XLOC_xxxxxx转换 - 简书 (jianshu.com)<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/44f8081e332d">https://www.jianshu.com/p/44f8081e332d</a></p>
<p>29，stringtie merge与gene_id - 简书 (jianshu.com)<br>这个网站只是给个思路。<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/d828d45d8b6c">https://www.jianshu.com/p/d828d45d8b6c</a></p>
<p>30，gffcompare的使用说明 - 简书 (jianshu.com)<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/08bb35f8e7c2">https://www.jianshu.com/p/08bb35f8e7c2</a></p>
<p>重要31.lncRNA分析<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/b6c40e9612c4">https://www.jianshu.com/p/b6c40e9612c4</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">grep -Ff transcript_ID -w mergelist原来.gtf &gt; lncRNA.gtf</span><br><span class="line"></span><br><span class="line">gffread protein_coding_gene.gtf -g /home/data1/Ghl/GHL/gtf/Gallus_gallus.GRCg6a.dna.toplevel.fa -w protein_coding_gene.fa</span><br><span class="line"></span><br><span class="line">gffread -w transcripts.fa -g Gallus_gallus.GRCg6a.dna.toplevel.fa merge_2.gtf</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>lncRNA组装流程的软件介绍之Stringtie (360doc.com)<br><a target="_blank" rel="noopener" href="http://www.360doc.com/content/21/0714/12/76149697_986501836.shtml">http://www.360doc.com/content/21/0714/12/76149697_986501836.shtml</a></p>
<p>非常详细介绍了转录本组装和定量，以及-e后不能筛选出新的</p>
<p>LncRNA的组装和鉴定（下游流程） (qq.com)<br><a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247504567&amp;idx=1&amp;sn=3ddf2df1cd9554e2718de0abff04959e&amp;chksm=9b4b920cac3c1b1a0ac7ae48b5de9802e7eeebf91a880fad9ffe13cd09132fb35a1405bd8d2f&amp;cur_album_id=1322384987060142080&amp;scene=189#wechat_redirect">https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247504567&amp;idx=1&amp;sn=3ddf2df1cd9554e2718de0abff04959e&amp;chksm=9b4b920cac3c1b1a0ac7ae48b5de9802e7eeebf91a880fad9ffe13cd09132fb35a1405bd8d2f&amp;cur_album_id=1322384987060142080&amp;scene=189#wechat_redirect</a></p>
<p>jackhmmer search | HMMER (ebi.ac.uk)<br><a target="_blank" rel="noopener" href="https://www.ebi.ac.uk/Tools/hmmer/search/jackhmmer">https://www.ebi.ac.uk/Tools/hmmer/search/jackhmmer</a></p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%9B%BE1.png" class="" title="可能有用的网站">

<p>图：来源lncRNA测序分析优势及分析流程-百迈客生物 - 知乎 (zhihu.com)</p>
<p>没必要全部，选重要的</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%9B%BE2.png" class="" title="可能有用的网站">

<p>图：长非编码RNA和mRNA的综合分析揭示了玉米幼苗根系对盐胁迫响应的调控网络 |BMC 基因组学 |全文 (biomedcentral.com)</p>
<p><a target="_blank" rel="noopener" href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-021-08286-7">https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-021-08286-7</a></p>
<h2 id="lncRNA检测工作流程，理论上的"><a href="#lncRNA检测工作流程，理论上的" class="headerlink" title="lncRNA检测工作流程，理论上的"></a>lncRNA检测工作流程，理论上的</h2><p>从最开始的数据下载到最后筛选出lncRNA的完整工作流程涉及多个步骤。下面的指南提供了每一步的代码示例，但请注意，这些示例需要根据实际环境进行调整，包括文件路径、软件安装路径等。</p>
<h3 id="步骤-1-下载测序数据"><a href="#步骤-1-下载测序数据" class="headerlink" title="步骤 1: 下载测序数据"></a>步骤 1: 下载测序数据</h3><ol>
<li><p><strong>下载SRA文件</strong>。使用SRA Toolkit的<code>prefetch</code>工具下载SRA格式的测序数据文件。例如，下载样本SRR4046839：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">prefetch SRR4046839</span><br></pre></td></tr></table></figure></li>
<li><p><strong>转换SRA文件为FASTQ格式</strong>。使用<code>fastq-dump</code>工具将SRA文件转换成FASTQ格式，便于后续分析。tbtools上有这个插件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fastq-dump --outdir /home/data1/Ghl/all --split-files SRR4046839</span><br></pre></td></tr></table></figure></li>
</ol>
<h3 id="步骤-2-质量控制"><a href="#步骤-2-质量控制" class="headerlink" title="步骤 2: 质量控制"></a>步骤 2: 质量控制</h3><ol>
<li><p><strong>质量检查</strong>。使用FastQC对FASTQ文件进行质量检查。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fastqc /home/data1/Ghl/all/SRR4046839_1.fastq /home/data1/Ghl/all/SRR4046839_2.fastq -o /home/data1/Ghl/fastqc_reports</span><br></pre></td></tr></table></figure></li>
<li><p><strong>质量修剪</strong>。使用Trimmomatic对数据进行质量修剪和适配器序列的去除。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">java -jar trimmomatic.jar PE -phred33 \</span><br><span class="line">/home/data1/Ghl/all/SRR4046839_1.fastq /home/data1/Ghl/all/SRR4046839_2.fastq \</span><br><span class="line">/home/data1/Ghl/trimmed/SRR4046839_1_paired.fastq /home/data1/Ghl/trimmed/SRR4046839_1_unpaired.fastq \</span><br><span class="line">/home/data1/Ghl/trimmed/SRR4046839_2_paired.fastq /home/data1/Ghl/trimmed/SRR4046839_2_unpaired.fastq \</span><br><span class="line">ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36</span><br></pre></td></tr></table></figure></li>
</ol>
<h3 id="步骤-3-序列映射"><a href="#步骤-3-序列映射" class="headerlink" title="步骤 3: 序列映射"></a>步骤 3: 序列映射</h3><ol>
<li><p><strong>构建HISAT2索引</strong>（如果尚未存在）。首先，为Gallus gallus参考基因组构建HISAT2索引。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">hisat2-build Ghl/Gallus_gallus.GRCg6a.dna.toplevel.fa Gallus_gallus_GRCg6a_hisat2_index</span><br></pre></td></tr></table></figure></li>
</ol>
<ol start="2">
<li><p><strong>使用HISAT2进行序列映射</strong>。将预处理后的读取映射到参考基因组。应该也有tbtools插件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">hisat2 -p 8 -x Gallus_gallus_GRCg6a_hisat2_index \</span><br><span class="line">-1 /home/data1/Ghl/trimmed/SRR4046839_1_paired.fastq \</span><br><span class="line">-2 /home/data1/Ghl/trimmed/SRR4046839_2_paired.fastq \</span><br><span class="line">-S /home/data1/Ghl/hisat2_out/SRR4046839_aligned.sam</span><br></pre></td></tr></table></figure>
<p>我的：</p>
</li>
</ol>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sudo hisat2 -p 16 --dta /home/dell/data/ghl/chicken/Gallus_index -1 /home/dell/data/ghl/SRR1_2/SRR6116749_1.fastq -2 /home/dell/data/ghl/SRR1_2/SRR6116749_2.fastq -S /home/dell/data/ghl/chicken/SRR6116749.sam</span><br><span class="line"></span><br></pre></td></tr></table></figure>


<h3 id="步骤-4-转录本组装和定量"><a href="#步骤-4-转录本组装和定量" class="headerlink" title="步骤 4: 转录本组装和定量"></a>步骤 4: 转录本组装和定量</h3><ol>
<li><p><strong>转换SAM为BAM格式</strong>，并对其进行排序。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools <span class="built_in">sort</span> -o /home/data1/Ghl/hisat2_out/SRR4046839_aligned.sorted.bam /home/data1/Ghl/hisat2_out/SRR4046839_aligned.sam</span><br></pre></td></tr></table></figure></li>
</ol>
<p>我的：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sudo samtools view -@ 16 -bS /home/dell/data/ghl/chicken/SRR6116777.sam -o /home/data1/Ghl/bam/SRR6116777.bam</span><br><span class="line"></span><br><span class="line">samtools <span class="built_in">sort</span> /home/data1/Ghl/bam/SRR6116777.bam -o /home/data1/Ghl/bam/SRR6116777.short.bam  </span><br></pre></td></tr></table></figure>


<ol start="2">
<li><p><strong>使用StringTie进行转录本组装</strong>。tbtools有插件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">stringtie -p 8 -G Ghl/Gallus_gallus.GRCg6a.98.gtf -o /home/data1/Ghl/stringtie_out/SRR4046839.gtf /home/data1/Ghl/hisat2_out/SRR4046839_aligned.sorted.bam</span><br></pre></td></tr></table></figure>
<p>我的：</p>
</li>
</ol>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">stringtie SRR6116749.short.bam -o /home/data1/Ghl/TAB/SRR6116749.gtf -p 28 -G Gallus_gallus.GRCg6a.98.gtf -A /home/data1/Ghl/TAB/SRR6116749_gene_abund.tab -B -e（不是这个）</span><br><span class="line"></span><br><span class="line">sudo /usr/bin/stringtie -p 16 -G /home/dell/data/ghl/chicken/Gallus_gallus.GRCg6a.98.gtf -o /home/dell/data/ghl/chicken/Gallus00.gtf -l Gallus00 /home/dell/data/ghl/chicken/SRR9109597/Gallus00.short.bam</span><br><span class="line"></span><br><span class="line">sudo /usr/bin/stringtie -p 16 -G /home/data1/Ghl/lab4046839/Gallus_gallus.GRCg6a.98.gtf -o /home/data1/Ghl/GTF/SRR6116749.gtf -l Gallus00 /home/data1/Ghl/bam/SRR6116749.short.bam</span><br><span class="line"></span><br><span class="line">&#123;% asset_img <span class="string">&quot;stringtie.png&quot;</span> <span class="string">&quot;stringTie&quot;</span>  %&#125;</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h3 id="步骤-5-lncRNA的筛选和注释，详细步骤在后面"><a href="#步骤-5-lncRNA的筛选和注释，详细步骤在后面" class="headerlink" title="步骤 5: lncRNA的筛选和注释，详细步骤在后面"></a>步骤 5: lncRNA的筛选和注释，详细步骤在后面</h3><p>命令 <code>cuffcompare</code> 是 Cufflinks 套件中的一个工具，用于比较多个转录组装文件（GTF/GFF格式）之间的差异，并且可以将转录本与参考注释进行比较。这个命令对于识别新的转录本、发现已知转录本的新变体，以及分类未知转录本非常有用。</p>
<p>让我们逐部分解析您提供的命令：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">cuffcompare -r Gallus_gallus.GRCg6a.98.gtf -o /home/data1/Ghl/lab4046839/cuffcompare mergelist.gtf</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">mergelist.gtf是组装的文件，例如</span><br><span class="line"></span><br><span class="line">stringtie --merge -p 8 \</span><br><span class="line">-o stringtie_merge.gtf \</span><br><span class="line">CK18.gtf T18.gtf CK69.gtf T69.gtf</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">运行 `cuffcompare` 命令后，通常会生成以下类型的文件：</span><br><span class="line"></span><br><span class="line">1. **摘要统计文件 (`cuffcmp.stats`)：** 包含比较结果的总体统计信息。</span><br><span class="line"></span><br><span class="line">2. **参考注释文件 (`Gallus_gallus.GRCg6a.98.gtf`)：** 你指定的参考注释文件。</span><br><span class="line"></span><br><span class="line">3. **查询注释文件 (`mergelist.gtf`)：** 你指定的查询注释文件。</span><br><span class="line"></span><br><span class="line">4. **关系跟踪文件 (`cuffcmp.tracking`)：** 详细描述了输入注释中转录本、基因等特征之间的关系，通常以表格形式呈现。</span><br><span class="line"></span><br><span class="line">5. **参考映射文件 (`cmp.refmap`)：** 显示输入注释中转录本与参考注释之间关系的映射文件。</span><br><span class="line"></span><br><span class="line">6. **参考映射统计文件 (`refmap.stats`)：** 总结输入注释与参考注释之间映射的统计信息。</span><br><span class="line"></span><br><span class="line">7. **转录本映射文件 (`mergelist.tmap`)：** 包含查询注释中转录本与参考注释之间的映射关系。</span><br><span class="line"></span><br><span class="line">这些文件中的内容和格式会根据具体的 `cuffcompare` 命令参数和版本而有所不同。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/cuffcompare.png" class="" title="cuffcompare">

<p> 上面代码解释</p>
<ul>
<li><p><code>-r Gallus_gallus.GRCg6a.98.gtf</code>: 这部分指定参考注释文件。<code>-r</code> 参数后面跟着的是参考基因组的注释文件（GTF格式），在这个例子中是 <code>Gallus gallus</code>（鸡）的参考注释，版本 GRCg6a.98。这个文件包含了鸡基因组的已知基因和转录本的位置、外显子边界等信息。</p>
</li>
<li><p><code>-o /home/data1/Ghl/lab4046839/cuffcompare</code>: 这部分指定输出目录和输出文件的前缀。<code>-o</code> 参数后面跟着的是输出目录的路径，以及用于存放 <code>cuffcompare</code> 产生的所有输出文件的前缀。在这个例子中，所有的输出文件将被保存在 <code>/home/data1/Ghl/lab4046839/cuffcompare</code> 目录下，并且文件名将以 “cuffcompare” 为前缀。</p>
</li>
<li><p><code>mergelist.gtf</code>: 这是输入文件，通常是由多个转录本组装结果文件（GTF格式）合并而成的文件，或者是单个转录本组装结果文件。<code>cuffcompare</code> 将会使用这个文件中的转录本信息，并将其与参考注释进行比较。</p>
</li>
</ul>
<p>总结来说，这个命令的目的是将 <code>mergelist.gtf</code> 文件中的转录本与 <code>Gallus gallus</code> GRCg6a.98版本的参考注释进行比较。通过这种比较，您可以识别出新的转录本、未注释的基因，或者是与参考注释相比有差异的转录本。输出结果将包括一系列文件，例如统计信息、转录本分类汇总等，这些文件将提供关于比较结果的详细信息，有助于进一步的生物信息学分析和研究。</p>
<p>或者使用gffcompare</p>
<p>是的，<code>gffcompare</code>命令与<code>cuffcompare</code>命令的作用相似，它们都用于比较GTF/GFF文件中的转录本与参考注释之间的差异。<code>gffcompare</code>是<code>cuffcompare</code>的后继工具，提供了类似的功能，但通常认为在某些方面<code>gffcompare</code>更加灵活和高效。这两个工具都能帮助研究人员分析和理解转录本组装的结果，识别新的转录本，以及评估转录本组装的质量。</p>
<p>下面是<code>gffcompare</code>命令的详细解释：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">nohup</span> gffcompare -R -r <span class="variable">$gtf</span> -o ./merged ../05.stringtie/02.merge_gtf/stringtie_merged.gtf &gt; gffcompare.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">运行这个命令后，你会得到以下结果：</span><br><span class="line"></span><br><span class="line">1. **REFMAP 文件 (`merged.refmap`)**：包含参考注释与预测转录本之间的映射关系。</span><br><span class="line"></span><br><span class="line">2. **TRANSFRAG 文件 (`merged.annotated.gtf`)**：一个包含了所有输入注释文件中的所有转录本以及它们在参考基因组上的对应关系的 GTF 文件。</span><br><span class="line"></span><br><span class="line">3. **Stats 文件 (`merged.stats`)**：包含了比较结果的统计信息。</span><br><span class="line"></span><br><span class="line">4. **Log 文件 (`gffcompare.log`)**：包含了 `gffcompare` 命令的输出信息，包括程序运行时的任何警告或错误信息。</span><br><span class="line"></span><br><span class="line">这些文件将保存在当前目录下，并且由于使用了 `<span class="built_in">nohup</span>` 命令，即使你关闭了终端，这些文件也会继续存在，直到该命令执行完毕。</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<ul>
<li><p><code>nohup</code>: 这是一个Unix命令，用于运行另一个命令，同时忽略所有挂起（hangup）信号。这意味着即使你关闭了终端或者断开了SSH连接，命令仍然在后台运行。</p>
</li>
<li><p><code>gffcompare</code>: 这是执行比较的主命令，用于比较GTF/GFF文件。</p>
</li>
<li><p><code>-R</code>: 这个选项告诉<code>gffcompare</code>在比较时考虑参考注释中的转录本重叠（包含在比较的转录本内的参考转录本）。</p>
</li>
<li><p><code>-r $gtf</code>: <code>-r</code>选项后跟参考注释文件的路径。这里使用了变量<code>$gtf</code>来指定参考GTF文件的路径。<code>$gtf</code>应该在执行命令前定义，指向正确的参考注释文件。</p>
</li>
<li><p><code>-o ./merged</code>: <code>-o</code>选项指定输出文件的前缀和输出目录。在这个例子中，所有的输出文件将被保存在当前目录下的<code>merged</code>前缀。</p>
</li>
<li><p><code>../05.stringtie/02.merge_gtf/stringtie_merged.gtf</code>: 这是要比较的GTF文件，通常是使用StringTie或其他转录本组装工具生成的合并转录本文件。</p>
</li>
<li><p><code>&gt; gffcompare.log 2&gt;&amp;1 &amp;</code>: 这部分将命令的标准输出（stdout）和标准错误（stderr）都重定向到<code>gffcompare.log</code>文件，并且在后台执行。这意味着你可以运行这个命令，然后继续在同一个终端中执行其他命令，而<code>gffcompare</code>的输出和错误都会被记录在日志文件中。</p>
</li>
</ul>
<p>总的来说，这个<code>gffcompare</code>命令的作用是将<code>stringtie_merged.gtf</code>文件中的转录本与指定的参考注释进行比较，并且考虑转录本重叠情况，其输出结果将帮助识别新的或是未注释的转录本，以及提供转录本组装质量的评估。</p>
<p>cat merged.stats</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/cat_merged.stats.png" class="" title="cat merged.stats">


<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/gffcompare%E6%9F%A5%E7%9C%8B%E6%AF%94%E5%AF%B9%E5%87%86%E7%A1%AE%E6%80%A7.png" class="" title="查看比对准确性">

<p>这个图像显示的是<code>gffcompare</code>工具生成的统计报告文件，通常命名为<code>.stats</code>，它提供了关于转录本组装质量和准确性的详细信息。报告分为几个部分，下面是对各部分内容的解释：</p>
<ol>
<li><p><strong>数据集摘要</strong>:</p>
<ul>
<li><strong>Query mRNAs</strong>: 表示从输入数据集（如<code>stringtie_merged.gtf</code>）中识别出的mRNA转录本数量，以及这些转录本分布在的基因座位（loci）数量。</li>
<li><strong>Reference mRNAs</strong>: 表示在参考注释（如基因组注释文件）中的mRNA转录本数量和基因座位数量。</li>
</ul>
</li>
<li><p><strong>转录本级别的敏感度和精确度</strong>:</p>
<ul>
<li><strong>Base level</strong>: 序列对比的基础级别，通常指核苷酸级别的准确性。</li>
<li><strong>Exon level</strong>: 外显子级别的对比，包括敏感度和精确度。敏感度指的是能够正确识别外显子的比例，而精确度指的是所有预测外显子中正确的比例。</li>
<li><strong>Intron level</strong>: 内含子级别的对比，同样包括敏感度和精确度。</li>
<li><strong>Intron chain level</strong>: 多个连续内含子的链的对比。</li>
<li><strong>Transcript level</strong>: 整个转录本级别的对比。</li>
<li><strong>Locus level</strong>: 基因座位级别的对比。</li>
</ul>
</li>
<li><p><strong>匹配和未匹配的统计</strong>:</p>
<ul>
<li><strong>Matching intron chains</strong>: 正确匹配的内含子链的数量。</li>
<li><strong>Matching transcripts</strong>: 正确匹配的转录本数量。</li>
<li><strong>Matching loci</strong>: 正确匹配的基因座位数量。</li>
<li><strong>Missed exons</strong>: 未能检测到的外显子数量和总外显子数量的比例。</li>
<li><strong>Novel exons</strong>: 新发现的外显子数量和总外显子数量的比例。</li>
<li><strong>Missed introns</strong>: 未能检测到的内含子数量和总内含子数量的比例。</li>
<li><strong>Novel introns</strong>: 新发现的内含子数量和总内含子数量的比例。</li>
<li><strong>Missed loci</strong>: 未能检测到的基因座位数量和总基因座位数量的比例。</li>
<li><strong>Novel loci</strong>: 新发现的基因座位数量和总基因座位数量的比例。</li>
</ul>
</li>
<li><p><strong>总的合并基因座位数</strong>:</p>
<ul>
<li>这表示所有输入数据集合并后的总基因座位数量。</li>
</ul>
</li>
</ol>
<p>这个报告对了解转录本组装的性能和准确性非常有用，可以帮助研究人员了解哪些区域被准确地重建，哪些区域可能需要进一步的验证或分析。敏感度（Sensitivity）是指正确识别的元素占所有应被识别元素的比例，而精确度（Precision）是指正确识别的元素占所有被识别元素的比例。通常，高敏感度和高精确度是期望达到的，但在实际应用中往往需要在二者之间寻求平衡。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">&#x27;$3!~/class/ &#123;print $3&#125;&#x27;</span> merged.stringtie_merged.gtf.tmap | <span class="built_in">sort</span> -V | <span class="built_in">uniq</span> -c</span><br></pre></td></tr></table></figure>



<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E7%BB%9F%E8%AE%A1class_code%E7%B1%BB%E5%9E%8B.png" class="" title="统计class_code类型">

<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%BE%97%E5%88%B0%E7%9A%84%E6%96%87%E4%BB%B6.png" class="" title="得到的文件">



<p>接下来主要的操作对象是 merged.stringtie_merged.gtf.tmap 文件。</p>
<h4 id="step1：保留指定class-code的transcripts"><a href="#step1：保留指定class-code的transcripts" class="headerlink" title="step1：保留指定class_code的transcripts"></a>step1：保留指定class_code的transcripts</h4><p>过滤，只保留class_code=”u”,”x”,”i”,”j”,”o”的 transcripts ，这个时候需要参考 stringtie官网提供的分类 ：</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%88%86%E7%B1%BB.png" class="" title="stringtie官网提供的分类">

<p>图片是关于<code>gffcompare</code>或<code>cuffcompare</code>工具用于转录本分类的代码说明。这些代码帮助用户了解转录本相对于参考基因组注释的情况。每个代码代表了转录本与参考序列之间的特定关系。当<code>gffcompare</code>运行时，它会为每个转录本分配一个类别代码，表示其与参考基因组的对应关系。以下是您所询问的类别代码的解释：</p>
<ul>
<li><p><code>u</code>：<strong>未知的，基因间</strong>。这个代码意味着转录本不能与参考基因组中的任何已知基因相匹配，通常位于已注释基因之间的区域。</p>
</li>
<li><p><code>x</code>：<strong>异染色体上的外显子重叠</strong>。指的是转录本的外显子与参考基因组上的外显子重叠，但是在异染色体上（如在X或Y染色体上的基因与常染色体上的基因重叠）。</p>
</li>
<li><p><code>i</code>：<strong>内含子</strong>。转录本完全位于一个已知基因的内含子区域内。</p>
</li>
<li><p><code>j</code>：<strong>至少一个结合位点匹配的多外显子转录本</strong>。这个代码表示转录本是多外显子的，并且至少有一个外显子的结合位点（即剪接位点）与参考基因组的外显子结合位点相匹配。</p>
</li>
<li><p><code>o</code>：<strong>其他链上的外显子重叠</strong>。转录本的外显子与参考基因组中的外显子在序列上重叠，但是位于反向链上。</p>
</li>
</ul>
<p>这些类别代码可以帮助研究者识别潜在的新基因、非编码RNA或错误注释的基因。例如，分类为“u”的转录本可能代表未知的基因间区域的转录活动或假阳性信号；而分类为“i”的转录本可能指示内含子保留事件或潜在的新基因。</p>
<p>在<code>gffcompare</code>的输出中，每个转录本的类别代码将在其对应的行的属性字段中标注，这些信息通常被用来进一步筛选和验证实验数据。在后续的分析中，研究者可能会关注那些被标记为“u”或“x”的转录本，因为它们可能表示新的转录活动区域。</p>
<h4 id="借鉴步骤：对比3-0-转录本比对组装HISAT2-StringTie-cuffcompare-简书-jianshu-com-，这个可以直接筛选过来"><a href="#借鉴步骤：对比3-0-转录本比对组装HISAT2-StringTie-cuffcompare-简书-jianshu-com-，这个可以直接筛选过来" class="headerlink" title="借鉴步骤：对比3.0 转录本比对组装HISAT2+StringTie+cuffcompare - 简书 (jianshu.com)，这个可以直接筛选过来"></a>借鉴步骤：对比3.0 转录本比对组装HISAT2+StringTie+cuffcompare - 简书 (jianshu.com)，这个可以直接筛选过来</h4><p><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/2b6aba1263d1">https://www.jianshu.com/p/2b6aba1263d1</a>  重要</p>
<p>从组装开始，后续重要步骤</p>
<p>下载参考注释：Phytozome Sbicolor_454_v3.1.1.gene_exons.gff3.gz</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#转格式：</span></span><br><span class="line">gffread Sbicolor_454_v3.1.1.gene_exons.gff3.gz -T -o sbi311gx.gtf</span><br><span class="line"></span><br><span class="line"><span class="comment">#组装转录本：</span></span><br><span class="line">stringtie -p 16 -G ~/sbi311gx.gtf -o T18.gtf T18.bam</span><br><span class="line">stringtie -p 16 -G ~/sbi311gx.gtf -o CK18.gtf CK18.bam</span><br><span class="line">stringtie -p 16 -G ~/sbi311gx.gtf -o CK69.gtf CK69.bam</span><br><span class="line">stringtie -p 16 -G ~/sbi311gx.gtf -o T69.gtf T69.bam</span><br><span class="line"></span><br><span class="line"><span class="comment">#四样本合并，得到非冗余转录本集</span></span><br><span class="line"> stringtie --merge -p 8 \</span><br><span class="line">-o stringtie_merge.gtf \</span><br><span class="line">CK18.gtf T18.gtf CK69.gtf T69.gtf</span><br><span class="line"></span><br><span class="line"><span class="comment">#以stringtie_merge.gtf 为参考注释重新组装</span></span><br><span class="line">stringtie -e -B -p 16 -G ~/4_stringtie/stringtie_merge.gtf -o T18.gtf ~/T18.bam</span><br><span class="line"><span class="comment">#-B代表输出Ballgown可读文件 </span></span><br><span class="line"></span><br><span class="line">Cuffcompare</span><br><span class="line">将各样品的注释文件与参考基因组+参考注释比较，得到新转录本（lncRNA)</span><br><span class="line"></span><br><span class="line">筛选条件：</span><br><span class="line"><span class="comment"># 过滤，只保留exon&gt;1并且长度&gt;200bp的transcripts</span></span><br><span class="line">1）FPKM&gt;=0.5 2）coverage &gt;1 3）Length &gt; 200 4）分类为i j o u x</span><br><span class="line">输入文件：cufcomp.CK18.gtf.tmap 等</span><br><span class="line">结果文件：class.CK18等 （内容为注释，格式gtf）</span><br><span class="line">提取序列</span><br><span class="line">awk命令，比对转录本<span class="built_in">id</span></span><br><span class="line">以class.CK18 从 CK18.gtf 中提取，文件名为CK18class.gtf</span><br><span class="line">gffread： 注释文件 、 参考基因组.fa 提取序列</span><br><span class="line">结果文件：CK18class.fa （G：备份数据）</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">cuffcompare -r ~/sbi311gx.gtf -s ~/Sbicolor_454_v3.0.1.fa -o cufcomp ../T18.gtf </span><br><span class="line">awk <span class="string">&#x27;&#123;if($7&gt;=0.5 &amp;&amp; $10 &gt; 1 &amp;&amp; $11 &gt;200) print $0&#125;&#x27;</span> cufcomp.T18.gtf.tmap &gt; filter.T18</span><br><span class="line">awk <span class="string">&#x27;&#123;if ($3==&quot;u&quot; || $3==&quot;x&quot; || $3==&quot;i&quot; || $3==&quot;j&quot; || $3==&quot;o&quot;)&#123;print $0&#125;&#125;&#x27;</span> filter.T18 &gt; class.T18</span><br><span class="line"></span><br><span class="line">$ <span class="built_in">wc</span> -l class.T18</span><br><span class="line">150458 class.T18</span><br><span class="line"><span class="comment">#获取剩余transcripts的exons位置信息，，计算class.T18的行数，提取序列并组装成转录本序列</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 获取剩余的transcripts的ID</span></span><br><span class="line">awk <span class="string">&#x27;&#123;print $5&#125;&#x27;</span> class.T18 &gt; filter1_transcript_ID</span><br><span class="line">$ <span class="built_in">wc</span> -l filter1_transcript_ID</span><br><span class="line">150458 filter1_transcript_ID</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">awk <span class="string">&#x27;NR==FNR&#123;a[$5]=1&#125; NR!=FNR&#123;if(a[substr($12,2,length($12)-3)]==1) print $0&#125;&#x27;</span> class.T18 T18.gtf &gt; T18class.gtf</span><br><span class="line"><span class="comment">#从参考基因组中提取序列</span></span><br><span class="line">gffread T18class.gtf -g ~/Sbicolor_454_v3.0.1.fa -w T18class.fa</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>filter1_by_uxijo.tmap——filter1_transcript_ID—— filter1_transcript.gtf</p>
<ol>
<li><strong>筛选潜在的lncRNA转录本</strong>。这可能需要基于转录本的特性（如长度、外显子数）以及编码潜</li>
</ol>
<p>力（使用CPC2、CNCI等工具）进行筛选。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">转录本编码能力预测，主要是4个软件，需要取交集：</span><br><span class="line"><span class="built_in">nohup</span> cpat.py -x ../dat/Human_Hexamer.tsv \</span><br><span class="line">-d ../dat/Human_logitModel.RData  \</span><br><span class="line">-g ~/lncRNA_project/07.identification/step2/filter2_transcript_exon.fa \</span><br><span class="line">-o ~/lncRNA_project/07.identification/step3/CPAT/cpat_result.txt &gt; cpat.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="built_in">nohup</span> Rscript ./LncFinder.R &gt; LncFinder.<span class="built_in">log</span> 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>lnc_finder功能 - 研发 (rdocumentation.org)<br><a target="_blank" rel="noopener" href="https://www.rdocumentation.org/packages/LncFinder/versions/1.1.5/topics/lnc_finder">https://www.rdocumentation.org/packages/LncFinder/versions/1.1.5/topics/lnc_finder</a><br>🤩 LncFinder | 非编码RNA的识别与分析神器！！！~ - 知乎</p>
<p>lncRNA组装流程的软件介绍之lncFinder-腾讯云开发者社区-腾讯云 (tencent.com)</p>
<p><a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1842771">https://cloud.tencent.com/developer/article/1842771</a></p>
<p>官方<br><a target="_blank" rel="noopener" href="https://cran.r-project.org/web/packages/LncFinder/index.html">https://cran.r-project.org/web/packages/LncFinder/index.html</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">nohup</span> python CNCI.py \</span><br><span class="line">-f ~/lncRNA_project/07.identification/step2/filter2_transcript_exon.fa \</span><br><span class="line">-o ~/lncRNA_project/07.identification/step3/CNCI \</span><br><span class="line">-m ve \</span><br><span class="line">-p 4 &gt; cnci.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="built_in">nohup</span> python PLEK.py \</span><br><span class="line">-fasta ~/lncRNA_project/07.identification/step2/filter2_transcript_exon.fa \</span><br><span class="line">-out ~/lncRNA_project/07.identification/step3/plek/plek \</span><br><span class="line">-thread 4 &gt; plek.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line">可能是这个原因吧</span><br><span class="line">PLek下载</span><br><span class="line">https://sourceforge.net/projects/plek/</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">python2 PLEK.py -fasta filter2_transcript_exon.fa -out plek -thread 4</span><br><span class="line">改下</span><br><span class="line">python PLEK.py -fasta merged_2.fa -out plek_merged_2 -thread 4</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/plek.png" class="" title="plek">


<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">less CPC2_result.txt|grep <span class="string">&#x27;noncoding&#x27;</span>|awk <span class="string">&#x27;&#123;print $1&#125;&#x27;</span>&gt; CPC2_id.txt</span><br><span class="line">$ <span class="built_in">wc</span> -l CPC2_id.txt</span><br><span class="line">55359 CPC2_id.txt</span><br><span class="line"></span><br><span class="line">less -S cpat_result.txt|awk <span class="string">&#x27;($6&lt;0.364)&#123;print $1&#125;&#x27;</span> &gt; cpat_id.txt</span><br><span class="line">sed -i <span class="string">&#x27;1d&#x27;</span> file</span><br><span class="line">$ <span class="built_in">wc</span> -l cpat_id.txt</span><br><span class="line">51956 cpat_id.txt</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">less lncFinder_result.txt|grep -w <span class="string">&#x27;NonCoding&#x27;</span> &gt; lncFinder_id.txt</span><br><span class="line">$ <span class="built_in">wc</span> -l lncFinder_id.txt</span><br><span class="line">52442 lncFinder_id.txt</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">less -S plek | grep -w <span class="string">&#x27;Non-coding&#x27;</span> |awk <span class="string">&#x27;&#123;print $3&#125;&#x27;</span> |sed <span class="string">&#x27;s/&gt;//g&#x27;</span> &gt; plek_id.txt</span><br><span class="line">$ <span class="built_in">wc</span> -l plek_id.txt </span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">55996 plek_id.txt</span><br><span class="line"></span><br><span class="line"><span class="comment"># 4个软件取交集</span></span><br><span class="line"><span class="built_in">cat</span> *txt |<span class="built_in">sort</span> |<span class="built_in">uniq</span>  -c |awk <span class="string">&#x27;&#123;if( $1==4)&#123;print&#125;&#125;&#x27;</span>|<span class="built_in">wc</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">&#123;% asset_img <span class="string">&quot;4个软件取交集.png&quot;</span> <span class="string">&quot;4个软件的交集&quot;</span>  %&#125;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">CNCI</span><br><span class="line">一共39060筛选出11059个nocoding</span><br><span class="line"></span><br><span class="line">CPC2</span><br><span class="line">一个39154筛选出11356个nocoding</span><br><span class="line"></span><br><span class="line">PLEK</span><br><span class="line"></span><br><span class="line">一个20935筛选出4097个nocoding</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%90%88%E5%B9%B6.png" class="" title="合并">
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%90%88%E5%B9%B62.png" class="" title="合并">

<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%A4%9A%E7%A7%8D%E6%96%B9%E6%B3%95.png" class="" title="mRNA">
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%A4%9A%E7%A7%8D%E6%96%B9%E6%B3%952.png" class="" title="lncRNA">



<p>最终筛选出21872个codeing RNA和4932个 nocoding<br>然后想得到这些lncRNA的fa文件。用tbtools的提取功能</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/tbtools2.png" class="" title="tbtools">




<p>最后得到的为lncRNA.fa</p>
<p>step4：比对到Pfam据库（我觉得filter3_by_noncoding_exon.fa可能是上面的交集的id转换来的）<br>filter3_by_noncoding_exon.fa——filter4_protein.fa——Pfam_scan.out——coding.ID，filter3_transcript_ID（未知，会不会是四个的合集啊？）——filter4_transcript_ID</p>
<p>比对到Pfam据库，过滤 (E-value &lt; 1e-5)</p>
<h4 id="pfem使用Transeq-将转录本序列翻译为6个可能的蛋白序列"><a href="#pfem使用Transeq-将转录本序列翻译为6个可能的蛋白序列" class="headerlink" title="pfem使用Transeq 将转录本序列翻译为6个可能的蛋白序列"></a>pfem使用Transeq 将转录本序列翻译为6个可能的蛋白序列</h4><p>transeq ../step3/intersection/filter3_by_noncoding_exon.fa filter4_protein.fa -frame=6</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/transeq.png" class="" title="transeq">




<p>conda install pfam_scan</p>
<p>nohup pfam_scan.pl -fasta ./filter4_protein.fa -dir /home/data/lihe/database/Pfam/ -out Pfam_scan.out &gt; Pfam_scan.log 2&gt;&amp;1 &amp;</p>
<p>pfam_scan.pl -fasta merged_2.fa -dir /home/dell/data/ghl/hmmer/hmmer-3.3.2 -outfile results_3.fa -as</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%A4%A7%E6%A6%82%E6%83%85%E5%86%B5.png" class="" title="大概情况">

<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%A4%A7%E6%A6%82%E6%83%85%E5%86%B52.png" class="" title="大概情况">


<p>transeq ../step3/merged_2.fa merged_2_protein.fa -frame=6<br>pfam_scan.pl -fasta merged_2_protein.fa -dir  -out Pfam_scan.out<br>我要暂停了，等不了了。。。<br>图：<br>pfam_scan.pl -fasta merged_2_protein.fa -dir . -out Pfam_scan.out<br>后来运行2天，能得到</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E4%BB%A3%E7%A0%81%E8%A7%A3%E9%87%8A.png" class="" title="代码解释">


<h3 id="过滤-E-value-lt-1e-5"><a href="#过滤-E-value-lt-1e-5" class="headerlink" title="过滤 (E-value &lt; 1e-5)"></a>过滤 (E-value &lt; 1e-5)</h3><p>grep -v ‘^#’ Pfam_scan.out | grep -v ‘^\s*$’ | awk ‘($13&lt; 1e-5){print $1}’| awk -F “_” ‘{print$1}’ | sort | uniq &gt; coding.ID</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/pfan.png" class="" title="过滤">



<p>wc coding.ID<br> 3751 </p>
<h3 id="反向过滤"><a href="#反向过滤" class="headerlink" title="反向过滤"></a>反向过滤</h3><p>grep -v -f coding.ID ../step3/intersection/filter3_transcript_ID &gt; filter4_transcript_ID</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%8F%8D%E5%90%91%E8%BF%87%E6%BB%A4.png" class="" title="反向过滤">





<p>dos2unix filter4_transcript_ID.txt</p>
<p>grep -Ff transcript_ID -w Gallus_gallus.GRCg6a.98.gtf &gt; lncRNA.gtf</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/%E5%BE%97%E5%88%B0lncRNA.png" class="" title="lncRNA.gtf">










<ol start="2">
<li><strong>注释和进一步的分析</strong>。将筛选出的lncRNA候选与已知的lncRNA数据库进行比对，如NONCODE或ALDB，以确定新发现的lncRNA。</li>
</ol>
<p>这个过程可能涉及到多个自定义脚本和具体的分析策略，具体取决于您的研究目标和可用资源。</p>
<h3 id="注意："><a href="#注意：" class="headerlink" title="注意："></a>注意：</h3><ul>
<li>上述步骤中的文件路径、索引名称和软件参数都需要根据您的实际环境进行调整。</li>
<li>某些步骤（如lncRNA的筛选和注释）需要您根据具体的研究需求来定制分析策略和使用相应的工具。</li>
<li>确保所有使用的工具都已正确安装，并且您有适当的权限执行这些命令。</li>
</ul>
<h2 id="第二种，lncRNA检测工作流程"><a href="#第二种，lncRNA检测工作流程" class="headerlink" title="第二种，lncRNA检测工作流程"></a>第二种，lncRNA检测工作流程</h2><p>测序数据分析 干净数据使用Tophat2 [19]程序按照以下参数映射到Gallus gallus参考基因组(gal6) 长非编码RNA (lncRNA) 的鉴定与注释在RNA-seq数据分析中是至关重要的一环，它分为两大部分：一个是处理通用RNA-seq数据，另一个是专门针对lncRNA的注释和分析。</p>
<h2 id="0-前期准备："><a href="#0-前期准备：" class="headerlink" title="0,前期准备："></a>0,前期准备：</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">下载方法</span><br><span class="line">prefetch SRR4046839</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">fastq-dump --outdir /home/data1/Ghl/all --split-files SRR6116749.sra</span><br><span class="line"></span><br></pre></td></tr></table></figure>


<h1 id="通用RNA-seq数据处理流程："><a href="#通用RNA-seq数据处理流程：" class="headerlink" title="通用RNA-seq数据处理流程："></a>通用RNA-seq数据处理流程：</h1><h2 id="1，质量控制-使用FastQC软件来生成初始序列数据的质量报告。"><a href="#1，质量控制-使用FastQC软件来生成初始序列数据的质量报告。" class="headerlink" title="1，质量控制:使用FastQC软件来生成初始序列数据的质量报告。"></a>1，质量控制:使用FastQC软件来生成初始序列数据的质量报告。</h2><p>按照3.0 转录本比对组装HISAT2+StringTie+cuffcompare - 简书 (jianshu.com)进行质控<br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/2b6aba1263d1">https://www.jianshu.com/p/2b6aba1263d1</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">fastp --detect_adapter_for_pe -l 100 -q 19 -u 50 -i ~/CK18_1.fq.gz -o CK18c_1.fq.gz -I ~/CK18_2.fq.gz -O CK18c_2.fq.gz \</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">fastp --detect_adapter_for_pe -l 100 -q 19 -u 50 -i /home/dell/data/ghl/SRR1_2/SRR6116775_1.fastq -o /home/dell/data/ghl/SRR1_2/CK18c1_1.fastq \</span><br><span class="line">-I /home/dell/data/ghl/SRR1_2/SRR6116775_2.fastq -O /home/dell/data/ghl/SRR1_2/CK18c1_2.fastq \</span><br><span class="line">-f 5</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>步骤：<br>（0）perl 手动去除第一行index六个碱基<br>（1）去除低质量的reads（质量值Q≤19的碱基占总碱基的50%以上）；<br>（2）默认参数滑窗去掉质量低的碱基，最终去除长度小于100的reads<br>（3）去除含N比例大于5%的Reads；<br>（4）去除与核糖体RNA（rRNA）匹配的Reads #sortmerna<br>（5）质量评估：fastqc</p>
<p>下面部分是我修改的代码</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">fastp --detect_adapter_for_pe -l 100 -q 19 -u 50 -i /home/dell/data/ghl/SRR1_2/SRR6116775_1.fastq -o /home/dell/data/ghl/SRR1_2/CK18c_1.fastq \</span><br><span class="line">-I /home/dell/data/ghl/SRR1_2/SRR6116775_2.fastq -O /home/dell/data/ghl/SRR1_2/CK18c_2.fastq</span><br></pre></td></tr></table></figure>

<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/3.png" class="" title="质量控制">


<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/4.png" class="" title="质量控制">


<p>数据过滤（这个要做吗）<br>并且使用Fastp工具对输入的双端测序数据进行质量控制和适配器修剪。<br>步骤：<br>（0）perl 手动去除第一行index六个碱基<br>（1）去除低质量的reads（质量值Q≤19的碱基占总碱基的50%以上）；<br>（2）默认参数滑窗去掉质量低的碱基，最终去除长度小于100的reads<br>（3）去除含N比例大于5%的Reads；<br>（4）去除与核糖体RNA（rRNA）匹配的Reads #sortmerna<br>（5）质量评估：fastqc</p>
<p>#工具：选一种即可</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">conda install -c trimmomatic</span><br><span class="line">conda install fastp</span><br><span class="line"><span class="comment">#参数：</span></span><br><span class="line">fastp --detect_adapter_for_pe -l 100 -q 19 -u 50 -i ~/CK18_1.fq.gz -o CK18c_1.fq.gz -I ~/CK18_2.fq.gz -O CK18c_2.fq.gz \</span><br><span class="line"><span class="comment">#或者：</span></span><br><span class="line">trimmomatic PE -phred33 CK18_1.fq.gz CK18_2.fq.gz CK18_1paired.fq.gz CK18_1unpaired.fq.gz CK18_2paired.fq.gz CK18_2unpaired.fq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:19 MINLEN:50</span><br><span class="line"><span class="comment">#质量评估：fastqc</span></span><br><span class="line">fastqc -o ./ -t 4 ./ CK18c_11.fq.gz</span><br><span class="line"></span><br></pre></td></tr></table></figure>















<h2 id="2，序列预处理-借助Trimmomatic工具-20-对原始读取进行清洗，这包括移除来自测序适配器的污染物和修剪低质量的读取。常规的读取质量分数阈值是20。"><a href="#2，序列预处理-借助Trimmomatic工具-20-对原始读取进行清洗，这包括移除来自测序适配器的污染物和修剪低质量的读取。常规的读取质量分数阈值是20。" class="headerlink" title="2，序列预处理:借助Trimmomatic工具[20]对原始读取进行清洗，这包括移除来自测序适配器的污染物和修剪低质量的读取。常规的读取质量分数阈值是20。"></a>2，序列预处理:借助Trimmomatic工具[20]对原始读取进行清洗，这包括移除来自测序适配器的污染物和修剪低质量的读取。常规的读取质量分数阈值是20。</h2><h2 id="3，序列映射-使用如Tophat2-19-、Bowtie2-21-或HISAT2-22-等工具将预处理后的读取映射到参考基因组。"><a href="#3，序列映射-使用如Tophat2-19-、Bowtie2-21-或HISAT2-22-等工具将预处理后的读取映射到参考基因组。" class="headerlink" title="3，序列映射:使用如Tophat2 [19]、Bowtie2 [21]或HISAT2 [22]等工具将预处理后的读取映射到参考基因组。"></a>3，序列映射:使用如Tophat2 [19]、Bowtie2 [21]或HISAT2 [22]等工具将预处理后的读取映射到参考基因组。</h2><h2 id="4，转录本组装-通常使用Cufflinks-23-或StringTie-24-来进行序列读取的组装。"><a href="#4，转录本组装-通常使用Cufflinks-23-或StringTie-24-来进行序列读取的组装。" class="headerlink" title="4，转录本组装:通常使用Cufflinks [23]或StringTie [24]来进行序列读取的组装。"></a>4，转录本组装:通常使用Cufflinks [23]或StringTie [24]来进行序列读取的组装。</h2><p>cuffcompare -r ~/sbi311gx.gtf -s ~/Sbicolor_454_v3.0.1.fa -o cufcomp ../T18.gtf </p>
<p>awk ‘{if($7&gt;=0.5 &amp;&amp; $10 &gt; 1 &amp;&amp; $11 &gt;200) print $0}’ cufcomp.T18.gtf.tmap &gt; filter.T18</p>
<p>awk ‘{if ($3==”u” || $3==”x” || $3==”i” || $3==”j” || $3==”o”){print $0}}’ filter.T18 &gt; class.T18</p>
<p>awk ‘NR==FNR{a[$5]=1} NR!=FNR{if(a[substr($12,2,length($12)-3)]==1) print $0}’ class.T18 T18.gtf &gt; T18class.gtf</p>
<p>#从参考基因组中提取序列</p>
<p>gffread T18class.gtf -g ~/Sbicolor_454_v3.0.1.fa -w T18class.fa</p>
<p>最后的<br>awk ‘{if($7&gt;=0.5 &amp;&amp; $10 &gt; 1 &amp;&amp; $11 &gt;200) print $0}’ /home/data1/Ghl/lab4046839/TSSR/TSRR6116778/6116778.TSRR6116778.gtf.tmap &gt; filter.6116778</p>
<p>到目前为止没有问题<br>1，merged.gtf没问题,<br>2，合并FPKM，没问题，并且所有的都合并了<br>3，tmap没问题，并且都生成了个</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/awk.png" class="" title="图4">



<p>(6116778.TSRR6116778.gtf.tmap– filter.6116778)<br>awk ‘{if($7&gt;=0.5 &amp;&amp; $10 &gt; 1 &amp;&amp; $11 &gt;200) print $0}’ /home/data1/Ghl/lab4046839/TSSR/TSRR6116778/6116778.TSRR6116778.gtf.tmap &gt; filter.6116778</p>
<p>4，筛选tmap（filter.6116749– class.T6116749）<br>awk ‘{if ($3==”u” || $3==”x” || $3==”i” || $3==”j” || $3==”o”){print $0}}’ filter.6116749 &gt; class.T6116749</p>
<p>5，又得到gtf（TSRR6116749.gtf与class.T6116749– T6116749class.gtf）<br>awk ‘NR==FNR{a[$5]=1} NR!=FNR{if(a[substr($12,2,length($12)-3)]==1) print $0}’ class.T18 T18.gtf &gt; T18class.gtf</p>
<p>下面这个代码的运行后<br>awk ‘NR==FNR{a[$5]=1} NR!=FNR{if(a[substr($12,2,length($12)-3)]==1) print $0}’ class.T6116749 /home/data1/Ghl/lab4046839/TSSR/TSRR6116749/TSRR6116749.gtf &gt; T6116749class.gtf</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/awk2.png" class="" title="awk">

<p>awk ‘{if ($3==”u” || $3==”x” || $3==”i” || $3==”j” || $3==”o”){print $0}}’ filter.6116778 &gt; class.T6116778</p>
<p>又是每一个都保存</p>
<p>6，得到fa文件（T6116749class.gtf- T6116749class.fa）<br>gffread T6116749class.gtf -g /home/data1/Ghl/lab4046839/Gallus_gallus.GRCg6a.dna.toplevel.fa -w T6116749class.fa</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/class.gtf.png" class="" title="awk">

<p>7，<br>7.1，fa文件合并到一起<br>cat *.fa | sort -u &gt; merged.fa</p>
<p>专门针对lncRNA的分析流程：<br>1，转录本筛选:使用Cufflinks的Cuffcompare功能，筛选出那些与参考基因组中的蛋白编码区域不重叠的新转录本，如此筛选出的转录本不包括已经注释过的lncRNA。<br>2，lncRNA候选过滤:潜在的lncRNA候选转录本会经过一系列过滤步骤。这包括按长度、按序列内容、按表达水平和按蛋白编码潜力进行过滤。为了实现这些过滤，会使用到如Transeq、HMME、PfamScan、BLASTX、RefSeq、UniRef90、CPC [25]、CNCI [26]和PLEK [27],CPAT [28]等工具和数据库。<br>3，数据整合:将新鉴定的lncRNA与已知的lncRNA数据库，例如ALDB [29]或NONCODE进行合并。这一步通常涉及根据序列同一性或其他标准来确定是否认为一个lncRNA是“已知”的。<br>4， 下游分析:这取决于具体的研究目的。常见的策略包括使用DESeq2 [30]比较lncRNA的表达、识别lncRNA的潜在靶基因，以及利用如DAVID、KOBAS、Gene Ontology或Reactome数据库的工具进行功能注释。</p>
<h2 id="我的批量运行的代码，有的在服务器上，以后补充"><a href="#我的批量运行的代码，有的在服务器上，以后补充" class="headerlink" title="我的批量运行的代码，有的在服务器上，以后补充"></a>我的批量运行的代码，有的在服务器上，以后补充</h2><h2 id="第九步9-stringtie-batch-sh-txt"><a href="#第九步9-stringtie-batch-sh-txt" class="headerlink" title="第九步9_stringtie_batch.sh.txt"></a>第九步9_stringtie_batch.sh.txt</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"><span class="comment">#!/bin/bash</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 您提供的SRR IDs列表</span></span><br><span class="line"><span class="built_in">declare</span> -a SRR_ids=(</span><br><span class="line"><span class="string">&quot;SRR6116749&quot;</span> <span class="string">&quot;SRR6116750&quot;</span> <span class="string">&quot;SRR6116751&quot;</span> <span class="string">&quot;SRR6116752&quot;</span> <span class="string">&quot;SRR6116753&quot;</span> <span class="string">&quot;SRR6116754&quot;</span> </span><br><span class="line"><span class="string">&quot;SRR6116761&quot;</span> <span class="string">&quot;SRR6116762&quot;</span> <span class="string">&quot;SRR6116763&quot;</span> <span class="string">&quot;SRR6116764&quot;</span> <span class="string">&quot;SRR6116765&quot;</span> <span class="string">&quot;SRR6116766&quot;</span> </span><br><span class="line"><span class="string">&quot;SRR6116773&quot;</span> <span class="string">&quot;SRR6116774&quot;</span> <span class="string">&quot;SRR6116775&quot;</span> <span class="string">&quot;SRR6116776&quot;</span> <span class="string">&quot;SRR6116777&quot;</span> <span class="string">&quot;SRR6116778&quot;</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 合并GTF文件的路径</span></span><br><span class="line">MERGED_GTF=<span class="string">&quot;/path_to/mergelist.gtf&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 输入和输出文件夹的路径</span></span><br><span class="line">BAM_DIR=<span class="string">&quot;/home/data1/Ghl/GHL/short_bam&quot;</span></span><br><span class="line">OUT_DIR=<span class="string">&quot;/home/data1/Ghl/GHL/_e_gtf&quot;</span>  <span class="comment"># 指定输出文件夹，已经按您提供的路径调整</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 确保输出目录存在</span></span><br><span class="line"><span class="built_in">mkdir</span> -p <span class="string">&quot;<span class="variable">$&#123;OUT_DIR&#125;</span>&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 循环处理每一个SRR ID</span></span><br><span class="line"><span class="keyword">for</span> srr <span class="keyword">in</span> <span class="string">&quot;<span class="variable">$&#123;SRR_ids[@]&#125;</span>&quot;</span></span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Processing <span class="variable">$&#123;srr&#125;</span> with StringTie...&quot;</span></span><br><span class="line">    stringtie -e -B -p 16 -G <span class="string">&quot;<span class="variable">$&#123;MERGED_GTF&#125;</span>&quot;</span> -o <span class="string">&quot;<span class="variable">$&#123;OUT_DIR&#125;</span>/e_<span class="variable">$&#123;srr&#125;</span>.gtf&quot;</span> <span class="string">&quot;<span class="variable">$&#123;BAM_DIR&#125;</span>/<span class="variable">$&#123;srr&#125;</span>.short.bam&quot;</span></span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;StringTie processing done for all samples!&quot;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>



<h2 id="第10步，10-cuffcompare-sh-txt"><a href="#第10步，10-cuffcompare-sh-txt" class="headerlink" title="第10步，10_cuffcompare.sh.txt"></a>第10步，10_cuffcompare.sh.txt</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#!/bin/bash</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 您提供的SRR IDs列表</span></span><br><span class="line"><span class="built_in">declare</span> -a SRR_ids=(</span><br><span class="line"><span class="string">&quot;SRR6116749&quot;</span> <span class="string">&quot;SRR6116750&quot;</span> <span class="string">&quot;SRR6116751&quot;</span> <span class="string">&quot;SRR6116752&quot;</span> <span class="string">&quot;SRR6116753&quot;</span> <span class="string">&quot;SRR6116754&quot;</span> </span><br><span class="line"><span class="string">&quot;SRR6116761&quot;</span> <span class="string">&quot;SRR6116762&quot;</span> <span class="string">&quot;SRR6116763&quot;</span> <span class="string">&quot;SRR6116764&quot;</span> <span class="string">&quot;SRR6116765&quot;</span> <span class="string">&quot;SRR6116766&quot;</span> </span><br><span class="line"><span class="string">&quot;SRR6116773&quot;</span> <span class="string">&quot;SRR6116774&quot;</span> <span class="string">&quot;SRR6116775&quot;</span> <span class="string">&quot;SRR6116776&quot;</span> <span class="string">&quot;SRR6116777&quot;</span> <span class="string">&quot;SRR6116778&quot;</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 参考GTF和参考基因组序列的路径</span></span><br><span class="line">REF_GTF=<span class="string">&quot;/home/data1/Ghl/GHL/Gallus_gallus.GRCg6a.98.gtf&quot;</span></span><br><span class="line">REF_SEQ=<span class="string">&quot;/home/data1/Ghl/GHL/Gallus_gallus.GRCg6a.dna.toplevel.fa&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 输入文件夹的路径，每个SRR的GTF文件都在自己的子文件夹下</span></span><br><span class="line">INPUT_DIR=<span class="string">&quot;/home/data1/Ghl/GHL/_e_gtf&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 输出文件夹的路径</span></span><br><span class="line">OUTPUT_DIR=<span class="string">&quot;/home/data1/Ghl/GHL/10_cuffcompare&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 确保输出目录存在</span></span><br><span class="line"><span class="built_in">mkdir</span> -p <span class="string">&quot;<span class="variable">$&#123;OUTPUT_DIR&#125;</span>&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 循环处理每一个SRR ID</span></span><br><span class="line"><span class="keyword">for</span> srr <span class="keyword">in</span> <span class="string">&quot;<span class="variable">$&#123;SRR_ids[@]&#125;</span>&quot;</span></span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">    <span class="comment"># 为每个SRR_id创建一个单独的输出文件夹</span></span><br><span class="line">    SRR_OUT_DIR=<span class="string">&quot;<span class="variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="variable">$&#123;srr&#125;</span>&quot;</span></span><br><span class="line">    <span class="built_in">mkdir</span> -p <span class="string">&quot;<span class="variable">$&#123;SRR_OUT_DIR&#125;</span>&quot;</span></span><br><span class="line">    </span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Running cuffcompare for <span class="variable">$&#123;srr&#125;</span>...&quot;</span></span><br><span class="line">    cuffcompare -r <span class="string">&quot;<span class="variable">$&#123;REF_GTF&#125;</span>&quot;</span> -s <span class="string">&quot;<span class="variable">$&#123;REF_SEQ&#125;</span>&quot;</span> -o <span class="string">&quot;<span class="variable">$&#123;SRR_OUT_DIR&#125;</span>/<span class="variable">$&#123;srr&#125;</span>&quot;</span> <span class="string">&quot;<span class="variable">$&#123;INPUT_DIR&#125;</span>/<span class="variable">$&#123;srr&#125;</span>/<span class="variable">$&#123;srr&#125;</span>.gtf&quot;</span></span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;Cuffcompare processing done for all samples!&quot;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>



<h2 id="11"><a href="#11" class="headerlink" title="11"></a>11</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"><span class="comment">#!/bin/bash</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 您提供的SRR IDs列表</span></span><br><span class="line"><span class="built_in">declare</span> -a SRR_ids=(</span><br><span class="line"><span class="string">&quot;SRR6116749&quot;</span> <span class="string">&quot;SRR6116750&quot;</span> <span class="string">&quot;SRR6116751&quot;</span> <span class="string">&quot;SRR6116752&quot;</span> <span class="string">&quot;SRR6116753&quot;</span> <span class="string">&quot;SRR6116754&quot;</span> </span><br><span class="line"><span class="string">&quot;SRR6116761&quot;</span> <span class="string">&quot;SRR6116762&quot;</span> <span class="string">&quot;SRR6116763&quot;</span> <span class="string">&quot;SRR6116764&quot;</span> <span class="string">&quot;SRR6116765&quot;</span> <span class="string">&quot;SRR6116766&quot;</span> </span><br><span class="line"><span class="string">&quot;SRR6116773&quot;</span> <span class="string">&quot;SRR6116774&quot;</span> <span class="string">&quot;SRR6116775&quot;</span> <span class="string">&quot;SRR6116776&quot;</span> <span class="string">&quot;SRR6116777&quot;</span> <span class="string">&quot;SRR6116778&quot;</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 基础GTF文件夹路径</span></span><br><span class="line">BASE_GTF_DIR=<span class="string">&quot;/home/data1/Ghl/GHL/_e_gtf&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 基础输出文件夹路径</span></span><br><span class="line">BASE_OUT_DIR=<span class="string">&quot;/home/data1/Ghl/GHL/11_gffread&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 参考基因组序列的路径</span></span><br><span class="line">REF_SEQ=<span class="string">&quot;~/Sbicolor_454_v3.0.1.fa&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 确保输出目录存在</span></span><br><span class="line"><span class="built_in">mkdir</span> -p <span class="string">&quot;<span class="variable">$&#123;BASE_OUT_DIR&#125;</span>&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 循环处理每一个SRR ID</span></span><br><span class="line"><span class="keyword">for</span> srr <span class="keyword">in</span> <span class="string">&quot;<span class="variable">$&#123;SRR_ids[@]&#125;</span>&quot;</span></span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Processing <span class="variable">$&#123;srr&#125;</span>...&quot;</span></span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 定义输入文件路径</span></span><br><span class="line">    TMAP_FILE=<span class="string">&quot;<span class="variable">$&#123;BASE_GTF_DIR&#125;</span>/<span class="variable">$&#123;srr&#125;</span>/<span class="variable">$&#123;srr&#125;</span>.gtf.tmap&quot;</span></span><br><span class="line">    GTF_FILE=<span class="string">&quot;<span class="variable">$&#123;BASE_GTF_DIR&#125;</span>/<span class="variable">$&#123;srr&#125;</span>/<span class="variable">$&#123;srr&#125;</span>.gtf&quot;</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># 定义输出文件的前缀路径</span></span><br><span class="line">    PREFIX=<span class="string">&quot;<span class="variable">$&#123;BASE_OUT_DIR&#125;</span>/<span class="variable">$&#123;srr&#125;</span>&quot;</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># 过滤tmap文件</span></span><br><span class="line">    awk <span class="string">&#x27;&#123;if($7&gt;=0.5 &amp;&amp; $10 &gt; 1 &amp;&amp; $11 &gt;200) print $0&#125;&#x27;</span> <span class="string">&quot;<span class="variable">$&#123;TMAP_FILE&#125;</span>&quot;</span> &gt; <span class="string">&quot;<span class="variable">$&#123;PREFIX&#125;</span>_filter.T18&quot;</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># 提取特定分类的转录本</span></span><br><span class="line">    awk <span class="string">&#x27;&#123;if ($3==&quot;u&quot; || $3==&quot;x&quot; || $3==&quot;i&quot; || $3==&quot;j&quot; || $3==&quot;o&quot;)&#123;print $0&#125;&#125;&#x27;</span> <span class="string">&quot;<span class="variable">$&#123;PREFIX&#125;</span>_filter.T18&quot;</span> &gt; <span class="string">&quot;<span class="variable">$&#123;PREFIX&#125;</span>_class.T18&quot;</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># 根据class文件从GTF中提取相关转录本</span></span><br><span class="line">    awk <span class="string">&#x27;NR==FNR&#123;a[$5]=1&#125; NR!=FNR&#123;if(a[substr($12,2,length($12)-3)]==1) print $0&#125;&#x27;</span> <span class="string">&quot;<span class="variable">$&#123;PREFIX&#125;</span>_class.T18&quot;</span> <span class="string">&quot;<span class="variable">$&#123;GTF_FILE&#125;</span>&quot;</span> &gt; <span class="string">&quot;<span class="variable">$&#123;PREFIX&#125;</span>_T18class.gtf&quot;</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># 从参考基因组中提取序列</span></span><br><span class="line">    gffread <span class="string">&quot;<span class="variable">$&#123;PREFIX&#125;</span>_T18class.gtf&quot;</span> -g <span class="string">&quot;<span class="variable">$&#123;REF_SEQ&#125;</span>&quot;</span> -w <span class="string">&quot;<span class="variable">$&#123;PREFIX&#125;</span>_T18class.fa&quot;</span></span><br><span class="line"></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Processing for <span class="variable">$&#123;srr&#125;</span> done!&quot;</span></span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;All processing done!&quot;</span></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>











<h2 id="12-cpc2-sh-1-txt"><a href="#12-cpc2-sh-1-txt" class="headerlink" title="12_cpc2.sh(1).txt"></a>12_cpc2.sh(1).txt</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"><span class="comment">#!/bin/bash</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 输入文件夹路径</span></span><br><span class="line">INPUT_DIR=<span class="string">&quot;/home/data1/Ghl/GHL/11_gffread&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 输出文件夹路径</span></span><br><span class="line">OUTPUT_DIR=<span class="string">&quot;/home/data1/Ghl/GHL/CPC_CNCI_CPAT/cpc2&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 确保输出目录存在</span></span><br><span class="line"><span class="built_in">mkdir</span> -p <span class="string">&quot;<span class="variable">$&#123;OUTPUT_DIR&#125;</span>&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 循环遍历输入目录下所有_Tclass.fa文件</span></span><br><span class="line"><span class="keyword">for</span> input_file <span class="keyword">in</span> <span class="variable">$&#123;INPUT_DIR&#125;</span>/*_Tclass.fa; <span class="keyword">do</span></span><br><span class="line">    <span class="comment"># 获取不带路径和扩展名的基础文件名</span></span><br><span class="line">    base_filename=$(<span class="built_in">basename</span> <span class="string">&quot;<span class="variable">$&#123;input_file&#125;</span>&quot;</span> _Tclass.fa)</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 定义输出文件路径</span></span><br><span class="line">    output_file=<span class="string">&quot;<span class="variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="variable">$&#123;base_filename&#125;</span>_cpc2_output.txt&quot;</span></span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 运行CPC2</span></span><br><span class="line">    python3 CPC2.py -i <span class="string">&quot;<span class="variable">$&#123;input_file&#125;</span>&quot;</span> -o <span class="string">&quot;<span class="variable">$&#123;output_file&#125;</span>&quot;</span></span><br><span class="line">    </span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;CPC2 analysis done for <span class="variable">$&#123;base_filename&#125;</span>&quot;</span></span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;All CPC2 processing done!&quot;</span></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>













<h2 id="把所有转录本注释分类为mRNA-lncRNA、新的鉴定的lncRNA"><a href="#把所有转录本注释分类为mRNA-lncRNA、新的鉴定的lncRNA" class="headerlink" title="把所有转录本注释分类为mRNA\lncRNA、新的鉴定的lncRNA"></a>把所有转录本注释分类为mRNA\lncRNA、新的鉴定的lncRNA</h2><p><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/b6c40e9612c4">https://www.jianshu.com/p/b6c40e9612c4</a></p>
<h1 id="WGCNA"><a href="#WGCNA" class="headerlink" title="WGCNA"></a>WGCNA</h1><h2 id="count"><a href="#count" class="headerlink" title="count"></a>count</h2><p>图：featureCounts -T 8 -a /home/data1/Ghl/bam/lncRNA.gtf  -o lncRNA_raw_count.txt -p -B -C -f -t transcript -g transcript_id /home/data1/Ghl/bam/*.short.bam<br>生成了lncRNA_raw_count.txt</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/featureCounts.png" class="" title="featureCounts">
<p>您提供的命令是使用 <code>featureCounts</code> 工具来计算基因表达量，特别是在这个上下文中，它被用来对长非编码RNA（lncRNA）进行计数。<code>featureCounts</code> 是 <code>Subread</code> 软件包的一部分，广泛用于对RNA-Seq数据进行定量分析。</p>
<p>下面是该命令中每个参数的解释：</p>
<ul>
<li><p><code>featureCounts</code>: 这是执行定量的主命令。</p>
</li>
<li><p><code>-T 8</code>: 指定要使用的线程数，这里设置为8，意味着<code>featureCounts</code>会并行运行以加快处理速度。</p>
</li>
<li><p><code>-a /home/data1/Ghl/bam/lncRNA.gtf</code>: <code>-a</code> 参数后跟注释文件的路径，这里指定了包含lncRNA注释信息的GTF文件。</p>
</li>
<li><p><code>-o lncRNA_raw_count.txt</code>: <code>-o</code> 参数指定输出文件的名称，这里的输出文件将包含原始的计数数据。</p>
</li>
<li><p><code>-p</code>: 这个参数告诉 <code>featureCounts</code> 输入的是配对端测序数据。</p>
</li>
<li><p><code>-B</code>: 要求计数时只考虑每对reads中都成功比对的情况。</p>
</li>
<li><p><code>-C</code>: 不计算跨越基因间区域的read对。</p>
</li>
<li><p><code>-f</code>: 使用<code>-f</code>参数意味着计数是基于特征（例如外显子）而不是基于元注释的（如转录本或基因），这对于lncRNA这样的非编码RNA特别有用。</p>
</li>
<li><p><code>-t transcript</code>: <code>-t</code> 参数指定计数时考虑的注释特征类型，这里设置为“transcript”，意味着计数将基于转录本进行。</p>
</li>
<li><p><code>-g transcript_id</code>: <code>-g</code> 参数指定用于分组的特征属性，这里使用“transcript_id”表示计数将根据每个转录本的ID来汇总。</p>
</li>
<li><p><code>/home/data1/Ghl/bam/*.short.bam</code>: 这是输入文件的路径，<code>*.short.bam</code> 表示该命令将处理指定路径下所有以<code>.short.bam</code>结尾的文件。</p>
</li>
</ul>
<p>综上所述，这个 <code>featureCounts</code> 命令将对所有在指定路径下的BAM文件中的lncRNA进行计数，使用多线程来加速处理，并将结果输出到一个文本文件中。这个结果文件可以用于后续的差异表达分析或其他类型的生物信息学分析。</p>
<h2 id="FPKEM"><a href="#FPKEM" class="headerlink" title="FPKEM"></a>FPKEM</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">2,R代码</span><br><span class="line"><span class="built_in">rm</span>(list=<span class="built_in">ls</span>())</span><br><span class="line"><span class="comment"># make count table </span></span><br><span class="line">raw_df &lt;- read.table(file = <span class="string">&quot;lncRNA_raw_count.txt&quot;</span>,header = T,skip = 1,sep = <span class="string">&quot;\t&quot;</span>)</span><br><span class="line"></span><br><span class="line">count_df &lt;- raw_df[ ,c(7:ncol(raw_df))]</span><br><span class="line">metadata &lt;- raw_df[ ,1:6] <span class="comment"># 提取基因信息count数据前的几列</span></span><br><span class="line">rownames(count_df) &lt;- as.character(raw_df[,1])</span><br><span class="line">colnames(count_df) &lt;- paste0(<span class="string">&quot;SRR10744&quot;</span>,1:18)</span><br><span class="line"></span><br><span class="line"><span class="comment"># calculate FPKM</span></span><br><span class="line">countToFpkm &lt;- <span class="keyword">function</span>(counts, effLen)</span><br><span class="line">&#123;</span><br><span class="line">  N &lt;- colSums(counts)</span><br><span class="line">  exp( <span class="built_in">log</span>(counts) + <span class="built_in">log</span>(1e9) - <span class="built_in">log</span>(effLen) - <span class="built_in">log</span>(N) )</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">options(scipen = 200) <span class="comment"># 表示在200个数字以内都不使用科学计数法</span></span><br><span class="line">fpkm = countToFpkm(count_df, metadata<span class="variable">$Length</span>)</span><br><span class="line"><span class="comment"># View(fpkm)</span></span><br><span class="line"></span><br><span class="line">write.table(fpkm, file = <span class="string">&quot;fpkm_data.txt&quot;</span>, sep = <span class="string">&quot;\t&quot;</span>, quote = FALSE)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># FPKM &gt; 0 in at least one sample</span></span><br><span class="line">count_df.filter &lt;- count_df[rowSums(fpkm)&gt;0,]</span><br><span class="line"></span><br><span class="line"><span class="comment">#write.table(rownames(count_df.filter,file=&quot;filter6_by_fpkm_id&quot;, sep=&quot;\t&quot;,quote=F))</span></span><br><span class="line">write.table(rownames(count_df.filter), file = <span class="string">&quot;filter6_by_fpkm_id&quot;</span>, sep = <span class="string">&quot;\t&quot;</span>, quote = FALSE)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/progrem2.png" class="" title="progrem">


<h3 id="得到fpkm-data-txt，进行WGCNA"><a href="#得到fpkm-data-txt，进行WGCNA" class="headerlink" title="得到fpkm_data.txt，进行WGCNA"></a>得到fpkm_data.txt，进行WGCNA</h3><p>把txt,转为xlsx,在转txt后没报错了</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/WGCNA4.png" class="" title="WGCNA">


<p>一共fpkm_data.txt有5439行，但很多是零，大概删去0的，有2987个</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/WGCNA5.png" class="" title="WGCNA">

<p>WGCNA筛选2000个</p>
<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/WGCNA6.png" class="" title="WGCNA">

<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/WGCNA7.png" class="" title="WGCNA">

<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/WGCNA8.png" class="" title="WGCNA">

<img src="/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/WGCNA9.png" class="" title="WGCNA">







































































<h2 id="lncRNA分析-第三种"><a href="#lncRNA分析-第三种" class="headerlink" title="lncRNA分析,第三种"></a>lncRNA分析,第三种</h2><p><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/b6c40e9612c4">https://www.jianshu.com/p/b6c40e9612c4</a></p>
<p>为了处理和压缩高通量测序数据，您可以按照以下步骤操作：</p>
<p>在您提到的场景中，下载数据（如<code>SRR4046839</code>）通常是使用NCBI的SRA Toolkit中的<code>prefetch</code>命令的第一步。在此之后，您需要通过多个步骤来处理和分析这些数据，直至达到上述步骤中提到的质控和数据清洗。下面是从下载<code>SRR4046839</code>到执行质控脚本的完整步骤：</p>
<h2 id="1-下载数据"><a href="#1-下载数据" class="headerlink" title="1. 下载数据"></a>1. 下载数据</h2><p>使用<code>prefetch</code>命令从NCBI下载SRA文件：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">prefetch SRR4046839</span><br></pre></td></tr></table></figure>

<p>这个命令会下载<code>SRR4046839</code>的SRA文件到本地。</p>
<h2 id="2-转换SRA文件到FASTQ格式"><a href="#2-转换SRA文件到FASTQ格式" class="headerlink" title="2. 转换SRA文件到FASTQ格式"></a>2. 转换SRA文件到FASTQ格式</h2><p>下载完SRA文件后，使用<code>fastq-dump</code>（SRA Toolkit的一部分）或更现代的工具如<code>fasterq-dump</code>来转换SRA文件为FASTQ格式。<code>fasterq-dump</code>比<code>fastq-dump</code>更快且更高效：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fasterq-dump SRR4046839 --split-files -O /path/to/output/directory</span><br></pre></td></tr></table></figure>

<p>这个命令会生成两个FASTQ文件（如果是双端数据的话），分别对应每个读段的数据，例如<code>SRR4046839_1.fastq</code>和<code>SRR4046839_2.fastq</code>，并将这些文件保存到指定的输出目录中。</p>
<h2 id="3-准备质控分析"><a href="#3-准备质控分析" class="headerlink" title="3. 准备质控分析"></a>3. 准备质控分析</h2><h4 id="创建数据目录"><a href="#创建数据目录" class="headerlink" title="创建数据目录"></a>创建数据目录</h4><p>创建一个存储清洗后数据的目录：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> cleandata</span><br></pre></td></tr></table></figure>

<h4 id="生成文件列表-等等就可以了"><a href="#生成文件列表-等等就可以了" class="headerlink" title="生成文件列表,等等就可以了"></a>生成文件列表,等等就可以了</h4><p>根据下载并转换的FASTQ文件位置，使用<code>ls</code>和<code>paste</code>命令生成包含读段文件路径的列表：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">ls</span> /path/to/output/directory/*.R1.fastq.gz &gt; rawdata_fastq_1</span><br><span class="line"><span class="built_in">ls</span> /path/to/output/directory/*.R2.fastq.gz &gt; rawdata_fastq_2</span><br><span class="line"><span class="built_in">paste</span> rawdata_fastq_1 rawdata_fastq_2 &gt; rawdata_fastq</span><br></pre></td></tr></table></figure>

<p>请将<code>/path/to/output/directory/</code>替换为实际存放<code>SRR4046839_1.fastq</code>和<code>SRR4046839_2.fastq</code>文件的目录路径。</p>
<h2 id="4，合并FASTQ文件"><a href="#4，合并FASTQ文件" class="headerlink" title="4，合并FASTQ文件"></a>4，合并FASTQ文件</h2><ol>
<li><strong>合并R1读取文件</strong>：使用<code>cat</code>命令将两个或多个R1读取的补测数据文件合并成一个文件。<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cat</span> V1-2_L4_160A60.R1.fastq1 V1-2_L4_160A60.R1.fastq2 &gt; V1-2_L4_160A60.R1.fastq</span><br></pre></td></tr></table></figure></li>
<li><strong>合并R2读取文件</strong>：同样地，使用<code>cat</code>命令合并R2读取的补测数据文件。<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cat</span> V1-2_L4_160A60.R2.fastq1 V1-2_L4_160A60.R2.fastq2 &gt; V1-2_L4_160A60.R2.fastq</span><br></pre></td></tr></table></figure></li>
</ol>
<p>对其他样本重复以上步骤。</p>
<h2 id="5，压缩合并后的文件"><a href="#5，压缩合并后的文件" class="headerlink" title="5，压缩合并后的文件"></a>5，压缩合并后的文件</h2><ol start="3">
<li><strong>压缩新合并的文件</strong>：使用<code>gzip</code>命令压缩合并后的文件以节省存储空间。<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">gzip V1-2_L4_160A60.R1.fastq</span><br><span class="line">gzip V1-2_L4_160A60.R2.fastq</span><br></pre></td></tr></table></figure></li>
</ol>
<p>对所有合并后的文件重复此步骤。</p>
<h2 id="6，重命名和压缩原始文件"><a href="#6，重命名和压缩原始文件" class="headerlink" title="6，重命名和压缩原始文件"></a>6，重命名和压缩原始文件</h2><ol start="4">
<li><strong>重命名原始压缩文件</strong>：使用<code>mv</code>命令简化文件名。<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mv</span> V0-1_L4_157A57.R1.fastq.gz V0-1.R1.fastq.gz</span><br></pre></td></tr></table></figure></li>
</ol>
<p>对所有需要重命名的文件重复此步骤。</p>
<ol start="5">
<li><strong>（可选）压缩重命名的文件</strong>：如果重命名的文件未被压缩（适用于未压缩的文件），使用<code>gzip</code>进行压缩。<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 仅当文件未压缩时执行</span></span><br><span class="line">gzip V0-1.R1.fastq</span><br></pre></td></tr></table></figure></li>
</ol>
<p>请注意，上述第5步仅适用于未压缩的文件。如果文件已经以<code>.gz</code>结尾，说明它们已经被压缩，无需再次压缩。</p>
<p>以上步骤将帮助您有效地管理和存储高通量测序数据。</p>
<h2 id="7，rawdata质控，或者其他质控方法"><a href="#7，rawdata质控，或者其他质控方法" class="headerlink" title="7，rawdata质控，或者其他质控方法"></a>7，rawdata质控，或者其他质控方法</h2> <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> cleandata</span><br><span class="line"><span class="built_in">ls</span> `<span class="built_in">pwd</span>`/*.R1.fastq.gz &gt; rawdata_fastq_1</span><br><span class="line"><span class="built_in">ls</span> `<span class="built_in">pwd</span>`/*.R2.fastq.gz &gt; rawdata_fastq_2</span><br><span class="line"><span class="built_in">paste</span> rawdata_fastq_1 rawdata_fastq_2&gt; rawdata_fastq</span><br><span class="line">------------------------------------------------------------------------</span><br><span class="line">vim qc.sh</span><br><span class="line"><span class="built_in">cat</span> rawdata_fastq|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span></span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">          arr=(<span class="variable">$&#123;id&#125;</span>)</span><br><span class="line">         fq1=<span class="variable">$&#123;arr[0]&#125;</span></span><br><span class="line">         fq2=<span class="variable">$&#123;arr[1]&#125;</span></span><br><span class="line">         trim_galore -q 25 --phred33 --length 36 -e 0.1 --stringency 3 --paired -o cleandata/ <span class="variable">$fq1</span> <span class="variable">$fq2</span> &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">chmod</span> a+x qc.sh</span><br><span class="line"><span class="built_in">nohup</span> ./qc.sh</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>这段代码是用于在生物信息学分析中进行序列数据的预处理，特别是用于质量控制和修剪原始的FASTQ格式的测序数据。这些操作通常是高通量测序数据分析流程的初步步骤，用以确保数据的质量足够好，适合进行后续的分析，如比对、变异检测等。下面是代码的逐步解释：</p>
<h3 id="创建目录和准备文件列表"><a href="#创建目录和准备文件列表" class="headerlink" title="创建目录和准备文件列表"></a>创建目录和准备文件列表</h3><ol>
<li><p><code>mkdir cleandata</code>：创建一个名为<code>cleandata</code>的目录，用于存放处理后的数据。</p>
</li>
<li><p><code>ls $(pwd)/*.R1.fastq.gz &gt; rawdata_fastq_1</code>：列出当前目录下所有以<code>.R1.fastq.gz</code>结尾的文件（代表配对测序的第一个文件），并将这些文件的完整路径保存到<code>rawdata_fastq_1</code>文件中。</p>
</li>
<li><p><code>ls $(pwd)/*.R2.fastq.gz &gt; rawdata_fastq_2</code>：同样，列出所有以<code>.R2.fastq.gz</code>结尾的文件（配对测序的第二个文件），并将路径保存到<code>rawdata_fastq_2</code>文件中。</p>
</li>
<li><p><code>paste rawdata_fastq_1 rawdata_fastq_2 &gt; rawdata_fastq</code>：将<code>rawdata_fastq_1</code>和<code>rawdata_fastq_2</code>文件中的内容合并（每行对应一个配对的样本），每对文件的路径由制表符分隔，结果保存到<code>rawdata_fastq</code>文件中。</p>
</li>
</ol>
<h3 id="编写和执行质量控制脚本"><a href="#编写和执行质量控制脚本" class="headerlink" title="编写和执行质量控制脚本"></a>编写和执行质量控制脚本</h3><ol start="5">
<li><code>vim qc.sh</code>：使用<code>vim</code>编辑器创建一个名为<code>qc.sh</code>的脚本文件。此步骤需要手动输入脚本内容，脚本的具体内容在下一部分解释。</li>
</ol>
<p>脚本内容解释：</p>
<ul>
<li>读取<code>rawdata_fastq</code>文件中的每一行，每行包含一对配对的FASTQ文件路径。</li>
<li>使用<code>trim_galore</code>命令对每对FASTQ文件进行质量控制和修剪。<code>trim_galore</code>是一个基于<code>Cutadapt</code>和<code>FastQC</code>的自动化质控工具，用于自动检测质量门槛、移除低质量的序列和adapter序列。</li>
<li>参数解释：<ul>
<li><code>-q 25</code>：设置质量分数门槛为25。</li>
<li><code>--phred33</code>：指定质量分数使用Phred+33编码。</li>
<li><code>--length 36</code>：移除修剪后长度小于36的读取。</li>
<li><code>-e 0.1</code>：允许的最大误差率为0.1。</li>
<li><code>--stringency 3</code>：在adapter修剪中，重叠长度需至少为3。</li>
<li><code>--paired</code>：指定输入文件为配对的。</li>
<li><code>-o cleandata/</code>：指定输出目录为<code>cleandata</code>。</li>
</ul>
</li>
<li><code>&amp;</code>：在后台执行每个<code>trim_galore</code>命令，允许同时处理多个文件对。</li>
</ul>
<ol start="6">
<li><p><code>chmod a+x qc.sh</code>：给予<code>qc.sh</code>脚本执行权限。</p>
</li>
<li><p><code>nohup ./qc.sh &amp;</code>：在后台执行<code>qc.sh</code>脚本，并通过<code>nohup</code>命令确保即使关闭终端会话，脚本也会继续运行。</p>
</li>
</ol>
<p>这个流程是高通量测序数据分析的常规质量控制步骤，旨在提高数据质量，为后续分析提供更准确的输入数据。</p>
<h2 id="8，统计cleandatabase"><a href="#8，统计cleandatabase" class="headerlink" title="8，统计cleandatabase"></a>8，统计cleandatabase</h2> <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> cleandataBase</span><br><span class="line"><span class="built_in">ls</span> `<span class="built_in">pwd</span>`/*.R1_val_1.fq.gz &gt; cleandata_fastq_1</span><br><span class="line"><span class="built_in">ls</span> `<span class="built_in">pwd</span>`/*.R2_val_2.fq.gz &gt; cleandata_fastq_2</span><br><span class="line">find *_1.fq.gz &gt;sample</span><br><span class="line">sed -i <span class="string">&quot;s/.R1_val_1.fq.gz/ /g&quot;</span>  sample    </span><br><span class="line"><span class="built_in">paste</span> cleandata_fastq_1 cleandata_fastq_2 sample&gt; cleandata_fastq</span><br><span class="line">vim fc.sh</span><br><span class="line"><span class="built_in">cat</span> cleandata_fastq|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span></span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">          arr=(<span class="variable">$&#123;id&#125;</span>)</span><br><span class="line">         fq1=<span class="variable">$&#123;arr[0]&#125;</span></span><br><span class="line">         fq2=<span class="variable">$&#123;arr[1]&#125;</span></span><br><span class="line">         sample=<span class="variable">$&#123;arr[2]&#125;</span></span><br><span class="line">fastq-count <span class="variable">$fq1</span> <span class="variable">$fq2</span> &gt;cleandataBase/<span class="variable">$sample</span> &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">chmod</span> a+x fc.sh</span><br><span class="line"><span class="built_in">nohup</span> ./fc.sh</span><br><span class="line"></span><br><span class="line"><span class="built_in">cd</span> cleandataBase</span><br><span class="line"><span class="built_in">cat</span> * &gt; cleandataBase</span><br><span class="line">sed -i  <span class="string">&#x27;s/&#125;/ /g&#x27;</span> cleandataBase</span><br><span class="line">sed -i  <span class="string">&#x27;s/:/\t/g&#x27;</span> cleandataBase</span><br><span class="line">sed -i  <span class="string">&#x27;s/,/\t/g&#x27;</span> cleandataBase</span><br><span class="line"><span class="built_in">cat</span> cleandataBase | awk <span class="string">&#x27;&#123;print $2&#125;&#x27;</span> &gt; cleandatabases</span><br><span class="line"><span class="built_in">cat</span> cleandataBase | awk <span class="string">&#x27;&#123;print $4&#125;&#x27;</span> &gt; cleandatareads</span><br><span class="line"><span class="built_in">cat</span> cleandatabases</span><br><span class="line"><span class="built_in">cat</span> cleandatareads</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>这个过程涉及一系列的命令，用于处理和统计经过质量控制后的FASTQ文件中的数据。下面是每一步的解释：</p>
<h3 id="准备工作和文件组织"><a href="#准备工作和文件组织" class="headerlink" title="准备工作和文件组织"></a>准备工作和文件组织</h3><ol>
<li><p><strong>创建目录</strong>:</p>
<ul>
<li><code>mkdir cleandataBase</code> 创建一个名为<code>cleandataBase</code>的目录，用于存储最终的处理结果。</li>
</ul>
</li>
<li><p><strong>生成FASTQ文件列表</strong>:</p>
<ul>
<li>使用<code>ls</code>命令列出当前工作目录下所有经过质量控制处理的R1和R2文件，并分别将这些文件的路径保存到<code>cleandata_fastq_1</code>和<code>cleandata_fastq_2</code>文件中。</li>
</ul>
</li>
<li><p><strong>生成样本名称列表</strong>:</p>
<ul>
<li><code>find *_1.fq.gz &gt;sample</code> 查找所有以<code>_1.fq.gz</code>结尾的文件（通常代表R1的质控后文件），并将这些文件名输出到<code>sample</code>文件中。</li>
<li>使用<code>sed</code>命令修改<code>sample</code>文件中的内容，去掉文件名中的<code>.R1_val_1.fq.gz</code>部分，仅留下样本名称。</li>
</ul>
</li>
<li><p><strong>合并文件列表和样本名称</strong>:</p>
<ul>
<li><code>paste cleandata_fastq_1 cleandata_fastq_2 sample &gt; cleandata_fastq</code> 将R1和R2的文件路径列表以及对应的样本名称合并到一个文件中，每行包含一对质控后的FASTQ文件和对应的样本名称。</li>
</ul>
</li>
</ol>
<h3 id="执行统计分析"><a href="#执行统计分析" class="headerlink" title="执行统计分析"></a>执行统计分析</h3><ol start="5">
<li><p><strong>编写并执行统计脚本</strong> (<code>fc.sh</code>):</p>
<ul>
<li>脚本读取<code>cleandata_fastq</code>文件中的每一行，这一行包含了一对配对的FASTQ文件路径和样本名称。</li>
<li>对于文件中的每一对FASTQ文件，脚本使用<code>fastq-count</code>（一个假设的命令，用于示例说明）来统计其中的序列数量，并将统计结果输出到<code>cleandataBase</code>目录下，文件名为样本名称。</li>
<li>脚本以后台方式执行，允许同时处理多个文件。</li>
</ul>
</li>
<li><p><strong>给脚本执行权限并运行</strong>:</p>
<ul>
<li><code>chmod a+x fc.sh</code> 使脚本可执行。</li>
<li><code>nohup ./fc.sh &amp;</code> 在后台运行脚本，<code>nohup</code>确保即使终端关闭，脚本也会继续运行。</li>
</ul>
</li>
</ol>
<h3 id="数据汇总和格式化"><a href="#数据汇总和格式化" class="headerlink" title="数据汇总和格式化"></a>数据汇总和格式化</h3><ol start="7">
<li><p><strong>汇总统计结果</strong>:</p>
<ul>
<li><code>cd cleandataBase</code> 切换到存储结果的目录。</li>
<li><code>cat * &gt; cleandataBase</code> 将目录中所有文件的内容合并到一个名为<code>cleandataBase</code>的文件中。</li>
</ul>
</li>
<li><p><strong>格式化统计结果</strong>:</p>
<ul>
<li>使用<code>sed</code>命令调整<code>cleandataBase</code>文件的格式，例如去除不需要的字符，将特定符号替换为制表符，以便更容易地读取和分析数据。</li>
<li>使用<code>awk</code>命令从格式化后的文件中提取特定的列（如序列基数和读数），并将这些数据保存到<code>cleandatabases</code>和<code>cleandatareads</code>文件中，分别用于进一步分析。</li>
</ul>
</li>
</ol>
<p>这个过程从组织经过质量控制的FASTQ文件开始，通过统计分析，最终生成易于阅读和分析的数据汇总。这对于理解每个样本的数据质量和量化信息非常有用。</p>
<p>最终得到的数据<br>最终，你会得到两个关键的数据文件：</p>
<p>cleandatabases：包含每个样本的序列基数（即序列的唯一标识符数目），这有助于了解样本的多样性和复杂性。<br>cleandatareads：包含每个样本的总读数，即每个样本中序列的总数量。</p>
<h2 id="9，建立索引，用上面的就行，下载fa和gtf文件"><a href="#9，建立索引，用上面的就行，下载fa和gtf文件" class="headerlink" title="9，建立索引，用上面的就行，下载fa和gtf文件"></a>9，建立索引，用上面的就行，下载fa和gtf文件</h2> <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span>  /public/jychu/reference/index/hisat/sheep</span><br><span class="line">wget http://ftp.ensembl.org/pub/release-104/fasta/anser_brachyrhynchus/dna/Anser_brachyrhynchus.ASM259213v1.dna.toplevel.fa.gz</span><br><span class="line">gzip -d Anser_brachyrhynchus.ASM259213v1.dna.toplevel.fa.gz     <span class="comment">#解压参考基因组文件</span></span><br><span class="line"><span class="built_in">mkdir</span> index                      <span class="comment">#新建一个存放构建参考索引的目录</span></span><br><span class="line">hisat2-build Anser_brachyrhynchus.ASM259213v1.dna.toplevel.fa index/index      <span class="comment">#构建以index为前缀的索引文件，并存放到index目录下    </span></span><br><span class="line"><span class="built_in">cd</span> /public/jychu/reference/annotate/chicken</span><br><span class="line">wget http://ftp.ensembl.org/pub/release-104/gtf/anser_brachyrhynchus/Anser_brachyrhynchus.ASM259213v1.104.gtf.gz</span><br><span class="line">gzip -d Anser_brachyrhynchus.ASM259213v1.104.gtf.gz</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h2 id="10，hisat2比对，批量比对，上面也有"><a href="#10，hisat2比对，批量比对，上面也有" class="headerlink" title="10，hisat2比对，批量比对，上面也有"></a>10，hisat2比对，批量比对，上面也有</h2> <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> /public/jychu/lncRNA-chicken/cleandata</span><br><span class="line"><span class="built_in">mkdir</span> bam</span><br><span class="line"><span class="built_in">ls</span> `<span class="built_in">pwd</span>`/*_1.fq.gz &gt; cleandata_fastq_1</span><br><span class="line"><span class="built_in">ls</span> `<span class="built_in">pwd</span>`/*_2.fq.gz &gt; cleandata_fastq_2</span><br><span class="line">find *_1.fq.gz &gt;sample</span><br><span class="line">sed -i <span class="string">&quot;s/.R1_val_1.fq.gz/ /g&quot;</span>  sample</span><br><span class="line"><span class="built_in">paste</span> cleandata_fastq_1 cleandata_fastq_2 sample&gt; cleandata_fastq</span><br><span class="line">vim compare.sh</span><br><span class="line">INDEX=<span class="string">&#x27;/public/jychu/reference/index/hisat/chicken/index/index&#x27;</span></span><br><span class="line"><span class="built_in">cat</span> cleandata_fastq|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span></span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">          arr=(<span class="variable">$&#123;id&#125;</span>)</span><br><span class="line">         fq1=<span class="variable">$&#123;arr[0]&#125;</span></span><br><span class="line">         fq2=<span class="variable">$&#123;arr[1]&#125;</span></span><br><span class="line">         sample=<span class="variable">$&#123;arr[2]&#125;</span></span><br><span class="line">hisat2  -p 8 -x <span class="variable">$INDEX</span> -1 <span class="variable">$fq1</span> -2 <span class="variable">$fq2</span> | samtools view -S -b - &gt; bam/<span class="variable">$sample</span>.bam &amp;</span><br><span class="line"></span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">chmod</span> a+x  compare.sh</span><br><span class="line"></span><br><span class="line"><span class="built_in">nohup</span> ./compare.sh</span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>


<h2 id="11，排序，批量排序，上面有"><a href="#11，排序，批量排序，上面有" class="headerlink" title="11，排序，批量排序，上面有"></a>11，排序，批量排序，上面有</h2>  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> bam</span><br><span class="line"></span><br><span class="line"><span class="built_in">ls</span> *.bam|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span></span><br><span class="line"><span class="keyword">do</span> </span><br><span class="line">samtools <span class="built_in">sort</span> -@ 4  <span class="variable">$id</span> -o <span class="variable">$&#123;id%.*&#125;</span>.sorted</span><br><span class="line"><span class="built_in">rm</span> -f <span class="variable">$&#123;id%.*&#125;</span>.bam</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h2 id="12，统计比对率"><a href="#12，统计比对率" class="headerlink" title="12，统计比对率"></a>12，统计比对率</h2> <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"><span class="built_in">mkdir</span> flagstat</span><br><span class="line"></span><br><span class="line"><span class="built_in">ls</span> *.bam &gt; flagstat_bam</span><br><span class="line"></span><br><span class="line">find *.bam&gt;bam_sample</span><br><span class="line"></span><br><span class="line">sed -i <span class="string">&quot;s/.sorted.bam/ /g&quot;</span> bam_sample</span><br><span class="line"></span><br><span class="line"><span class="built_in">paste</span> flagstat_bam bam_sample &gt;flagstatbam_list</span><br><span class="line"></span><br><span class="line">vim flagstatbam.sh</span><br><span class="line"></span><br><span class="line"><span class="built_in">cat</span> flagstatbam_list |<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">do</span></span><br><span class="line"></span><br><span class="line">          arr=(<span class="variable">$&#123;id&#125;</span>)</span><br><span class="line"></span><br><span class="line">         bam=<span class="variable">$&#123;arr[0]&#125;</span></span><br><span class="line"></span><br><span class="line">         sample=<span class="variable">$&#123;arr[1]&#125;</span></span><br><span class="line"></span><br><span class="line">         samtools flagstat <span class="variable">$bam</span> &gt; flagstat/<span class="variable">$sample</span>.flagstat.txt &amp;</span><br><span class="line"></span><br><span class="line">         <span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">chmod</span> a+x flagstatbam.sh</span><br><span class="line"></span><br><span class="line"><span class="built_in">nohup</span> ./flagstatbam.sh</span><br><span class="line"></span><br><span class="line"><span class="built_in">cd</span> flagstat</span><br><span class="line"></span><br><span class="line"><span class="built_in">cat</span> * &gt; flagstat</span><br><span class="line"></span><br><span class="line">grep <span class="string">&quot;0 mapped&quot;</span> flagstat  &gt; flagstats</span><br><span class="line"></span><br><span class="line"><span class="built_in">cat</span> flagstats | awk <span class="string">&#x27;&#123;print $5&#125;&#x27;</span> &gt; flag</span><br><span class="line"></span><br><span class="line">sed -i <span class="string">&quot;s/(/ /g&quot;</span> flag</span><br><span class="line"></span><br><span class="line">sed -i <span class="string">&quot;s/%/ /g&quot;</span> flag</span><br><span class="line"></span><br><span class="line"><span class="built_in">cat</span> flag</span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>


<h2 id="13，stringTie组装，批量组装"><a href="#13，stringTie组装，批量组装" class="headerlink" title="13，stringTie组装，批量组装"></a>13，stringTie组装，批量组装</h2> <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"><span class="built_in">mkdir</span> gtfs</span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> `<span class="built_in">ls</span> *.sorted`;</span><br><span class="line"><span class="keyword">do</span> new=<span class="variable">$&#123;i%.*&#125;</span>.sorted.bam</span><br><span class="line"><span class="built_in">mv</span> <span class="string">&quot;<span class="variable">$i</span>&quot;</span> <span class="string">&quot;<span class="variable">$new</span>&quot;</span></span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">ls</span> *.sorted.bam|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span></span><br><span class="line"><span class="keyword">do</span> <span class="built_in">nohup</span> stringtie <span class="variable">$id</span> -p 4 -G /public/jychu/reference/annotate/chicken/Anser_brachyrhynchus.ASM259213v1.104.gtf -o gtfs/<span class="variable">$&#123;id&#125;</span>.gtf -l <span class="variable">$&#123;id%.*&#125;</span> &amp;</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>



<h2 id="14，合并转录本"><a href="#14，合并转录本" class="headerlink" title="14，合并转录本"></a>14，合并转录本</h2> <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"> <span class="built_in">cd</span> gtfs</span><br><span class="line"><span class="built_in">mkdir</span> merged</span><br><span class="line"><span class="built_in">ls</span> *.gtf &gt; merge.list</span><br><span class="line">stringtie --merge -G /public/jychu/reference/annotate/chicken/Anser_brachyrhynchus.ASM259213v1.104.gtf merge.list -o merged/merged.gtf</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h2 id="15，定量，获得转录本表达矩阵"><a href="#15，定量，获得转录本表达矩阵" class="headerlink" title="15，定量，获得转录本表达矩阵"></a>15，定量，获得转录本表达矩阵</h2> <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> /public/jychu/lncRNA-chicken/cleandata/bam</span><br><span class="line"><span class="built_in">mkdir</span> ballgown</span><br><span class="line"><span class="built_in">ls</span> *sorted.bam|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span></span><br><span class="line"><span class="keyword">do</span> stringtie -e <span class="variable">$id</span> -p 4 -G gtfs/merged/merged.gtf -o ballgown/<span class="variable">$&#123;id%%.*&#125;</span>.bg.gtf -b ballgown/<span class="variable">$&#123;id&#125;</span>_out_dir</span><br><span class="line"><span class="keyword">done</span>    <span class="comment">#应该是为了取所有样本都注释到的转录本进行定量</span></span><br><span class="line"><span class="comment">#制作转录本表达矩阵   用ballgrown的transcript_fpkm = texpr(bg, &#x27;FPKM&#x27;)  可以直接得出表达矩阵，不过需要把编号和转录本ID对应替换一下</span></span><br><span class="line"><span class="built_in">cd</span> ballgown</span><br><span class="line"><span class="built_in">mkdir</span> quality</span><br><span class="line"><span class="comment">#获取每个样本的转录本ID</span></span><br><span class="line"><span class="built_in">ls</span> *.bg.gtf|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> <span class="built_in">cat</span> <span class="variable">$id</span>| awk <span class="string">&#x27;$3==&quot;transcript&quot;&#x27;</span>|grep -Eo <span class="string">&#x27;transcript_id \&quot;\w+|transcript_id \&quot;\w+\.\w+\.\w+&#x27;</span>|<span class="built_in">cut</span> -d\&quot; -f 2 &gt; quality/<span class="variable">$&#123;id%%.*&#125;</span>.t; <span class="keyword">done</span></span><br><span class="line"><span class="comment">#获取每个样本的FPKM值</span></span><br><span class="line"><span class="built_in">ls</span> *.bg.gtf|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span> </span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">less <span class="variable">$id</span>|awk <span class="string">&#x27;$3==&quot;transcript&quot;&#x27;</span>|grep -Eo <span class="string">&#x27;FPKM \&quot;\w+\.\w+&#x27;</span>|awk -F\&quot; <span class="string">&#x27;&#123;print$2&#125;&#x27;</span> &gt; quality/<span class="variable">$&#123;id%%.*&#125;</span>.value</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">cd</span> quality</span><br><span class="line"><span class="comment">#合并每个样本转录本和对应的FPKM值</span></span><br><span class="line"><span class="built_in">ls</span> *.t|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> <span class="built_in">paste</span> <span class="variable">$&#123;id%%.*&#125;</span>.t <span class="variable">$&#123;id%%.*&#125;</span>.value &gt; <span class="variable">$&#123;id%%.*&#125;</span>.FPKM ; <span class="keyword">done</span></span><br><span class="line"><span class="comment">#按转录本字母排序</span></span><br><span class="line"><span class="built_in">ls</span> *.FPKM|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>;<span class="keyword">do</span> less <span class="variable">$id</span>|<span class="built_in">sort</span> -k 1 &gt; <span class="variable">$&#123;id%.*&#125;</span>.sorted.FPKM;<span class="keyword">done</span></span><br><span class="line"><span class="comment">#提取样本的第一列并加列名</span></span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;transcript&quot;</span> &gt; FPKM</span><br><span class="line">less `<span class="built_in">ls</span> *.sorted.FPKM|<span class="built_in">shuf</span> -n1`|<span class="built_in">cut</span> -f 1 &gt;&gt; FPKM</span><br><span class="line"><span class="comment">#提取每个样本的第二列并加列名</span></span><br><span class="line"><span class="built_in">ls</span> *.sorted.FPKM|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>;<span class="keyword">do</span> <span class="built_in">echo</span> <span class="variable">$&#123;id%%.*&#125;</span> &gt; <span class="variable">$&#123;id%%.*&#125;</span>.tmp;less <span class="variable">$id</span>|<span class="built_in">cut</span> -f 2 &gt;&gt; <span class="variable">$&#123;id%%.*&#125;</span>.tmp;<span class="keyword">done</span></span><br><span class="line"><span class="comment">#合并转录ID列的文件以及所有样本的FPKM值</span></span><br><span class="line"><span class="built_in">paste</span> FPKM *.tmp &gt; all.FPKM</span><br><span class="line"><span class="comment">#删除除了all.FPKM的所有文件</span></span><br><span class="line"><span class="built_in">ls</span> *.value *.t *.FPKM *.tmp FPKM|grep -v all.FPKM |xargs <span class="built_in">rm</span></span><br><span class="line"><span class="comment">#至此获得了转录本的表达矩阵  并下载至桌面上</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>


<h2 id="16，ballgrown差异分析"><a href="#16，ballgrown差异分析" class="headerlink" title="16，ballgrown差异分析"></a>16，ballgrown差异分析</h2> <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> /public/jychu/lncRNA-chicken/cleandata/bam/ballgown</span><br><span class="line">conda activate dna</span><br><span class="line">R</span><br><span class="line"><span class="keyword">if</span> (!requireNamespace(<span class="string">&quot;BiocManager&quot;</span>, quietly = TRUE))</span><br><span class="line">    install.packages(<span class="string">&quot;BiocManager&quot;</span>)</span><br><span class="line">BiocManager::install(<span class="string">&quot;ballgown&quot;</span>)</span><br><span class="line">library(ballgown)</span><br><span class="line"><span class="built_in">bg</span> = ballgown(samples = c(<span class="string">&quot;V0-1.sorted.bam_out_dir&quot;</span>, <span class="string">&quot;V0-3.sorted.bam_out_dir&quot;</span>,<span class="string">&quot;V0-4.sorted.bam_out_dir&quot;</span>,<span class="string">&quot;V1-2.sorted.bam_out_dir&quot;</span>,<span class="string">&quot;V1-3.sorted.bam_out_dir&quot;</span>,<span class="string">&quot;V1-4.sorted.bam_out_dir&quot;</span>),meas=<span class="string">&#x27;all&#x27;</span>)</span><br><span class="line"><span class="comment">#查看转录本水平的表达量的代码示例如下</span></span><br><span class="line">transcript_fpkm = texpr(<span class="built_in">bg</span>, <span class="string">&#x27;FPKM&#x27;</span>)</span><br><span class="line"><span class="built_in">head</span>(transcript_fpkm)</span><br><span class="line">pData(<span class="built_in">bg</span>) &lt;- data.frame(<span class="built_in">id</span>=sampleNames(<span class="built_in">bg</span>),group=rep(c(1,0), each=3))    <span class="comment">#分组</span></span><br><span class="line"><span class="comment"># 转录本水平的差异分析</span></span><br><span class="line">stat_results = stattest(<span class="built_in">bg</span>,feature=<span class="string">&#x27;transcript&#x27;</span>,meas=<span class="string">&#x27;FPKM&#x27;</span>,covariate=<span class="string">&#x27;group&#x27;</span>)</span><br><span class="line">write.csv(stat_results, <span class="string">&quot;ballgrown_transcript_results.csv&quot;</span>,row.names=FALSE)</span><br><span class="line">------------------------------------------------------------------</span><br><span class="line"><span class="built_in">cd</span> /public/jychu/lncRNA-chicken/cleandata/bam/ballgown/V0-1.sorted.bam_out_dir</span><br><span class="line"><span class="built_in">cat</span> t_data.ctab|<span class="built_in">cut</span> -f2-8&gt;<span class="built_in">test</span>  <span class="comment">#下载到桌面上</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">------------------------------------------------------------------</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">在你提供的命令和上下文中，`V0-1.sorted.bam_out_dir` 是一个目录，用于存储特定样本的`ballgown`分析数据。具体来说，这个目录通常包含`ballgown`需要读取的文件，用于进行转录组数据的表达量分析和差异表达分析。而这些文件通常是由`StringTie`生成的，用于描述样本中转录本的表达量信息。</span><br><span class="line"></span><br><span class="line">在`V0-1.sorted.bam_out_dir`这样的目录中，你可能会找到以下类型的文件：</span><br><span class="line"></span><br><span class="line">1. **e_data.ctab**：包含每个外显子的表达量信息，如外显子ID、长度、覆盖度等。</span><br><span class="line">2. **i_data.ctab**：包含每个内含子的信息。</span><br><span class="line">3. **t_data.ctab**：包含每个转录本的表达量信息，如转录本ID、长度、FPKM（Fragment Per Kilobase Million）值等。</span><br><span class="line">4. **e2t.ctab** 和 **i2t.ctab**：分别是外显子到转录本和内含子到转录本的映射文件。</span><br><span class="line"></span><br><span class="line">以`t_data.ctab`为例，它的内容可能如下：</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>t_id    gene_id    length    eff_length    est_counts    abundance<br>1       Gene1      1000      950           150           20<br>2       Gene1      800       750           100           15</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">- `t_id`：转录本ID</span><br><span class="line">- `gene_id`：基因ID</span><br><span class="line">- `length`：转录本长度</span><br><span class="line">- `eff_length`：考虑到测序读长后的有效长度</span><br><span class="line">- `est_counts`：估计的计数，基于对该转录本的覆盖度</span><br><span class="line">- `abundance`：表达量估计，比如FPKM值</span><br><span class="line"></span><br><span class="line">你最后的命令`cat t_data.ctab | cut -f2-8 &gt; test` 是将`t_data.ctab`文件中的第2到第8列的内容提取出来，并将这部分内容重定向到一个名为`test`的新文件中。这意味着你会得到一个包含转录本ID、基因ID、长度、有效长度、估计计数和丰度等信息的文件，但是以一种更简化的格式。</span><br><span class="line"></span><br><span class="line">这些步骤和文件是转录组分析流程中的一部分，用于评估和比较不同样本中的基因和转录本的表达水平。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">这段代码主要用于进行转录组数据的差异表达分析，使用的是`ballgown`包在R语言环境下。它从一系列的`.bam`文件中读取数据，然后进行统计分析，最终生成差异表达的结果。这里简单解释一下输入和输出文件的内容：</span><br><span class="line"></span><br><span class="line">### 输入文件：</span><br><span class="line"></span><br><span class="line">1. **.bam文件**：这些是经过排序的二进制版本的序列对齐/比对文件（例如，`V0-1.sorted.bam`），通常由比对工具（如`TopHat`、`HISAT2`等）生成。`.bam`文件包含了序列数据（如RNA序列）与参考基因组的对齐信息。</span><br><span class="line"></span><br><span class="line">2. **t_data.ctab**：这是`ballgown`需要的文件之一，包含了转录本的表达量和其他统计数据。它位于每个样本的`ballgown`输出目录中（如`V0-1.sorted.bam_out_dir`）。`ctab`文件包括转录本ID、基因ID、长度、覆盖度、FPKM等信息。</span><br><span class="line"></span><br><span class="line">### 输出文件：</span><br><span class="line"></span><br><span class="line">1. **ballgrown_transcript_results.csv**：这个文件包含了转录本水平的差异表达分析结果。对于每个转录本，它可能包含列如转录本ID、基因名、统计值（如p值）、FPKM值等，用于指示在不同条件或组之间表达量的显著差异。</span><br><span class="line"></span><br><span class="line">2. **test**（通过`cut`命令从`t_data.ctab`生成）：这个文件是`t_data.ctab`的简化版本，只保留了部分列（第2到第8列），通常包括转录本ID、位置信息、编码区长度、外显子数量等。这个文件更便于用户进行快速查看或特定分析。</span><br><span class="line"></span><br><span class="line">### 例子：</span><br><span class="line"></span><br><span class="line">假设一个简化的`t_data.ctab`文件如下：</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>t_id    gene_id    length    exon_count    FPKM<br>1       geneA      1000      5             50.0<br>2       geneB      800       3             30.0</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">经过上述代码处理后，得到的`test`文件将会是：</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>gene_id    length    exon_count    FPKM<br>geneA      1000      5             50.0<br>geneB      800       3             30.0</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">而`ballgrown_transcript_results.csv`将包含差异表达分析的结果，比如：</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>transcript_id    gene_name    p_value    q_value    FPKM_group1    FPKM_group2<br>1                geneA        0.05       0.1        50.0           20.0<br>2                geneB        0.01       0.02       30.0           15.0</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">这个文件显示了每个转录本在不同组之间的表达量差异，包括统计显著性和表达量变化。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>





<h2 id="17，初步筛选lncRNA-，上面也有"><a href="#17，初步筛选lncRNA-，上面也有" class="headerlink" title="17，初步筛选lncRNA#，上面也有"></a>17，初步筛选lncRNA#，上面也有</h2> <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### compare to reference transcripts, parameter M refer to remove single exon transcripts</span></span><br><span class="line"><span class="built_in">cd</span> /public/jychu/lncRNA-chicken/cleandata/bam/gtfs/merged</span><br><span class="line"><span class="built_in">mkdir</span> Identification</span><br><span class="line"><span class="built_in">cd</span> Identification</span><br><span class="line">gffcompare -R -M -r /public/jychu/reference/annotate/chicken/Anser_brachyrhynchus.ASM259213v1.104.gtf -s /public/jychu/reference/index/hisat/chicken/Anser_brachyrhynchus.ASM259213v1.dna.toplevel.fa ../merged.gtf </span><br><span class="line"></span><br><span class="line"><span class="comment">## class code filteration</span></span><br><span class="line"><span class="comment">#进入soft目录下载gtftools</span></span><br><span class="line"><span class="built_in">cd</span> ~/soft/</span><br><span class="line">wget https://github.com/Dukunbioinfo/in-house-gtftools/archive/master.zip</span><br><span class="line">unzip master.zip   <span class="comment">#生成in-house-gtftools-master文件夹</span></span><br><span class="line"><span class="built_in">cd</span> in-house-gtftools-master/</span><br><span class="line">g++ *.cpp -o inhouseGTFtools   <span class="comment">#安装，得到一个绿色的可执行文件</span></span><br><span class="line">~/soft/in-house-gtftools-master/inhouseGTFtools STscreen -classCode u -i gffcmp.annotated.gtf &gt; u.gtf          <span class="comment">#代表基因间</span></span><br><span class="line">~/soft/in-house-gtftools-master/inhouseGTFtools STscreen -classCode i -i gffcmp.annotated.gtf &gt; i.gtf    <span class="comment">#代表内含子</span></span><br><span class="line">~/soft/in-house-gtftools-master/inhouseGTFtools STscreen -classCode x -i gffcmp.annotated.gtf &gt; x.gtf     <span class="comment">#代表反义</span></span><br><span class="line"><span class="built_in">cat</span> i.gtf u.gtf x.gtf &gt; iux.gtf</span><br><span class="line"></span><br><span class="line"><span class="comment">#得到read序列长度</span></span><br><span class="line"><span class="comment">#下载gffread</span></span><br><span class="line">conda install gffread -y</span><br><span class="line"><span class="comment"># read sequence</span></span><br><span class="line">gffread -w iux.fa -g /public/jychu/reference/index/hisat/chicken/Anser_brachyrhynchus.ASM259213v1.dna.toplevel.fa iux.gtf</span><br><span class="line">samtools faidx iux.fa</span><br><span class="line"><span class="comment"># length filteration  #转录本的长度是指转录本中所有外显子加起来的长度</span></span><br><span class="line">less iux.fa.fai |awk <span class="string">&#x27;$2&gt;200&#123;print$1&#125;&#x27;</span> &gt; iuxL200.MSTRG    <span class="comment">#筛选长度大于200的转录本</span></span><br><span class="line"><span class="comment">## 根据FPKM值过滤</span></span><br><span class="line"><span class="built_in">cat</span> iuxL200.MSTRG | <span class="built_in">tr</span> <span class="string">&quot;.&quot;</span> <span class="string">&quot;\t&quot;</span> &gt; iuxL200.tmp</span><br><span class="line"><span class="built_in">cat</span> iuxL200.tmp | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>;</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">               arr=(<span class="variable">$&#123;id&#125;</span>)</span><br><span class="line">         MSTRG=<span class="variable">$&#123;arr[0]&#125;</span></span><br><span class="line">         N1=<span class="variable">$&#123;arr[1]&#125;</span></span><br><span class="line">              N2=<span class="variable">$&#123;arr[2]&#125;</span></span><br><span class="line">grep  -w <span class="string">&quot;<span class="variable">$MSTRG</span>\.<span class="variable">$N1</span>\.<span class="variable">$N2</span>&quot;</span> ../../../ballgown/quality/all.FPKM &gt;&gt; iux200FPKM.tmp</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">cat</span> iux200FPKM.tmp | awk <span class="string">&#x27;$2&gt;0.1 || $3&gt;0.1 || $4&gt;0.1 || $5&gt;0.1 || $6&gt;0.1 || $7&gt;0.1 &#123;print$1&#125;&#x27;</span> &gt; iux200LFPKM0.1.name  <span class="comment">#提取至少在一个样本里有表达量的转录本</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">cat</span> iux200LFPKM0.1.name | <span class="built_in">tr</span> <span class="string">&quot;.&quot;</span> <span class="string">&quot;\t&quot;</span> &gt; iux200LFPKM0.1.tmp</span><br><span class="line"><span class="built_in">cat</span> iux200LFPKM0.1.tmp | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>;</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">         arr=(<span class="variable">$&#123;id&#125;</span>)</span><br><span class="line">         MSTRG=<span class="variable">$&#123;arr[0]&#125;</span></span><br><span class="line">         N1=<span class="variable">$&#123;arr[1]&#125;</span></span><br><span class="line">         N2=<span class="variable">$&#123;arr[2]&#125;</span></span><br><span class="line">grep  -w <span class="string">&quot;<span class="variable">$MSTRG</span>\.<span class="variable">$N1</span>\.<span class="variable">$N2</span>&quot;</span> gffcmp.annotated.gtf &gt;&gt; iux200LFPKM0.1.gtf</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line">gffread -w iux200LFPKM0.1.fa -g /public/jychu/reference/index/hisat/chicken/Anser_brachyrhynchus.ASM259213v1.dna.toplevel.fa iux200LFPKM0.1.gtf</span><br><span class="line"><span class="built_in">rm</span> -f *.tmp</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">执行这一系列的命令后，你将完成一整套基于转录组数据的过滤和分析流程，特别是针对长非编码RNA（lncRNA）的鉴定和分析。这个过程包括几个关键步骤：比对、分类、过滤和序列提取。具体来说，你会得到：</span><br><span class="line"></span><br><span class="line">1. **比对和分类（gffcompare）**：使用`gffcompare`比对你的转录本到参考基因组，考虑了移除单外显子转录本的参数（`-M`），并且只关注与参考基因组重叠的转录本（`-R`）。`gffcompare`会生成几个文件，包括`gffcmp.annotated.gtf`，其中包含了转录本的分类代码（class codes），这些代码描述了转录本相对于参考基因组的位置关系。</span><br><span class="line"></span><br><span class="line">2. **过滤（class code filtration）**：通过自定义工具`inhouseGTFtools`对`gffcmp.annotated.gtf`进行过滤，提取特定分类代码的转录本，如基因间(`u`)、内含子(`i`)和反义(`x`)转录本，然后将这些分类的转录本合并成一个新的GTF文件`iux.gtf`。</span><br><span class="line"></span><br><span class="line">3. **序列提取和长度过滤**：使用`gffread`根据`iux.gtf`从参考基因组提取相应的转录本序列到`iux.fa`，并利用`samtools faidx`和`awk`命令筛选出长度大于200bp的转录本，结果保存在`iuxL200.MSTRG`文件中。</span><br><span class="line"></span><br><span class="line">4. **基于FPKM值的过滤**：通过脚本过滤出至少在一个样本中FPKM值大于0.1的转录本，并将这些转录本的ID保存在`iux200LFPKM0.1.name`文件中。</span><br><span class="line"></span><br><span class="line">5. **最终提取和清理**：使用这个过滤后的转录本ID列表从`gffcmp.annotated.gtf`中提取相应的转录本信息，保存在`iux200LFPKM0.1.gtf`。然后，利用`gffread`再次从参考基因组中提取这些过滤后的转录本的序列，保存在`iux200LFPKM0.1.fa`。</span><br><span class="line"></span><br><span class="line">6. **清理临时文件**：最后，删除所有中间生成的临时文件。</span><br><span class="line"></span><br><span class="line">执行这些命令后，你将获得一组经过长度和表达量过滤后，可能包含潜在长非编码RNA的转录本序列文件（`iux200LFPKM0.1.fa`），以及相应的GTF文件（`iux200LFPKM0.1.gtf`），这为进一步的功能分析和实验验证提供了基础。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>



<h2 id="18，鉴定lncRNA-，CPC2-CPAT-详细"><a href="#18，鉴定lncRNA-，CPC2-CPAT-详细" class="headerlink" title="18，鉴定lncRNA#，CPC2,CPAT(详细)"></a>18，鉴定lncRNA#，CPC2,CPAT(详细)</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br></pre></td><td class="code"><pre><span class="line">1.CPC2</span><br><span class="line">先下载软件压缩包到桌面，地址（http://cpc2.gao-lab.org/data/CPC2-beta.tar.gz）</span><br><span class="line"> tar xzvf CPC2-beta.tar.gz</span><br><span class="line"> <span class="built_in">cd</span> CPC2-beta/</span><br><span class="line"> <span class="built_in">ls</span></span><br><span class="line"> <span class="built_in">pwd</span></span><br><span class="line"> <span class="built_in">export</span> CPC_HOME=<span class="string">&quot;/public/jychu/soft/CPC2-beta&quot;</span></span><br><span class="line"> <span class="built_in">cd</span> libs/libsvm</span><br><span class="line"> tar xzvf libsvm-3.18.tar.gz</span><br><span class="line"> <span class="built_in">cd</span> libsvm-3.18</span><br><span class="line"> <span class="built_in">ls</span></span><br><span class="line"> make clean &amp;&amp; make</span><br><span class="line">1.1 使用</span><br><span class="line">conda activate python2  <span class="comment">#cpc2脚本只能在python2环境下使用</span></span><br><span class="line"><span class="built_in">cd</span> /public/jychu/lncRNA-chicken/cleandata/bam/gtfs/merged/Identification</span><br><span class="line"><span class="built_in">mkdir</span> lncRNA</span><br><span class="line"></span><br><span class="line">python /public/jychu/soft/CPC2-beta/bin/CPC2.py -i iux200LFPKM0.1.fa -o lncRNA/chicken.</span><br><span class="line"></span><br><span class="line">cpc2      <span class="comment">#运行命令</span></span><br><span class="line">chicken.cpc2.txt             <span class="comment">#输出文件</span></span><br><span class="line">grep <span class="string">&quot;noncoding&quot;</span> chicken.cpc2.txt |<span class="built_in">cut</span> -f1 &gt;chicken.cpc2.lnc</span><br><span class="line"></span><br><span class="line">2.CPAT    使用方法http://rna-cpat.sourceforge.net/</span><br><span class="line">pip install CPAT      <span class="comment">#下载软件</span></span><br><span class="line"><span class="comment"># 已有的文件在这个网址  https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model/</span></span><br><span class="line"><span class="comment">#构建训练模型</span></span><br><span class="line">1.CAPT预测</span><br><span class="line"><span class="comment">#非模式动物构建训练集</span></span><br><span class="line"><span class="built_in">mkdir</span> -p /public/jychu/reference/index/hisat/chicken</span><br><span class="line">wget http://ftp.ensembl.org/pub/release-103/fasta/anser_brachyrhynchus/cdna/Anser_brachyrhynchus.ASM259213v1.cdna.all.fa.gz</span><br><span class="line">wget http://ftp.ensembl.org/pub/release-103/fasta/anser_brachyrhynchus/ncrna/Anser_brachyrhynchus.ASM259213v1.ncrna.fa.gz</span><br><span class="line"></span><br><span class="line">conda install seqkit</span><br><span class="line">grep <span class="string">&quot;protein&quot;</span> Anser_brachyrhynchus.ASM259213v1.cdna.all.fa | sed <span class="string">&quot;s/&gt;//g&quot;</span>  | <span class="built_in">tr</span> <span class="string">&quot; &quot;</span> <span class="string">&quot;\t&quot;</span>|<span class="built_in">cut</span> -f1 &gt; cdna.name  <span class="comment">#提取cdna序列名称</span></span><br><span class="line"><span class="built_in">cat</span> Anser_brachyrhynchus.ASM259213v1.cdna.all.fa  | seqkit grep -f cdna.name &gt; cdna.name.fa   <span class="comment">#提取cdna名称对应的序列</span></span><br><span class="line">grep <span class="string">&quot;lncRNA&quot;</span> Anser_brachyrhynchus.ASM259213v1.ncrna.fa| sed <span class="string">&quot;s/&gt;//g&quot;</span>  | <span class="built_in">tr</span> <span class="string">&quot; &quot;</span> <span class="string">&quot;\t&quot;</span>|<span class="built_in">cut</span> -f1 &gt; lncRNA.name    <span class="comment">#提取lncRNA序列名称</span></span><br><span class="line"><span class="built_in">cat</span> Anser_brachyrhynchus.ASM259213v1.ncrna.fa | seqkit grep -f lncRNA.name &gt; lncRNA.name.fa    <span class="comment">#提取lncRNA名称对应的序列</span></span><br><span class="line"><span class="built_in">mkdir</span> -p /public/jychu/soft/CPAT/chicken</span><br><span class="line">make_hexamer_tab.py -c cdna.name.fa  -n  lncRNA.name.fa &gt;/public/jychu/soft/CPAT/chicken/chicken_Hexamer.tsv</span><br><span class="line">make_logitModel.py  -x ~/soft/CPAT/chicken/chicken_Hexamer.tsv -c cdna.name.fa  -n lncRNA.name.fa -o /public/jychu/soft/CPAT/chicken/chicken</span><br><span class="line">cpat.py -x /public/jychu/soft/CPAT/chicken/chicken_Hexamer.tsv -d /public/jychu/soft/CPAT/chicken/chicken.logit.RData  -g iux200LFPKM0.1.fa -o lncRNA/chicken.CAPT</span><br><span class="line">awk  <span class="string">&#x27;$2&gt;=200&#x27;</span> chicken.CAPT.ORF_prob.tsv |<span class="built_in">cut</span> -f1|<span class="built_in">tr</span> <span class="string">&quot;_&quot;</span> <span class="string">&quot;\t&quot;</span>|<span class="built_in">cut</span> -f1|<span class="built_in">sort</span>|<span class="built_in">uniq</span> &gt; capt.alltrans    <span class="comment">#提取ORF长度大于200的所有转录本id</span></span><br><span class="line">awk <span class="string">&#x27;$10&gt;=0.3 &amp;&amp; $2&gt;=200&#x27;</span> chicken.CAPT.ORF_prob.tsv|<span class="built_in">cut</span> -f1|<span class="built_in">tr</span> <span class="string">&quot;_&quot;</span> <span class="string">&quot;\t&quot;</span>|<span class="built_in">cut</span> -f1|<span class="built_in">sort</span>|<span class="built_in">uniq</span> &gt; chicken.CAPT.unlncRNA     <span class="comment">#提取编码潜力大于0.3的转录本id</span></span><br><span class="line"><span class="built_in">cat</span> chicken.unlncRNA capt.alltrans|<span class="built_in">sort</span>|<span class="built_in">uniq</span> -u &gt; chicken.capt.lnc      <span class="comment">#提取编码潜力小于0.3，ORF长度大于200的转录本id</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">这些步骤涉及到使用CPAT (Coding Potential Assessment Tool) 来评估转录本的编码潜力，特别是为了区分编码蛋白的mRNA和非编码的长链非编码RNA（lncRNA）。这是一个重要的过程，因为lncRNAs在基因调控、疾病发展等方面扮演着关键角色，而它们不编码蛋白质。</span><br><span class="line"></span><br><span class="line">1. **安装seqkit**: 首先，安装`seqkit`，这是一个用于处理FASTA/FASTQ文件的高效工具。</span><br><span class="line"></span><br><span class="line">2. **提取cdna序列名称**: 使用`grep`、`sed`、`<span class="built_in">tr</span>`和`<span class="built_in">cut</span>`命令从`.cdna.all.fa`文件中提取包含<span class="string">&quot;protein&quot;</span>的序列名称，这些是编码蛋白质的mRNA序列。</span><br><span class="line"></span><br><span class="line">3. **提取cdna序列**: 通过`seqkit grep`命令，使用上一步提取的名称从相同的FASTA文件中提取对应的序列。</span><br><span class="line"></span><br><span class="line">4. **提取lncRNA序列名称**: 类似地，从`.ncrna.fa`文件中提取包含<span class="string">&quot;lncRNA&quot;</span>的序列名称。</span><br><span class="line"></span><br><span class="line">5. **提取lncRNA序列**: 使用`seqkit grep`命令，根据提取的lncRNA名称从FASTA文件中提取对应的序列。</span><br><span class="line"></span><br><span class="line">6. **生成CPAT所需的hexamer表和逻辑回归模型**: `make_hexamer_tab.py`和`make_logitModel.py`脚本分别用于生成hexamer频率表和逻辑回归模型，这两者都是CPAT评估转录本编码潜力所需的。输入文件包括编码蛋白的mRNA序列（cdna.name.fa）和非编码的lncRNA序列（lncRNA.name.fa）。</span><br><span class="line"></span><br><span class="line">7. **使用CPAT评估转录本的编码潜力**: `cpat.py`脚本根据之前生成的模型和hexamer表来评估`iux200LFPKM0.1.fa`文件中转录本的编码潜力，并输出结果。</span><br><span class="line"></span><br><span class="line">8. **筛选转录本**: 最后，使用`awk`和`<span class="built_in">cut</span>`命令根据ORF长度和编码潜力的阈值筛选转录本。这包括提取ORF长度大于200的转录本，以及编码潜力大于0.3的转录本（可能被认为是mRNA），然后识别出那些编码潜力小于0.3且ORF长度大于200的转录本（被认为是潜在的lncRNA）。</span><br><span class="line"></span><br><span class="line">通过这些步骤，研究人员可以有效地从大量转录本数据中识别出可能具有重要生物学功能的lncRNAs，这对于理解复杂的基因表达调控网络至关重要。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">3.EMBOSS transeq   把核酸序列转换成氨基酸序列</span><br><span class="line">网址：https://www.ebi.ac.uk/Tools/st/emboss_transeq/  在线版本,缺点是序列文件最大为1M，最多序列数为500</span><br><span class="line">把iux200LFPKM0.1.fa 下载到桌面</span><br><span class="line"><span class="built_in">cd</span> /public/jychu/lncRNA-chicken/cleandata/bam/gtfs/merged/Identification</span><br><span class="line">seqkit <span class="built_in">seq</span> iux200LFPKM0.1.fa -w  0  &gt; chicken.iux200LFPKM0.1.fa  <span class="comment">#把序列转换成一行</span></span><br><span class="line"><span class="built_in">wc</span> -l chicken.iux200LFPKM0.1.fa</span><br><span class="line">8034 chicken.iux200LFPKM0.1.fa</span><br><span class="line"></span><br><span class="line"><span class="built_in">cd</span> /mnt/c/Users/TE/Desktop/lncRNA/绍梅师姐数据</span><br><span class="line">sed -n <span class="string">&#x27;1,250p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-1.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;251,450p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-2.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;451,650p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-3.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;651,858p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-4.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;859,890p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-5.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;891,1160p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-6.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;1161,1420p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-7.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;1421,1700p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-8.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;1701,1810p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-9.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;1811,2070p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-10.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;2071,2290p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-11.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;2291,2476p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-12.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;2477,2680p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-13.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;2681,3000p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-14.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;3001,3300p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-15.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;3301,3562p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-16.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;3563,3800p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-17.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;3801,4100p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-18.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;4101,4400p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-19.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;4401,4840p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-20.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;4841,5200p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-21.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;5201,5550p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-22.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;5551,5776p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-23.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;5777,6100p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-24.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;6101,6500p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-25.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;6501,6870p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-26.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;6871,7500p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-27.iux200LFPKM0.1.fa</span><br><span class="line">sed -n <span class="string">&#x27;7501,8034p&#x27;</span> chicken.iux200LFPKM0.1.fa &gt;chicken-28.iux200LFPKM0.1.fa</span><br><span class="line">每一个结果保存为txt格式</span><br><span class="line"><span class="built_in">cat</span> *.txt&gt;chicken.iux.ped <span class="comment">#合并所有结果，上传至服务器/public/jychu/lncRNA-chicken/cleandata/bam/gtfs/merged/Identification/lncRNA下</span></span><br><span class="line"></span><br><span class="line">seqkit <span class="built_in">seq</span> chicken.iux.ped -w 0 &gt;chicken.iux.ped.fa  <span class="comment">#把序列转换成一行</span></span><br><span class="line">grep -rn <span class="string">&quot;MSTRG.13560.1&quot;</span> chicken.iux.ped.fa    <span class="comment">#查看要去除的转录本的行数,转录本翻译成的蛋白长度不能超过100k</span></span><br><span class="line">42493:&gt;MSTRG.13560.1_1</span><br><span class="line">42495:&gt;MSTRG.13560.1_2</span><br><span class="line">42497:&gt;MSTRG.13560.1_3</span><br><span class="line">42499:&gt;MSTRG.13560.1_4</span><br><span class="line">42501:&gt;MSTRG.13560.1_5</span><br><span class="line">42503:&gt;MSTRG.13560.1_6</span><br><span class="line">sed -i <span class="string">&quot;42493,42504d&quot;</span> chicken.iux.ped.fa    <span class="comment">#去除转录本和它的下一行序列行</span></span><br><span class="line">step1 通过hmmmerspress 来把下载的pfam数据建库</span><br><span class="line">使用conda安装Pfam_scan ,hmmpress包含在里面</span><br><span class="line">conda install pfam_scan</span><br><span class="line">下载pfam数据</span><br><span class="line"><span class="built_in">cd</span> database</span><br><span class="line"><span class="built_in">mkdir</span> pfam</span><br><span class="line"><span class="built_in">cd</span> pfam</span><br><span class="line">wget ftp://ftp.ebi.ac.uk:21/pub/databases/Pfam/current_release/Pfam-A.hmm.gz</span><br><span class="line">wget ftp://ftp.ebi.ac.uk:21/pub/databases/Pfam/current_release/Pfam-A.hmm.dat.gz</span><br><span class="line">gunzip *.gz</span><br><span class="line">hmmpress Pfam-A.hmm   <span class="comment">#通过hmmerspress来把下载的数据建库</span></span><br><span class="line"></span><br><span class="line">step2 用HMMER和pfam数据库比对，过滤比对上的序列（Evalue&lt;1e-5）</span><br><span class="line"><span class="built_in">cd</span> /public/jychu/lncRNA-chicken/cleandata/bam/gtfs/merged/Identification/lncRNA</span><br><span class="line"><span class="built_in">nohup</span> pfam_scan.pl -fasta chicken.iux.ped.fa -<span class="built_in">dir</span> /public/jychu/database/pfam -out chicken.pfam_scan.out &amp;</span><br><span class="line">grep -v <span class="string">&quot;#&quot;</span> chicken.pfam_scan.out | grep -v <span class="string">&quot;^\s*$&quot;</span> |awk <span class="string">&#x27;&#123;if($13&lt;1e-5)&#123;print$1&#125;&#125;&#x27;</span> | awk -F <span class="string">&quot;_&quot;</span> <span class="string">&#x27;&#123;print$1&#125;&#x27;</span>|<span class="built_in">sort</span>|<span class="built_in">uniq</span>&gt;pfam.coding.ID  <span class="comment">#提取有编码潜能的ID</span></span><br><span class="line"><span class="built_in">cat</span> pfam.coding.ID ../iux200LFPKM0.1.name|<span class="built_in">sort</span>|<span class="built_in">uniq</span> -u |grep -v <span class="string">&quot;MSTRG.13560.1&quot;</span> &gt;chicken.pfam.lnc  <span class="comment">#得到不具有编码能力的转录本id,也就是lncRNA</span></span><br><span class="line"></span><br><span class="line">step3 利用diamond将蛋白序列和nr-animal,uniRef100数据库比对，过滤比对上的序列(1e-5)</span><br><span class="line">1.和nr-animal比对</span><br><span class="line">~/soft/diamond/diamond blastp -d ~/database/nr/nr.animal -q chicken.iux.ped.fa --sensitive  --evalue 1e-5 -o chicken.nr-animal_matches_fmt6.txt</span><br><span class="line"><span class="built_in">cat</span> chicken.nr-animal_matches_fmt6.txt| awk <span class="string">&#x27;&#123;if($3&gt;80)&#123;print$1&#125;&#125;&#x27;</span>|awk -F <span class="string">&quot;_&quot;</span> <span class="string">&#x27;&#123;print$1&#125;&#x27;</span>|<span class="built_in">sort</span>|<span class="built_in">uniq</span>&gt;nr.coding.ID  <span class="comment">#选择比对上的序列片段相似性大于80%的作为有编码能力的</span></span><br><span class="line"><span class="built_in">cat</span> nr.coding.ID ../iux200LFPKM0.1.name|<span class="built_in">sort</span>|<span class="built_in">uniq</span> -u |grep -v <span class="string">&quot;MSTRG.13560.1&quot;</span> &gt;chicken.nr.lnc</span><br><span class="line">2.和uniRef100比对</span><br><span class="line"><span class="built_in">cd</span> ~/database</span><br><span class="line"><span class="built_in">mkdir</span> uniRef</span><br><span class="line"><span class="built_in">cd</span> uniRef</span><br><span class="line">wget https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref100/uniref100.fasta.gz</span><br><span class="line">gunzip uniref100.fasta.gz   <span class="comment">#解压</span></span><br><span class="line">~/soft/diamond/diamond makedb --<span class="keyword">in</span> uniref100.fasta --db uniref100   <span class="comment">#建立数据库索引</span></span><br><span class="line">~/soft/diamond/diamond blastp -d ~/database/uniRef/uniref100 -q chicken.iux.ped.fa --sensitive  --evalue 1e-5 -o chicken.uniref100_matches_fmt6.txt</span><br><span class="line"><span class="built_in">cat</span> chicken.uniref100_matches_fmt6.txt| awk <span class="string">&#x27;&#123;if($3&gt;80)&#123;print$1&#125;&#125;&#x27;</span>|awk -F <span class="string">&quot;_&quot;</span> <span class="string">&#x27;&#123;print$1&#125;&#x27;</span>|<span class="built_in">sort</span>|<span class="built_in">uniq</span>&gt;uniref100.coding.ID</span><br><span class="line"><span class="built_in">cat</span> uniref100.coding.ID ../iux200LFPKM0.1.name|<span class="built_in">sort</span>|<span class="built_in">uniq</span> -u |grep -v <span class="string">&quot;MSTRG.13560.1&quot;</span> &gt;chicken.uniref100.lnc</span><br><span class="line"></span><br><span class="line"><span class="built_in">mkdir</span> lnc</span><br><span class="line"><span class="built_in">cp</span> *.lnc lnc/     <span class="comment">#把所有预测的lncRNA文件放在一个文件夹下面，便于下载和查找</span></span><br><span class="line"><span class="built_in">cat</span> *.lnc|<span class="built_in">sort</span> |<span class="built_in">uniq</span> -c|<span class="built_in">tr</span> <span class="string">&quot; &quot;</span> <span class="string">&quot;\t&quot;</span>|<span class="built_in">cut</span> -f7,8|awk <span class="string">&#x27;&#123;if($1==5)&#123;print$2&#125;&#125;&#x27;</span>&gt;chicken.lncRNA.lnc  <span class="comment">#重叠的lncID,也就是鉴定出来的lncRNA</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">这个流程是用于鉴定和验证lncRNA（长非编码RNA）序列的一系列步骤，主要通过不同的生物信息学工具和数据库对给定的转录本序列进行编码潜力分析和同源性搜索。下面是这个流程的具体解释：</span><br><span class="line"></span><br><span class="line"><span class="comment">### 第1步：序列处理和分割</span></span><br><span class="line">- 首先，利用`seqkit`将`iux200LFPKM0.1.fa`文件中的序列格式化，每个序列一行，便于处理。</span><br><span class="line">- 因为EMBOSS transeq的在线版本限制了文件大小和序列数目，所以需要将大文件分割成多个小文件，每个包含一定数量的序列。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 第2步：EMBOSS transeq转换</span></span><br><span class="line">- 这些小文件准备上传到EMBOSS transeq工具在线转换核酸序列到氨基酸序列，由于大小和数量限制，这是必须的步骤。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 第3步：合并转换结果</span></span><br><span class="line">- 将所有分割文件的转换结果合并，准备进一步的分析。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 第4步：利用Pfam和HMMER进行同源性搜索</span></span><br><span class="line">- 使用`Pfam_scan.pl`脚本，结合Pfam数据库，对合并后的序列进行同源性搜索，筛选出可能具有编码潜力的序列（基于E-value的阈值）。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 第5步：利用DIAMOND进行同源性搜索</span></span><br><span class="line">- 进一步使用DIAMOND工具对序列进行同源性搜索，与nr-animal和uniRef100数据库比对，筛选出具有高度同源性的序列，即可能具有编码潜力的序列。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 第6步：鉴定lncRNA</span></span><br><span class="line">- 通过上述步骤筛选出的编码潜力序列，与原始转录本列表对比，排除这些可能编码蛋白的转录本，剩下的被认为是lncRNA。</span><br><span class="line">- 最终，通过各种方法的交叉验证，得到最终被认定为lncRNA的转录本列表。</span><br><span class="line"></span><br><span class="line">这个过程涉及到多个生物信息学工具和数据库，每一步都是为了确保所鉴定的lncRNA序列具有较高的可信度，排除那些可能具有编码潜力或与已知蛋白质高度同源的序列。通过这种方法，可以有效地识别出具有研究和临床价值的lncRNA序列。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>









<h1 id="19，画韦恩图，tbtools画就行"><a href="#19，画韦恩图，tbtools画就行" class="headerlink" title="19，画韦恩图，tbtools画就行"></a>19，画韦恩图，tbtools画就行</h1><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line">setwd(<span class="string">&quot;C:/Users/TE/Desktop/lncRNA/绍梅师姐数据/LNC&quot;</span>)</span><br><span class="line">V1 = read.table(<span class="string">&quot;chicken.capt.lnc&quot;</span>,header=F, sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line">capt=V1<span class="variable">$V1</span></span><br><span class="line">V2 = read.table(<span class="string">&quot;chicken.cpc2.lnc&quot;</span>,header=F, sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line">cpc2=V2<span class="variable">$V1</span></span><br><span class="line">V3 = read.table(<span class="string">&quot;chicken.nr.lnc&quot;</span>,header=F, sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line">nr=V3<span class="variable">$V1</span></span><br><span class="line">V4 = read.table(<span class="string">&quot;chicken.pfam.lnc&quot;</span>,header=F, sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line">pfam=V4<span class="variable">$V1</span></span><br><span class="line">V5 = read.table(<span class="string">&quot;chicken.uniref100.lnc&quot;</span>,header=F, sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line">uniref=V5<span class="variable">$V1</span></span><br><span class="line">library(VennDiagram)</span><br><span class="line">venn.diagram(list(<span class="string">&quot;CAPT&quot;</span> = capt,<span class="string">&quot;CPC2&quot;</span> =cpc2,<span class="string">&quot;NR&quot;</span> = nr,<span class="string">&quot;Pfam&quot;</span> =pfam,<span class="string">&quot;UniRef100&quot;</span> = uniref),height=6000,width=6000,resolution=840,</span><br><span class="line">filename=<span class="string">&quot;VennPlot.tiff&quot;</span>,col=<span class="string">&quot;white&quot;</span>,fill = c(<span class="string">&quot;#7fffd4&quot;</span>, <span class="string">&quot;#ff69b4&quot;</span>,<span class="string">&quot;#ffa07a&quot;</span>,<span class="string">&quot;#7cfc00&quot;</span>,<span class="string">&quot;#ff8c00&quot;</span>),</span><br><span class="line">alpha = 0.55,fontfamily = <span class="string">&quot;serif&quot;</span>,fontface = <span class="string">&quot;bold&quot;</span>,                </span><br><span class="line">cat.col = <span class="string">&quot;black&quot;</span>,cat.cex=1.6,cat.dist = c(0.18,0.192,0.22,0.22,0.196),</span><br><span class="line">cat.pos = c(0,330,250,115,25),cat.fontfamily = <span class="string">&quot;serif&quot;</span>)            </span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">这个R脚本是用来生成一个Venn图，目的是为了可视化不同方法预测出的长非编码RNA（lncRNA）之间的重叠情况。Venn图是展示集合间交集与并集的一种图形化方法。在这个脚本中，使用了五个不同的数据集，每个数据集代表一种预测lncRNA的方法或来源：</span><br><span class="line"></span><br><span class="line">1. **CAPT**：通过CPAT预测的lncRNA。</span><br><span class="line">2. **CPC2**：通过CPC2（Coding Potential Calculator 2）预测的lncRNA。</span><br><span class="line">3. **NR**：通过与NR（非冗余蛋白数据库）比对，预测的非编码RNA。</span><br><span class="line">4. **Pfam**：通过与Pfam数据库比对，预测的非编码RNA。</span><br><span class="line">5. **UniRef100**：通过与UniRef100数据库比对，预测的非编码RNA。</span><br><span class="line"></span><br><span class="line">脚本步骤如下：</span><br><span class="line"></span><br><span class="line">- 首先，设置工作目录到存储预测结果文件的文件夹。</span><br><span class="line">- 使用`read.table`函数读取每个预测结果文件，并将结果存储在变量中。</span><br><span class="line">- 载入`VennDiagram`包，用于生成Venn图。</span><br><span class="line">- 使用`venn.diagram`函数生成Venn图，输入是一个列表，列表中包含了上述五个变量。每个变量代表一组lncRNA的集合。</span><br><span class="line">- 在`venn.diagram`函数中，指定了各种参数，如图形的尺寸、分辨率、文件名、颜色、透明度、字体等，以定制Venn图的外观。</span><br><span class="line">- `filename=<span class="string">&quot;VennPlot.tiff&quot;</span>`参数指定输出文件的名称和格式。</span><br><span class="line">- `fill`参数定义了五个集合的颜色。</span><br><span class="line">- `alpha`参数设置颜色的透明度，使得重叠区域的颜色混合可以清晰可见。</span><br><span class="line">- `cat.col`、`cat.cex`、`cat.dist`、`cat.pos`和`cat.fontfamily`等参数用于定制分类标签的颜色、大小、距离、位置和字体。</span><br><span class="line"></span><br><span class="line">通过这个Venn图，研究人员可以直观地看到不同方法预测的lncRNA集合之间的相互关系，包括它们之间的独有和共有预测结果，这有助于评估预测方法的一致性和差异性，以及确定潜在的lncRNA候选。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>


<p>mRNA鉴定的思路：完全匹配到粉脚鹅已有的参考基因组上mRNA的转录本定义为已知的mRNA<br>而lncRNA分为已知的lncRNA和新的lncRNA</p>
<h2 id="20，把所有转录本注释分类为mRNA-lncRNA、新的鉴定的lncRNA等"><a href="#20，把所有转录本注释分类为mRNA-lncRNA、新的鉴定的lncRNA等" class="headerlink" title="20，把所有转录本注释分类为mRNA\lncRNA、新的鉴定的lncRNA等"></a>20，把所有转录本注释分类为mRNA\lncRNA、新的鉴定的lncRNA等</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">已知的</span><br><span class="line"><span class="built_in">cd</span> /public/jychu/reference/annotate/chicken</span><br><span class="line"><span class="built_in">cat</span> Anser_brachyrhynchus.ASM259213v1.104.gtf |awk <span class="string">&#x27;&#123;if($3==&quot;transcript&quot;)&#123;print$0&#125;&#125;&#x27;</span>|<span class="built_in">tr</span> <span class="string">&quot;;&quot;</span> <span class="string">&quot;\t&quot;</span> |<span class="built_in">sort</span> -k11|<span class="built_in">cut</span> -f1,4,5,7,9,11,13-15|sed <span class="string">&quot;s/gene_id //g&quot;</span>|sed <span class="string">&quot;s/transcript_id //g&quot;</span>|sed <span class="string">&quot;s/\&quot;//g&quot;</span>|<span class="built_in">tr</span> <span class="string">&quot; &quot;</span> <span class="string">&quot;:&quot;</span>|sed <span class="string">&quot;s/:ENSABRT/ENSABRT/g&quot;</span>|sed <span class="string">&quot;s/:gene/gene/g&quot;</span>|awk -F <span class="string">&quot;\t&quot;</span> <span class="string">&#x27;&#123;if($7==&quot;gene_source:ensembl&quot;)&#123;$7=$6&#125;&#125;&#123;print$0&#125;&#x27;</span>|<span class="built_in">tr</span> <span class="string">&quot; &quot;</span> <span class="string">&quot;\t&quot;</span>|awk -F <span class="string">&quot;\t&quot;</span> <span class="string">&#x27;&#123;if($8==&quot;gene_source:ensembl&quot;)&#123;$8=$9&#125;&#125;&#123;print$0&#125;&#x27;</span>|<span class="built_in">tr</span> <span class="string">&quot; &quot;</span> <span class="string">&quot;\t&quot;</span>|<span class="built_in">cut</span> -f1-8|sed -e <span class="string">&quot;s/gene_name://g;s/gene_biotype://g&quot;</span>&gt;chicken.transcript.type</span><br><span class="line">下载到桌面，用excel打开</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">新的lncRNA</span><br><span class="line"></span><br><span class="line"><span class="built_in">cd</span> /mnt/c/Users/TE/Desktop/lncRNA/绍梅师姐数据/LNC</span><br><span class="line"><span class="built_in">cat</span> *.lnc |<span class="built_in">sort</span>|<span class="built_in">uniq</span> -c|<span class="built_in">tr</span> <span class="string">&quot; &quot;</span> <span class="string">&quot;\t&quot;</span>|<span class="built_in">cut</span> -f7-8 |awk <span class="string">&#x27;&#123;if($1==5)&#123;print$0&#125;&#125;&#x27;</span>|<span class="built_in">cut</span> -f2&gt;novel.lnc</span><br><span class="line"><span class="built_in">cat</span> novel.lnc | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>;</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">         arr=(<span class="variable">$&#123;id&#125;</span>)</span><br><span class="line">         MSTRG=<span class="variable">$&#123;arr[0]&#125;</span></span><br><span class="line">         N1=<span class="variable">$&#123;arr[1]&#125;</span></span><br><span class="line">         N2=<span class="variable">$&#123;arr[2]&#125;</span></span><br><span class="line">grep  -w <span class="string">&quot;<span class="variable">$MSTRG</span>\.<span class="variable">$N1</span>\.<span class="variable">$N2</span>&quot;</span> ../ballgown/alltranscript.txt &gt;&gt; novel.lnc.expression</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">这段脚本是用于从之前通过多种方法预测得到的长非编码RNA（lncRNA）候选集中，进一步筛选出同时被所有方法预测为lncRNA的转录本，并提取这些转录本的表达数据。</span><br><span class="line"></span><br><span class="line">1. **合并并筛选lncRNA**: 首先，使用`<span class="built_in">cat</span> *.lnc | <span class="built_in">sort</span> | <span class="built_in">uniq</span> -c`命令合并所有的`.lnc`文件（这些文件包含了不同预测方法识别的lncRNA候选），然后通过管道（`|`）传递给`<span class="built_in">tr</span> <span class="string">&quot; &quot;</span> <span class="string">&quot;\t&quot;</span>`将空格转换为制表符，再通过`<span class="built_in">cut</span>`和`awk`命令来筛选出在所有五个预测方法（CAPT、CPC2、NR、Pfam、UniRef100）中都被识别为lncRNA的转录本。这些转录本的ID被保存在`novel.lnc`文件中。</span><br><span class="line"></span><br><span class="line">2. **提取表达数据**: 接下来，脚本通过读取`novel.lnc`文件中的每个转录本ID，使用`grep`命令在一个包含所有转录本表达信息的文件（`../ballgown/alltranscript.txt`）中搜索这些ID对应的表达数据。找到的表达数据被追加到`novel.lnc.expression`文件中。（筛选找到的lncRNA又没有已知的）</span><br><span class="line"></span><br><span class="line">这个过程允许研究人员聚焦于那些高度可能是真正的lncRNA的转录本，并进一步分析这些转录本的表达模式，这对于理解它们在生物过程中的作用是非常有价值的。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">当然可以。让我们通过一个简化的例子来说明这个过程。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 输入文件示例</span></span><br><span class="line"></span><br><span class="line">假设你有多个`.lnc`文件，每个文件都包含一系列转录本ID，这些ID由不同的方法预测为lncRNA。文件内容可能如下：</span><br><span class="line"></span><br><span class="line">**capt.lnc**</span><br></pre></td></tr></table></figure>
<p>MSTRG.1.1<br>MSTRG.2.1<br>MSTRG.3.1</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">**cpc2.lnc**</span><br></pre></td></tr></table></figure>
<p>MSTRG.1.1<br>MSTRG.3.1<br>MSTRG.4.1</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">**nr.lnc**</span><br></pre></td></tr></table></figure>
<p>MSTRG.1.1<br>MSTRG.2.1<br>MSTRG.4.1</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">**pfam.lnc**</span><br></pre></td></tr></table></figure>
<p>MSTRG.1.1<br>MSTRG.2.1<br>MSTRG.5.1</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">**uniref100.lnc**</span><br></pre></td></tr></table></figure>
<p>MSTRG.1.1<br>MSTRG.3.1<br>MSTRG.5.1</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">### 中间过程</span><br><span class="line"></span><br><span class="line">当执行`cat *.lnc | sort | uniq -c | ...`这一系列命令时，目的是找出在所有文件中都出现的转录本ID。在这个例子中，`MSTRG.1.1`是唯一出现在所有五个文件中的ID。</span><br><span class="line"></span><br><span class="line">因此，`novel.lnc`文件将包含：</span><br><span class="line"></span><br><span class="line">**novel.lnc**</span><br></pre></td></tr></table></figure>
<p>MSTRG.1.1</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">### 输出文件示例</span><br><span class="line"></span><br><span class="line">假设`../ballgown/alltranscript.txt`文件包含了转录本及其表达数据，如下所示：</span><br><span class="line"></span><br><span class="line">**alltranscript.txt**</span><br></pre></td></tr></table></figure>
<p>MSTRG.1.1    100    200<br>MSTRG.2.1    150    250<br>MSTRG.3.1    200    300</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">执行第二部分的脚本后，`novel.lnc.expression`文件将包含`MSTRG.1.1`的表达数据，因为它是被所有方法共同预测的lncRNA。</span><br><span class="line"></span><br><span class="line">**novel.lnc.expression**</span><br></pre></td></tr></table></figure>
<p>MSTRG.1.1    100    200</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">在这个简化的例子中，我们看到如何从多个预测方法中筛选出共同的lncRNA转录本，并从一个包含表达数据的文件中提取这些转录本的信息。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>




<h2 id="21，取新的转录本对应的xloc编号，重要，我也能做到，不过利用自己的那个插件"><a href="#21，取新的转录本对应的xloc编号，重要，我也能做到，不过利用自己的那个插件" class="headerlink" title="21，取新的转录本对应的xloc编号，重要，我也能做到，不过利用自己的那个插件"></a>21，取新的转录本对应的xloc编号，重要，我也能做到，不过利用自己的那个插件</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"><span class="built_in">cd</span> /public/jychu/lncRNA-chicken/cleandata/bam/gtfs/merged/Identification</span><br><span class="line"><span class="built_in">cat</span> iux.gtf|awk <span class="string">&#x27;&#123;if($3==&quot;transcript&quot;)&#123;print$0&#125;&#125;&#x27;</span>&gt;test.tmp</span><br><span class="line">vi novel.lnc</span><br><span class="line">Vim 的命令模式中输入 :%s/^M$//g 消除EXCEL模式</span><br><span class="line"><span class="built_in">cat</span> novel.lnc|<span class="built_in">tr</span> <span class="string">&quot;.&quot;</span> <span class="string">&quot;\t&quot;</span> &gt;novel.lnc.tmp</span><br><span class="line">vi test.sh</span><br><span class="line"><span class="built_in">cat</span> novel.lnc.tmp | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>;</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">         arr=(<span class="variable">$&#123;id&#125;</span>)</span><br><span class="line">         MSTRG=<span class="variable">$&#123;arr[0]&#125;</span></span><br><span class="line">         N1=<span class="variable">$&#123;arr[1]&#125;</span></span><br><span class="line">         N2=<span class="variable">$&#123;arr[2]&#125;</span></span><br><span class="line">grep  -w <span class="string">&quot;<span class="variable">$MSTRG</span>\.<span class="variable">$N1</span>\.<span class="variable">$N2</span>&quot;</span> test.tmp &gt;&gt; novel.lnc.tmp.tmp</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">chmod</span> a+x test.sh</span><br><span class="line">./test.sh</span><br><span class="line"><span class="built_in">cat</span> novel.lnc.tmp.tmp|<span class="built_in">cut</span> -f9|<span class="built_in">tr</span> <span class="string">&quot;;&quot;</span> <span class="string">&quot;\t&quot;</span>|<span class="built_in">cut</span> -f1,3- |sed <span class="string">&quot;s/\&quot;//g&quot;</span>|sed -e <span class="string">&quot;s/transcript_id //g;s/xloc //g&quot;</span>|awk <span class="string">&#x27;&#123;$NF=NULL;print&#125;&#x27;</span>|awk <span class="string">&#x27;&#123;$NF=NULL;print&#125;&#x27;</span>|sed <span class="string">&quot;s/class_code//g&quot;</span>|awk <span class="string">&#x27;&#123;if($2==&quot;gene_name&quot;)&#123;print$1&quot;\t&quot;$4&quot;\t&quot;$7&quot;\t&quot;$2&quot;:&quot;$3&quot;;&quot;$5&quot;:&quot;$6&#125;else&#123;print$1&quot;\t&quot;$2&quot;\t&quot;$3&#125;&#125;&#x27;</span>&gt;novel.lnc.test</span><br><span class="line">下载到桌面上</span><br><span class="line"><span class="built_in">rm</span> -f *.tmp</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">这个脚本的目的是处理特定的转录本数据，提取特定转录本的相关信息，并生成一个格式化的输出文件。下面是各个步骤的解释：</span><br><span class="line"></span><br><span class="line">1. **提取转录本信息**:</span><br><span class="line">   - `<span class="built_in">cat</span> iux.gtf | awk <span class="string">&#x27;&#123;if($3==&quot;transcript&quot;)&#123;print$0&#125;&#125;&#x27;</span> &gt; test.tmp`：从`iux.gtf`文件中提取所有表示转录本的行（第三个字段为`transcript`），并将这些行保存到`test.tmp`。</span><br><span class="line"></span><br><span class="line">2. **处理`novel.lnc`文件**:</span><br><span class="line">   - 使用`vi novel.lnc`打开`novel.lnc`文件，并在Vim的命令模式中输入`:%s/^M$//g`来删除由Excel导入时可能引入的特殊字符（`^M`）。</span><br><span class="line">   - `<span class="built_in">cat</span> novel.lnc | <span class="built_in">tr</span> <span class="string">&quot;.&quot;</span> <span class="string">&quot;\t&quot;</span> &gt; novel.lnc.tmp`：将`novel.lnc`文件中的点（`.`）转换为制表符（`\t`），并保存到临时文件`novel.lnc.tmp`中。</span><br><span class="line"></span><br><span class="line">3. **编写和执行`test.sh`脚本**:</span><br><span class="line">   - 在`test.sh`脚本中，通过`<span class="built_in">cat</span> novel.lnc.tmp | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> ... <span class="keyword">done</span>`循环读取`novel.lnc.tmp`中的每行，这些行包含了需要进一步搜索的转录本ID。</span><br><span class="line">   - 对于每个ID，使用`grep -w`从`test.tmp`中查找匹配的行，并将结果追加到`novel.lnc.tmp.tmp`。</span><br><span class="line">   - `<span class="built_in">chmod</span> a+x test.sh`：给`test.sh`脚本文件添加执行权限。</span><br><span class="line">   - `./test.sh`：执行`test.sh`脚本。</span><br><span class="line"></span><br><span class="line">4. **格式化输出文件**:</span><br><span class="line">   - `<span class="built_in">cat</span> novel.lnc.tmp.tmp | <span class="built_in">cut</span> -f9 | <span class="built_in">tr</span> <span class="string">&quot;;&quot;</span> <span class="string">&quot;\t&quot;</span> | <span class="built_in">cut</span> -f1,3- | sed <span class="string">&quot;s/\&quot;//g&quot;</span> | sed -e <span class="string">&quot;s/transcript_id //g;s/xloc //g&quot;</span> | awk <span class="string">&#x27;&#123;$NF=NULL;print&#125;&#x27;</span> | awk <span class="string">&#x27;&#123;$NF=NULL;print&#125;&#x27;</span> | sed <span class="string">&quot;s/class_code //g&quot;</span> | awk <span class="string">&#x27;&#123;...&#125;&#x27;</span> &gt; novel.lnc.test`：这一系列命令对匹配的转录本信息进行了多步处理，包括字段提取、替换、格式化等，最终生成的格式化输出保存在`novel.lnc.test`。</span><br><span class="line"></span><br><span class="line">5. **清理临时文件**:</span><br><span class="line">   - `<span class="built_in">rm</span> -f *.tmp`：删除所有临时文件。</span><br><span class="line"></span><br><span class="line">**输入输出文件例子**:</span><br><span class="line"></span><br><span class="line">- **输入** (`novel.lnc`原始格式):</span><br></pre></td></tr></table></figure>
<p>   MSTRG.1.1<br>   MSTRG.2.1<br>   <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">- **中间文件** (`novel.lnc.tmp`格式化后):</span><br></pre></td></tr></table></figure><br>   MSTRG    1    1<br>   MSTRG    2    1<br>   <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">- **输出** (`novel.lnc.test`):</span><br></pre></td></tr></table></figure><br>   MSTRG.1    gene_name    GeneX    transcript_id:MSTRG.1.1;xloc:X001<br>   MSTRG.2    gene_name    GeneY    transcript_id:MSTRG.2.1;xloc:X002<br>   <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">这个流程从`iux.gtf`文件中筛选特定的转录本，处理`novel.lnc`中的转录本ID，执行文本处理和格式化，最终生成一个包含所需转录本信息的文件。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure></p>
<h2 id="22，差异表达的lncRNA顺式靶基因预测"><a href="#22，差异表达的lncRNA顺式靶基因预测" class="headerlink" title="22，差异表达的lncRNA顺式靶基因预测"></a>22，差异表达的lncRNA顺式靶基因预测</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">先把差异lncRNA筛选出来</span><br><span class="line">在坐标前后加减100kb</span><br><span class="line">然后把所有的mRNA整理出来</span><br><span class="line">-wa -wb 输出overlap的区域所在-a和-b中的原内容：</span><br><span class="line"></span><br><span class="line">bedtools intersect -a lnc.bed.txt -b mRNA.bed.txt -wb |<span class="built_in">cut</span> -f4,8|<span class="built_in">sort</span>|<span class="built_in">uniq</span> &gt;cis.lnc   <span class="comment">#-wb会输出overlap的区域和其中-b文件中的内容</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">这条命令使用`bedtools intersect`来找出`lnc.bed.txt`（长非编码RNA, lncRNA）和`mRNA.bed.txt`（信使RNA, mRNA）文件中空间上相交（overlap）的区域。`-wb`选项意味着除了输出`-a`文件中的内容，还会输出与之相交的`-b`文件中的内容。接着，使用`<span class="built_in">cut</span>`命令提取第4和第8列，这通常代表着特定的标识符或名称（具体取决于BED文件的格式）。然后，`<span class="built_in">sort</span>`和`<span class="built_in">uniq</span>`命令用来排序并去除重复的行，最后的结果被重定向到`cis.lnc`文件中。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 输入文件内容示例</span></span><br><span class="line"></span><br><span class="line">- **lnc.bed.txt (输入文件a)**:</span><br></pre></td></tr></table></figure>
<p>  chr1  1000  2000  lncRNA1<br>  chr1  3000  4000  lncRNA2<br>  <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">  这个文件列出了lncRNA的位置和名称，其中每行包含染色体名、起始位置、结束位置和lncRNA的名称。</span><br><span class="line"></span><br><span class="line">- **mRNA.bed.txt (输入文件b)**:</span><br></pre></td></tr></table></figure><br>  chr1  1500  2500  mRNA1<br>  chr1  3500  4500  mRNA2<br>  <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">  类似地，这个文件列出了mRNA的位置和名称。</span><br><span class="line"></span><br><span class="line">### 输出文件内容示例</span><br><span class="line"></span><br><span class="line">- **cis.lnc (输出文件)**:</span><br></pre></td></tr></table></figure><br>  lncRNA1  mRNA1<br>  lncRNA2  mRNA2<br>  <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">  这个文件包含了空间上与mRNA相交的lncRNA的名称。在这个例子中，`lncRNA1`和`mRNA1`在染色体1上的位置有重叠，`lncRNA2`和`mRNA2`也是如此，这意味着它们可能在功能上有联系，比如在同一个调控网络中。</span><br><span class="line"></span><br><span class="line">这个过程对于研究lncRNA和mRNA之间的空间和功能关系非常重要，特别是在探索基因表达调控和基因间相互作用的研究中。通过识别空间上相交的lncRNA和mRNA，研究者可以进一步分析它们是否参与相同的生物过程或疾病。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure></p>
<h2 id="23，差异表达的lncRNA的反式靶基因准备文件如下格式"><a href="#23，差异表达的lncRNA的反式靶基因准备文件如下格式" class="headerlink" title="23，差异表达的lncRNA的反式靶基因准备文件如下格式"></a>23，差异表达的lncRNA的反式靶基因准备文件如下格式</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">R</span><br><span class="line">data&lt;-read.table(file=<span class="string">&quot;lnc-mRNA.fpkm&quot;</span>,row.names=1,header=F)</span><br><span class="line">corr&lt;- round(cor(t(data)),3)     <span class="comment">#round是保留小数位的函数</span></span><br><span class="line"><span class="built_in">head</span>(corr[,1:16])</span><br><span class="line"><span class="built_in">id</span> = row.names(corr)</span><br><span class="line">n = dim(corr)[1]</span><br><span class="line">re = data.frame(ID1= rep(<span class="built_in">id</span>,n),ID2=rep(<span class="built_in">id</span>,each = n),r = round(as.vector(corr),3))</span><br><span class="line">nrow(re)   <span class="comment">#查看有多少行</span></span><br><span class="line">res=subset(re,re<span class="variable">$r</span>&gt;0.95 | re<span class="variable">$r</span>&lt; -0.95)</span><br><span class="line">res1&lt;-res[grep(pattern=<span class="string">&quot;lncRNA&quot;</span>,res[,1]),]   <span class="comment">#提取第一列有lncRNA字符的行</span></span><br><span class="line">res2&lt;-res1[grep(pattern=<span class="string">&quot;protein_coding&quot;</span>,res1[,2]),]  <span class="comment">#在前面的基础上提取第二列有protein_coding的行（也就是去掉lnc与lnc有相关的行）</span></span><br><span class="line"> write.table(res2,file=<span class="string">&quot;trans.lnc&quot;</span>,sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">这段代码主要用于处理基因表达数据，计算转录本间的相关性，并筛选出与长非编码RNA（lncRNA）表达强相关的蛋白编码基因（protein_coding）。让我们一步步解析代码：</span><br><span class="line"></span><br><span class="line">1. **读取数据**:</span><br><span class="line">   - `data &lt;- read.table(file=<span class="string">&quot;lnc-mRNA.fpkm&quot;</span>, row.names=1, header=F)`：从文件`lnc-mRNA.fpkm`中读取数据，假设第一列为行名（转录本ID），没有列名（header=F）。</span><br><span class="line"></span><br><span class="line">2. **计算相关性**:</span><br><span class="line">   - `corr &lt;- round(cor(t(data)), 3)`：计算转置后的数据矩阵（`t(data)`）的相关性矩阵，并将结果四舍五入到小数点后三位。转置是因为`cor()`函数默认计算列与列之间的相关性，而这里的数据每一行代表一个转录本的表达，每一列代表一个样本。</span><br><span class="line"></span><br><span class="line">3. **查看前16列的相关性**:</span><br><span class="line">   - `<span class="built_in">head</span>(corr[, 1:16])`：查看相关性矩阵的前16列，这可以提供一些关于转录本间相关性的初步信息。</span><br><span class="line"></span><br><span class="line">4. **创建一个包含所有可能的转录本对及其相关性的数据框**:</span><br><span class="line">   - 使用`rep(<span class="built_in">id</span>, n)`和`rep(<span class="built_in">id</span>, each = n)`生成所有可能的转录本对组合。</span><br><span class="line">   - `r = round(as.vector(corr), 3)`：将相关性矩阵转换成一个向量，并四舍五入到小数点后三位。</span><br><span class="line">   - `re = data.frame(ID1= ..., ID2= ..., r = ...)`：创建一个新的数据框，包含每一对转录本的ID（ID1和ID2）及它们之间的相关性（r）。</span><br><span class="line"></span><br><span class="line">5. **筛选高度相关的转录本对**:</span><br><span class="line">   - `res = subset(re, re<span class="variable">$r</span> &gt; 0.95 | re<span class="variable">$r</span> &lt; -0.95)`：从创建的数据框中筛选出相关性大于0.95或小于-0.95的转录本对，即寻找非常强的正相关或负相关。</span><br><span class="line"></span><br><span class="line">6. **进一步筛选涉及lncRNA和蛋白编码基因的对**:</span><br><span class="line">   - `res1 &lt;- res[grep(pattern=<span class="string">&quot;lncRNA&quot;</span>, res[, 1]),]`：在筛选出的高度相关转录本对中，进一步筛选出第一列（ID1）包含<span class="string">&quot;lncRNA&quot;</span>的行。</span><br><span class="line">   - `res2 &lt;- res1[grep(pattern=<span class="string">&quot;protein_coding&quot;</span>, res1[, 2]),]`：在上一步的基础上，进一步筛选出第二列（ID2）包含<span class="string">&quot;protein_coding&quot;</span>的行，即保留那些lncRNA与蛋白编码基因之间高度相关的对。</span><br><span class="line"></span><br><span class="line">7. **输出结果**:</span><br><span class="line">   - `write.table(res2, file=<span class="string">&quot;trans.lnc&quot;</span>, sep=<span class="string">&quot;\t&quot;</span>)`：将最终筛选出的转录本对写入到文件`trans.lnc`中，使用制表符作为字段分隔符。</span><br><span class="line"></span><br><span class="line">这个过程通过筛选相关性高的转录本对，特别是关注于lncRNA与蛋白编码基因之间的相关性，有助于识别可能在调控路径或生物过程中发挥作用的关键lncRNA。</span><br><span class="line"></span><br><span class="line">这段代码旨在分析和筛选出具有高度相关性的长非编码RNA（lncRNA）和蛋白质编码基因（protein_coding）之间的表达量数据。下面是对这个流程的简要解释，以及输入和输出文件的示例。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 流程解释</span></span><br><span class="line"></span><br><span class="line">1. **读取数据**：`data &lt;- read.table(file=<span class="string">&quot;lnc-mRNA.fpkm&quot;</span>, row.names=1, header=F)` 从名为 <span class="string">&quot;lnc-mRNA.fpkm&quot;</span> 的文件中读取表达量数据，其中行名为基因ID，没有列名。</span><br><span class="line"></span><br><span class="line">2. **计算相关性矩阵**：`corr &lt;- round(cor(t(data)), 3)` 计算转录本表达数据的相关性矩阵，并将结果四舍五入到小数点后三位。这里使用 `t(data)` 是因为 `cor()` 函数需要在列之间计算相关性，而输入数据中的每一列代表一个样本，每一行代表一个基因。</span><br><span class="line"></span><br><span class="line">3. **格式化输出**：创建一个新的数据框 `re`，包含每对基因ID的相关性得分。</span><br><span class="line"></span><br><span class="line">4. **筛选高度相关的转录本对**：从 `re` 中筛选出相关性大于0.95或小于-0.95的条目，这意味着只保留极度正相关或负相关的转录本对。</span><br><span class="line"></span><br><span class="line">5. **进一步筛选**：首先筛选出第一列（ID1）包含 <span class="string">&quot;lncRNA&quot;</span> 的行，然后在这个子集中进一步筛选出第二列（ID2）包含 <span class="string">&quot;protein_coding&quot;</span> 的行，即寻找与蛋白质编码基因高度相关的lncRNA。</span><br><span class="line"></span><br><span class="line">6. **保存结果**：将筛选后的结果保存到名为 <span class="string">&quot;trans.lnc&quot;</span> 的文件中。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 输入文件示例 (`lnc-mRNA.fpkm`)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<pre><code>      Sample1 Sample2 Sample3
</code></pre>
<p>lncRNA1    2.5     3.2     2.8<br>lncRNA2    1.2     1.3     1.5<br>geneA      5.1     4.9     5.2<br>geneB      7.8     8.1     7.9</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">这个文件包含了每个转录本在不同样本中的表达量(FPKM值)。</span><br><span class="line"></span><br><span class="line">### 输出文件示例 (`trans.lnc`)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>“ID1”             “ID2”             “r”<br>“lncRNA1”         “geneA”           0.96<br>“lncRNA2”         “geneB”           -0.97</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">这个文件列出了那些与蛋白质编码基因高度相关（相关性大于0.95或小于-0.95）的lncRNA，以及它们之间的相关性系数。</span><br><span class="line"></span><br><span class="line">请注意，实际的输入和输出文件内容会根据实际数据而有所不同，这里只提供了一个简化的示例。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>


<h2 id="24，找有表达量的lncRNA做顺式反式靶基因预测"><a href="#24，找有表达量的lncRNA做顺式反式靶基因预测" class="headerlink" title="24，找有表达量的lncRNA做顺式反式靶基因预测"></a>24，找有表达量的lncRNA做顺式反式靶基因预测</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> /mnt/c/Users/TE/Desktop/lncRNA/绍梅师姐数据/ballgown</span><br><span class="line">bedtools intersect -a alllnc.bed -b mRNA.bed.txt -wa -wb |<span class="built_in">cut</span> -f4,5,9,10&gt;cis.alllnc</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="built_in">cd</span> /mnt/c/Users/TE/Desktop/lncRNA/绍梅师姐数据/ballgown</span><br><span class="line">awk <span class="string">&#x27;&#123;print $1&quot;:&quot;$2&quot;:&quot;$3&quot;\t&quot;$4&quot;\t&quot;$5&quot;\t&quot;$6&quot;\t&quot;$7&quot;\t&quot;$8&quot;\t&quot;$9&#125;&#x27;</span> alllnc-mRNA.fpkm.txt &gt;alllnc-mRNA.fpkm</span><br><span class="line">R</span><br><span class="line">data&lt;-read.table(file=<span class="string">&quot;alllnc-mRNA.fpkm&quot;</span>,row.names=1,header=F)</span><br><span class="line">corr&lt;- round(cor(t(data)),3)     <span class="comment">#round是保留小数位的函数</span></span><br><span class="line"><span class="built_in">head</span>(corr[,1:16])</span><br><span class="line"><span class="built_in">id</span> = row.names(corr)</span><br><span class="line">n = dim(corr)[1]</span><br><span class="line">re = data.frame(ID1= rep(<span class="built_in">id</span>,n),ID2=rep(<span class="built_in">id</span>,each = n),r = round(as.vector(corr),3))</span><br><span class="line">nrow(re)   <span class="comment">#查看有多少行</span></span><br><span class="line">res=subset(re,re<span class="variable">$r</span>&gt;0.95 | re<span class="variable">$r</span>&lt; -0.95)</span><br><span class="line">res1&lt;-res[grep(pattern=<span class="string">&quot;lncRNA&quot;</span>,res[,1]),]   <span class="comment">#提取第一列有lncRNA字符的行</span></span><br><span class="line">res2&lt;-res1[grep(pattern=<span class="string">&quot;protein_coding&quot;</span>,res1[,2]),]  <span class="comment">#在前面的基础上提取第二列有protein_coding的行（也就是去掉lnc与lnc有相关的行）</span></span><br><span class="line">write.table(res2,file=<span class="string">&quot;trans.alllnc&quot;</span>,sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">这个过程涉及几个步骤，旨在找出lncRNA和mRNA之间的相关性，并筛选出高度相关的对。让我们逐步解析：</span><br><span class="line"></span><br><span class="line"><span class="comment">### 第一步：使用`bedtools`寻找lncRNA和mRNA之间的空间重叠</span></span><br><span class="line"></span><br><span class="line">```bash</span><br><span class="line">bedtools intersect -a alllnc.bed -b mRNA.bed.txt -wa -wb | <span class="built_in">cut</span> -f4,5,9,10 &gt; cis.alllnc</span><br></pre></td></tr></table></figure>

<ul>
<li><code>bedtools intersect</code>用于找出<code>alllnc.bed</code>（lncRNA的位置信息）和<code>mRNA.bed.txt</code>（mRNA的位置信息）中空间上重叠的区域。</li>
<li><code>-wa -wb</code>选项表示同时输出<code>-a</code>和<code>-b</code>文件中重叠区域的详细信息。</li>
<li><code>cut -f4,5,9,10</code>提取重要的字段，假设是lncRNA和mRNA的标识符及其位置信息。</li>
<li>输出保存到<code>cis.alllnc</code>文件中，这个文件包含空间上相互作用的lncRNA和mRNA对。</li>
</ul>
<h3 id="第二步：准备相关性分析的数据"><a href="#第二步：准备相关性分析的数据" class="headerlink" title="第二步：准备相关性分析的数据"></a>第二步：准备相关性分析的数据</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">&#x27;&#123;print $1&quot;:&quot;$2&quot;:&quot;$3&quot;\t&quot;$4&quot;\t&quot;$5&quot;\t&quot;$6&quot;\t&quot;$7&quot;\t&quot;$8&quot;\t&quot;$9&#125;&#x27;</span> alllnc-mRNA.fpkm.txt &gt; alllnc-mRNA.fpkm</span><br></pre></td></tr></table></figure>

<ul>
<li>这一步使用<code>awk</code>处理<code>alllnc-mRNA.fpkm.txt</code>文件（包含lncRNA和mRNA的表达量数据），格式化输出以便于R语言读取。</li>
</ul>
<h3 id="第三步：在R中计算相关性并筛选高度相关的lncRNA-mRNA对"><a href="#第三步：在R中计算相关性并筛选高度相关的lncRNA-mRNA对" class="headerlink" title="第三步：在R中计算相关性并筛选高度相关的lncRNA-mRNA对"></a>第三步：在R中计算相关性并筛选高度相关的lncRNA-mRNA对</h3><ol>
<li><strong>读取数据</strong>：<code>data &lt;- read.table(file=&quot;alllnc-mRNA.fpkm&quot;, row.names=1, header=F)</code>读取处理过的表达量数据。</li>
<li><strong>计算相关性</strong>：<code>corr &lt;- round(cor(t(data)), 3)</code>计算转置后的数据矩阵的相关性，<code>round</code>用于保留三位小数。</li>
<li><strong>构建相关性数据框</strong>：构建一个包含所有可能对的数据框<code>re</code>，其中包含两个ID和它们之间的相关系数。</li>
<li><strong>筛选高度相关的对</strong>：<code>res</code>中筛选出相关系数大于0.95或小于-0.95的对，然后进一步筛选出第一列包含<code>lncRNA</code>和第二列包含<code>protein_coding</code>的行，这意味着找出了与mRNA高度相关的lncRNA。</li>
<li><strong>保存结果</strong>：将这些高度相关的lncRNA-mRNA对保存到<code>trans.alllnc</code>文件中。</li>
</ol>
<h3 id="输入文件示例"><a href="#输入文件示例" class="headerlink" title="输入文件示例"></a>输入文件示例</h3><ul>
<li><strong>alllnc.bed</strong>和<strong>mRNA.bed.txt</strong>文件包含BED格式的lncRNA和mRNA的位置信息。</li>
<li><strong>alllnc-mRNA.fpkm.txt</strong>包含lncRNA和mRNA的表达量数据。</li>
</ul>
<h3 id="输出文件示例"><a href="#输出文件示例" class="headerlink" title="输出文件示例"></a>输出文件示例</h3><ul>
<li><strong>cis.alllnc</strong>文件列出了空间上相互作用的lncRNA和mRNA对的标识符及其位置信息。</li>
<li><strong>trans.alllnc</strong>文件包含了基于表达量数据计算得出的高度相关的lncRNA-mRNA对。</li>
</ul>
<p>这个分析流程能够帮助研究者识别可能在基因表达调控中扮演重要角色的lncRNA和mRNA对，为进一步的生物学验证提供依据。</p>
<p>让我给你举个例子来说明这些输入和输出文件可能的内容。</p>
<h3 id="输入文件示例-1"><a href="#输入文件示例-1" class="headerlink" title="输入文件示例"></a>输入文件示例</h3><ol>
<li><p><strong>alllnc.bed</strong> (lncRNA的位置信息，BED格式):</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">chr1    1000    2000    lncRNA1</span><br><span class="line">chr1    5000    6000    lncRNA2</span><br></pre></td></tr></table></figure></li>
<li><p><strong>mRNA.bed.txt</strong> (mRNA的位置信息，BED格式):</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">chr1    1500    2500    mRNA1</span><br><span class="line">chr1    5500    6500    mRNA2</span><br></pre></td></tr></table></figure></li>
<li><p><strong>alllnc-mRNA.fpkm.txt</strong> (lncRNA和mRNA的表达量数据，假设格式为ID和FPKM值):</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">lncRNA1    10</span><br><span class="line">lncRNA2    20</span><br><span class="line">mRNA1      30</span><br><span class="line">mRNA2      40</span><br></pre></td></tr></table></figure></li>
</ol>
<h3 id="中间处理文件示例"><a href="#中间处理文件示例" class="headerlink" title="中间处理文件示例"></a>中间处理文件示例</h3><ul>
<li><strong>alllnc-mRNA.fpkm</strong> (处理后的表达量数据文件，假设格式转换为列格式): <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">ID        FPKM</span><br><span class="line">lncRNA1   10</span><br><span class="line">lncRNA2   20</span><br><span class="line">mRNA1     30</span><br><span class="line">mRNA2     40</span><br></pre></td></tr></table></figure></li>
</ul>
<h3 id="输出文件示例-1"><a href="#输出文件示例-1" class="headerlink" title="输出文件示例"></a>输出文件示例</h3><ol>
<li><p><strong>cis.alllnc</strong> (空间上相互作用的lncRNA和mRNA对及其位置信息):</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">lncRNA1    1000    mRNA1    1500</span><br><span class="line">lncRNA2    5000    mRNA2    5500</span><br></pre></td></tr></table></figure>
<p>这个文件说明<code>lncRNA1</code>与<code>mRNA1</code>、<code>lncRNA2</code>与<code>mRNA2</code>在基因组上有空间上的接近或重叠，可能暗示着它们之间的功能关联或调控作用。</p>
</li>
<li><p><strong>trans.alllnc</strong> (基于表达量数据计算得出的高度相关的lncRNA-mRNA对):</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">lncRNA1    mRNA1    0.95</span><br><span class="line">lncRNA2    mRNA2    -0.96</span><br></pre></td></tr></table></figure>
<p>这个文件列出了基于表达量数据（如FPKM值）计算得出的相关性很高（正相关或负相关）的lncRNA和mRNA对。在这个示例中，<code>lncRNA1</code>与<code>mRNA1</code>的相关性为0.95（正相关），而<code>lncRNA2</code>与<code>mRNA2</code>的相关性为-0.96（负相关），说明这些lncRNA可能通过某种方式影响或与这些mRNA的表达调控相关。</p>
</li>
</ol>
<p>这些示例简单展示了从位置信息和表达量数据出发，通过空间关系（cis作用）和表达量相关性（trans作用）分析，来探索lncRNA与mRNA之间的潜在功能联系的过程。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>




<h2 id="两种FPKM的区别"><a href="#两种FPKM的区别" class="headerlink" title="两种FPKM的区别"></a>两种FPKM的区别</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br></pre></td><td class="code"><pre><span class="line"> 下面两个FPKM值的获得方法，有什么区别或相同吗？</span><br><span class="line"></span><br><span class="line">第一种：</span><br><span class="line"><span class="built_in">cd</span> /public/jychu/lncRNA-chicken/cleandata/bam</span><br><span class="line"><span class="built_in">mkdir</span> ballgown</span><br><span class="line"><span class="built_in">ls</span> *sorted.bam|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span></span><br><span class="line"><span class="keyword">do</span> stringtie -e <span class="variable">$id</span> -p 4 -G gtfs/merged/merged.gtf -o ballgown/<span class="variable">$&#123;id%%.*&#125;</span>.bg.gtf -b ballgown/<span class="variable">$&#123;id&#125;</span>_out_dir</span><br><span class="line"><span class="keyword">done</span>    <span class="comment">#应该是为了取所有样本都注释到的转录本进行定量</span></span><br><span class="line"><span class="comment">#制作转录本表达矩阵   用ballgrown的transcript_fpkm = texpr(bg, &#x27;FPKM&#x27;)  可以直接得出表达矩阵，不过需要把编号和转录本ID对应替换一下</span></span><br><span class="line"><span class="built_in">cd</span> ballgown</span><br><span class="line"><span class="built_in">mkdir</span> quality</span><br><span class="line"><span class="comment">#获取每个样本的转录本ID</span></span><br><span class="line"><span class="built_in">ls</span> *.bg.gtf|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> <span class="built_in">cat</span> <span class="variable">$id</span>| awk <span class="string">&#x27;$3==&quot;transcript&quot;&#x27;</span>|grep -Eo <span class="string">&#x27;transcript_id \&quot;\w+|transcript_id \&quot;\w+\.\w+\.\w+&#x27;</span>|<span class="built_in">cut</span> -d\&quot; -f 2 &gt; quality/<span class="variable">$&#123;id%%.*&#125;</span>.t; <span class="keyword">done</span></span><br><span class="line"><span class="comment">#获取每个样本的FPKM值</span></span><br><span class="line"><span class="built_in">ls</span> *.bg.gtf|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span> </span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">less <span class="variable">$id</span>|awk <span class="string">&#x27;$3==&quot;transcript&quot;&#x27;</span>|grep -Eo <span class="string">&#x27;FPKM \&quot;\w+\.\w+&#x27;</span>|awk -F\&quot; <span class="string">&#x27;&#123;print$2&#125;&#x27;</span> &gt; quality/<span class="variable">$&#123;id%%.*&#125;</span>.value</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="built_in">cd</span> quality</span><br><span class="line"><span class="comment">#合并每个样本转录本和对应的FPKM值</span></span><br><span class="line"><span class="built_in">ls</span> *.t|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> <span class="built_in">paste</span> <span class="variable">$&#123;id%%.*&#125;</span>.t <span class="variable">$&#123;id%%.*&#125;</span>.value &gt; <span class="variable">$&#123;id%%.*&#125;</span>.FPKM ; <span class="keyword">done</span></span><br><span class="line"><span class="comment">#按转录本字母排序</span></span><br><span class="line"><span class="built_in">ls</span> *.FPKM|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>;<span class="keyword">do</span> less <span class="variable">$id</span>|<span class="built_in">sort</span> -k 1 &gt; <span class="variable">$&#123;id%.*&#125;</span>.sorted.FPKM;<span class="keyword">done</span></span><br><span class="line"><span class="comment">#提取样本的第一列并加列名</span></span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;transcript&quot;</span> &gt; FPKM</span><br><span class="line">less `<span class="built_in">ls</span> *.sorted.FPKM|<span class="built_in">shuf</span> -n1`|<span class="built_in">cut</span> -f 1 &gt;&gt; FPKM</span><br><span class="line"><span class="comment">#提取每个样本的第二列并加列名</span></span><br><span class="line"><span class="built_in">ls</span> *.sorted.FPKM|<span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>;<span class="keyword">do</span> <span class="built_in">echo</span> <span class="variable">$&#123;id%%.*&#125;</span> &gt; <span class="variable">$&#123;id%%.*&#125;</span>.tmp;less <span class="variable">$id</span>|<span class="built_in">cut</span> -f 2 &gt;&gt; <span class="variable">$&#123;id%%.*&#125;</span>.tmp;<span class="keyword">done</span></span><br><span class="line"><span class="comment">#合并转录ID列的文件以及所有样本的FPKM值</span></span><br><span class="line"><span class="built_in">paste</span> FPKM *.tmp &gt; all.FPKM</span><br><span class="line"><span class="comment">#删除除了all.FPKM的所有文件</span></span><br><span class="line"><span class="built_in">ls</span> *.value *.t *.FPKM *.tmp FPKM|grep -v all.FPKM |xargs <span class="built_in">rm</span></span><br><span class="line"><span class="comment">#至此获得了转录本的表达矩阵  并下载至桌面上</span></span><br><span class="line"></span><br><span class="line">第二种</span><br><span class="line">featureCounts -T 8 -a /home/data1/Ghl/bam/lncRNA.gtf  -o lncRNA_raw_count.txt -p -B -C -f -t transcript -g transcript_id /home/data1/Ghl/bam/*.short.bam</span><br><span class="line">生成了lncRNA_raw_count.txt</span><br><span class="line">2,R代码</span><br><span class="line"><span class="built_in">rm</span>(list=<span class="built_in">ls</span>())</span><br><span class="line"><span class="comment"># make count table </span></span><br><span class="line">raw_df &lt;- read.table(file = <span class="string">&quot;lncRNA_raw_count.txt&quot;</span>,header = T,skip = 1,sep = <span class="string">&quot;\t&quot;</span>)</span><br><span class="line"></span><br><span class="line">count_df &lt;- raw_df[ ,c(7:ncol(raw_df))]</span><br><span class="line">metadata &lt;- raw_df[ ,1:6] <span class="comment"># 提取基因信息count数据前的几列</span></span><br><span class="line">rownames(count_df) &lt;- as.character(raw_df[,1])</span><br><span class="line">colnames(count_df) &lt;- paste0(<span class="string">&quot;SRR10744&quot;</span>,1:18)</span><br><span class="line"></span><br><span class="line"><span class="comment"># calculate FPKM</span></span><br><span class="line">countToFpkm &lt;- <span class="keyword">function</span>(counts, effLen)</span><br><span class="line">&#123;</span><br><span class="line">  N &lt;- colSums(counts)</span><br><span class="line">  exp( <span class="built_in">log</span>(counts) + <span class="built_in">log</span>(1e9) - <span class="built_in">log</span>(effLen) - <span class="built_in">log</span>(N) )</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">options(scipen = 200) <span class="comment"># 表示在200个数字以内都不使用科学计数法</span></span><br><span class="line">fpkm = countToFpkm(count_df, metadata<span class="variable">$Length</span>)</span><br><span class="line"><span class="comment"># View(fpkm)</span></span><br><span class="line"></span><br><span class="line">write.table(fpkm, file = <span class="string">&quot;fpkm_data.txt&quot;</span>, sep = <span class="string">&quot;\t&quot;</span>, quote = FALSE)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># FPKM &gt; 0 in at least one sample</span></span><br><span class="line">count_df.filter &lt;- count_df[rowSums(fpkm)&gt;0,]</span><br><span class="line"></span><br><span class="line"><span class="comment">#write.table(rownames(count_df.filter,file=&quot;filter6_by_fpkm_id&quot;, sep=&quot;\t&quot;,quote=F))</span></span><br><span class="line">write.table(rownames(count_df.filter), file = <span class="string">&quot;filter6_by_fpkm_id&quot;</span>, sep = <span class="string">&quot;\t&quot;</span>, quote = FALSE)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">这两种方法用于获得FPKM（Fragment Per Kilobase of transcript per Million mapped reads）值，这是一种用来量化基因表达的方法。尽管两者的最终目的相同，即获得转录本的FPKM值，但它们在数据处理流程、工具使用和具体实现细节上有所不同。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 第一种方法：使用StringTie和Ballgown</span></span><br><span class="line"></span><br><span class="line">1. **数据准备与处理**：</span><br><span class="line">   - 使用StringTie进行转录本组装和表达量估计，以`-e`参数对所有样本中都存在的转录本进行定量，输出为`.bg.gtf`文件。</span><br><span class="line">   - 创建一个ballgown目录，用于存放每个样本的转录本定量结果。</span><br><span class="line"></span><br><span class="line">2. **提取转录本ID和FPKM值**：</span><br><span class="line">   - 通过处理`.bg.gtf`文件，分别提取每个样本的转录本ID和对应的FPKM值。</span><br><span class="line"></span><br><span class="line">3. **数据整合**：</span><br><span class="line">   - 将转录本ID和FPKM值合并，生成每个样本的FPKM文件，然后对这些文件进行排序、合并，最终形成一个包含所有样本FPKM值的矩阵。</span><br><span class="line"></span><br><span class="line">这个方法直接从转录本组装和定量的结果中提取FPKM值，依赖于StringTie的输出格式。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 第二种方法：使用featureCounts和R脚本</span></span><br><span class="line"></span><br><span class="line">1. **计数数据生成**：</span><br><span class="line">   - 使用featureCounts对BAM文件进行基因计数，得到原始的计数数据（`lncRNA_raw_count.txt`）。</span><br><span class="line"></span><br><span class="line">2. **FPKM值计算**：</span><br><span class="line">   - 在R中，首先读入featureCounts的输出，然后定义一个函数`countToFpkm`，将原始计数转换为FPKM值。</span><br><span class="line">   - 这个方法考虑了转录本的有效长度和总映射读数，手动计算FPKM值。</span><br><span class="line"></span><br><span class="line">3. **数据过滤和输出**：</span><br><span class="line">   - 对FPKM矩阵进行过滤，仅保留在至少一个样本中FPKM值大于0的转录本。</span><br><span class="line">   - 输出过滤后的转录本ID和对应的FPKM矩阵。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 相同点与不同点</span></span><br><span class="line"></span><br><span class="line">- **相同点**：两种方法最终都提供了转录本的FPKM值，用于量化基因表达水平。</span><br><span class="line">- **不同点**：</span><br><span class="line">  - **工具和流程**：第一种方法依赖于StringTie和Ballgown进行转录本的组装和表达量估计，而第二种方法使用featureCounts进行计数，然后在R中计算FPKM值。</span><br><span class="line">  - **数据处理**：第一种方法直接从转录本定量结果中提取FPKM值，第二种方法则从计数数据出发，使用自定义函数在R中计算FPKM值。</span><br><span class="line">  - **灵活性**：第二种方法在FPKM计算过程中提供了更多的自定义选项，例如可以调整有效长度等参数，而第一种方法的计算细节更多地依赖于StringTie的内部算法。</span><br><span class="line">  - **输出格式**：第一种方法的输出需要额外的步骤来整合和格式化数据，第二种方法则直接在R中进行数据处理和输出，可能对于熟悉R的用户来说更为方便。</span><br><span class="line"></span><br><span class="line">总的来说，选择哪种方法取决于你的具体需求、数据类型以及个人对工具的熟悉程度。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">两种方法都是用来计算FPKM（Fragments Per Kilobase of transcript per Million mapped reads）值，这是一种用来量化基因表达的方法。尽管目的相同，但它们在处理过程、工具使用和细节实现上有所不同。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 方法一：使用StringTie和Ballgown</span></span><br><span class="line"></span><br><span class="line">- **过程**：这种方法首先使用StringTie进行转录本的组装和定量，然后通过Ballgown进行表达矩阵的构建。这个过程更多地依赖于基因组注释文件（如GTF文件）来指导转录本的重建和定量。</span><br><span class="line">- **工具**：StringTie用于转录本的重建和定量，Ballgown用于表达数据的分析和可视化。</span><br><span class="line">- **特点**：这种方法可以提供更细致的转录本水平的分析，能够识别和量化新的或已知的转录本。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 方法二：使用featureCounts和R脚本</span></span><br><span class="line"></span><br><span class="line">- **过程**：这种方法直接使用featureCounts对比对后的BAM文件进行读数计数，然后通过R脚本计算FPKM值。这个过程侧重于已知转录本的定量。</span><br><span class="line">- **工具**：featureCounts用于读数计数，R脚本用于后续的FPKM计算和数据处理。</span><br><span class="line">- **特点**：这种方法相对简单直接，侧重于对已知基因或转录本的表达量进行定量分析。通过R脚本可以灵活地进行数据的进一步处理和分析。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 相同点</span></span><br><span class="line"></span><br><span class="line">- **目标**：两种方法的最终目的都是计算FPKM值，以量化基因或转录本的表达水平。</span><br><span class="line">- **依赖数据**：两者都需要基因组注释信息和比对后的序列数据（BAM格式）。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 不同点</span></span><br><span class="line"></span><br><span class="line">- **工具和流程**：第一种方法依赖于StringTie和Ballgown这两个专门的生物信息学工具，而第二种方法主要依赖于featureCounts和R脚本进行数据处理和FPKM的计算。</span><br><span class="line">- **数据处理深度**：第一种方法可能提供更深入的转录本级别的分析，包括新转录本的发现和注释，而第二种方法更多关注于对已知转录本的量化分析。</span><br><span class="line"></span><br><span class="line"><span class="comment">### 正确性</span></span><br><span class="line"></span><br><span class="line">- **两种方法都是正确的**，选择哪一种取决于研究的具体需求、可用的工具、以及个人对工具的熟悉程度。对于需要深入转录本级别分析的研究，第一种方法可能更合适；而对于关注已知基因或转录本表达量的分析，第二种方法可能更直接和高效。</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>





<h2 id="额外：stringtie-e与非-e"><a href="#额外：stringtie-e与非-e" class="headerlink" title="额外：stringtie -e与非-e"></a>额外：stringtie -e与非-e</h2><p>gffread Sbicolor_454_v3.1.1.gene_exons.gff3.gz -T -o sbi311gx.gtf</p>
<p>#组装转录本：<br>stringtie -p 16 -G ~/sbi311gx.gtf -o T18.gtf T18.bam<br>stringtie -p 16 -G ~/sbi311gx.gtf -o CK18.gtf CK18.bam<br>stringtie -p 16 -G ~/sbi311gx.gtf -o CK69.gtf CK69.bam<br>stringtie -p 16 -G ~/sbi311gx.gtf -o T69.gtf T69.bam</p>
<p>#四样本合并，得到非冗余转录本集<br> stringtie –merge -p 8 <br>-o stringtie_merge.gtf <br>CK18.gtf T18.gtf CK69.gtf T69.gtf</p>
<p>#以stringtie_merge.gtf 为参考注释重新组装<br>stringtie -e -B -p 16 -G ~/4_stringtie/stringtie_merge.gtf -o T18.gtf ~/T18.bam<br>#-B代表输出Ballgown可读文件 </p>
<p>Cuffcompare<br>将各样品的注释文件与参考基因组+参考注释比较，得到新转录本（lncRNA)</p>
<p>筛选条件：</p>
<h1 id="过滤，只保留exon-gt-1并且长度-gt-200bp的transcripts"><a href="#过滤，只保留exon-gt-1并且长度-gt-200bp的transcripts" class="headerlink" title="过滤，只保留exon&gt;1并且长度&gt;200bp的transcripts"></a>过滤，只保留exon&gt;1并且长度&gt;200bp的transcripts</h1><p>1）FPKM&gt;=0.5 2）coverage &gt;1 3）Length &gt; 200 4）分类为i j o u x<br>输入文件：cufcomp.CK18.gtf.tmap 等<br>结果文件：class.CK18等 （内容为注释，格式gtf）<br>提取序列<br>awk命令，比对转录本id<br>以class.CK18 从 CK18.gtf 中提取，文件名为CK18class.gtf<br>gffread： 注释文件 、 参考基因组.fa 提取序列<br>结果文件：CK18class.fa （G：备份数据）</p>
<p>上面的代码是筛选lncRNA的，为什么stringtie -p 16 -G ~/sbi311gx.gtf -o CK18.gtf CK18.bam</p>
<p>没有-e，而下面的有-e，stringtie -e -B -p 16 -G ~/4_stringtie/stringtie_merge.gtf -o T18.gtf ~/T18.bam</p>
<p>这段代码描述的是使用StringTie和其他工具对RNA-Seq数据进行转录本组装、合并和筛选lncRNA的流程。解释<code>-e</code>参数在不同步骤中的使用与不使用的原因，我们需要了解StringTie的参数含义和这些步骤的目的。</p>
<h3 id="StringTie-e参数"><a href="#StringTie-e参数" class="headerlink" title="StringTie -e参数"></a>StringTie <code>-e</code>参数</h3><p><code>-e</code>参数在StringTie中用于告诉程序仅使用提供的参考注释文件（GTF/GFF）中的已知转录本来估计表达量，而不尝试发现任何新的转录本。这通常用于更准确地量化已知转录本的表达，尤其是在后续分析步骤中需要精确的表达量估计时。</p>
<h3 id="不同步骤中-e参数的使用"><a href="#不同步骤中-e参数的使用" class="headerlink" title="不同步骤中-e参数的使用"></a>不同步骤中<code>-e</code>参数的使用</h3><ol>
<li><p><strong>初步组装转录本（没有使用<code>-e</code>）</strong>:</p>
<ul>
<li>在初步组装转录本的步骤中（如<code>stringtie -p 16 -G ~/sbi311gx.gtf -o T18.gtf T18.bam</code>），目的是从每个样本的RNA-Seq数据中尽可能地发现并组装转录本，包括新的和未注释的转录本。因此，不使用<code>-e</code>参数，以允许新转录本的发现。</li>
</ul>
</li>
<li><p><strong>合并转录本集（没有使用<code>-e</code>）</strong>:</p>
<ul>
<li>在合并不同样本的转录本以生成非冗余转录本集的步骤中，同样不使用<code>-e</code>参数，因为目标是整合所有样本中发现的转录本，包括新的和已知的。</li>
</ul>
</li>
<li><p><strong>重新组装并量化（使用<code>-e</code>）</strong>:</p>
<ul>
<li>在使用合并后的转录本集（<code>stringtie_merge.gtf</code>）作为参考来重新组装和量化每个样本的转录本的步骤中，使用了<code>-e</code>参数。这一步骤的目的是基于已经确定的转录本集（包括新发现的和已知的转录本）精确估计每个转录本的表达量。使用<code>-e</code>参数，因为此时关注的是对已识别转录本的精确量化，而不是发现新转录本。</li>
</ul>
</li>
</ol>
<h3 id="总结"><a href="#总结" class="headerlink" title="总结"></a>总结</h3><ul>
<li>在初期的转录本组装和合并步骤中，目标是尽可能地发现和组装转录本，包括新的和未注释的，因此不使用<code>-e</code>参数。</li>
<li>在基于合并后的转录本集重新组装和量化表达量的步骤中，使用<code>-e</code>参数，因为此时的目标是精确量化已经识别的转录本集中的转录本，而不再是发现新的转录本。</li>
</ul>
<h2 id="featureCounts与stringtie"><a href="#featureCounts与stringtie" class="headerlink" title="featureCounts与stringtie"></a>featureCounts与stringtie</h2><p>featureCounts -T 8 -a /home/data1/Ghl/bam/lncRNA.gtf -o lncRNA_raw_count.txt -p -B -C -f -t transcript -g transcript_id /home/data1/Ghl/bam/*.short.bam<br>后续R可以得到FPKM</p>
<p>这个用stringtie可以吗？而且也可以生成的FPKM</p>
<p><code>featureCounts</code>是一个专门用于计数RNA-Seq数据中的reads，映射到基因或转录本上的工具，它是Subread软件包的一部分。它主要用于定量阶段，用来统计对应于特定基因或转录本的reads数目。这个命令使用了多个参数来定制计数的行为，如<code>-T</code>指定线程数，<code>-a</code>指定注释文件，<code>-o</code>指定输出文件名，<code>-p</code>、<code>-B</code>、<code>-C</code>、<code>-f</code>等参数用于指定如何处理reads和如何匹配到注释等。</p>
<p><code>StringTie</code>，另一方面，主要用于组装和量化基于RNA-Seq数据的转录本。尽管它与<code>featureCounts</code>在RNA-Seq数据分析中的作用不同，<code>StringTie</code>确实可以用于生成转录本的表达量估计，包括FPKM（每千基因组片段每百万基对映射reads数）或TPM（每百万转录本片段每千碱基映射reads数）。</p>
<p>若想用<code>StringTie</code>来代替<code>featureCounts</code>的功能，并获取FPKM值，你可以按照以下步骤进行：</p>
<ol>
<li><p><strong>使用<code>StringTie</code>进行转录本的组装和量化</strong>:</p>
<ul>
<li>首先，使用<code>StringTie</code>来组装每个样本的转录本，并生成每个转录本的表达量估计，包括FPKM。这一步是使用<code>StringTie</code>的基本功能，不需要<code>featureCounts</code>参与。</li>
</ul>
</li>
<li><p><strong>命令示例</strong>:</p>
<figure class="highlight sh"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">stringtie -p 8 -G /home/data1/Ghl/bam/lncRNA.gtf -o output.gtf -A gene_abundance.txt -e -B /home/data1/Ghl/bam/sample.short.bam</span><br></pre></td></tr></table></figure>
<ul>
<li><code>-p 8</code>：指定使用的线程数。</li>
<li><code>-G /home/data1/Ghl/bam/lncRNA.gtf</code>：指定注释文件。</li>
<li><code>-o output.gtf</code>：指定输出的GTF文件，其中包含组装的转录本。</li>
<li><code>-A gene_abundance.txt</code>：输出基因丰度估计文件，其中包括FPKM值。</li>
<li><code>-e</code>：仅估计提供的注释文件中的已知转录本的表达量。</li>
<li><code>-B</code>：生成可以被Ballgown读取的文件，用于后续的表达量分析。</li>
</ul>
</li>
</ol>
<p>通过这种方式，你可以使用<code>StringTie</code>来进行转录本的量化，并获取每个转录本的FPKM值，类似于<code>featureCounts</code>的功能，但是是通过不同的方法实现的。需要注意的是，<code>StringTie</code>提供的是转录本级别的表达量估计，而<code>featureCounts</code>提供的是基于注释的reads计数，两者提供的信息虽然相关，但是侧重点和使用的方法不同。</p>
<h3 id="如果有多个bam文件，prepDE-py"><a href="#如果有多个bam文件，prepDE-py" class="headerlink" title="如果有多个bam文件，prepDE.py"></a>如果有多个bam文件，prepDE.py</h3><p>如果你有很多<code>.bam</code>文件需要用<code>StringTie</code>进行转录本的组装和量化，你可以采用批处理的方式来处理它们。这通常涉及编写一个简单的脚本，该脚本遍历所有的<code>.bam</code>文件，并为每个文件运行<code>StringTie</code>命令。这种方法可以自动化处理大量样本，节省时间和减少重复劳动。</p>
<p>下面是一个基于Unix/Linux shell的简单脚本示例，演示如何自动化这个过程：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#!/bin/bash</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置GTF文件的路径</span></span><br><span class="line">GTF_PATH=<span class="string">&quot;/home/data1/Ghl/bam/lncRNA.gtf&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置输出目录</span></span><br><span class="line">OUTPUT_DIR=<span class="string">&quot;/path/to/output/directory&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 创建输出目录</span></span><br><span class="line"><span class="built_in">mkdir</span> -p <span class="string">&quot;<span class="variable">$&#123;OUTPUT_DIR&#125;</span>&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 遍历当前目录下的所有.bam文件</span></span><br><span class="line"><span class="keyword">for</span> BAM_FILE <span class="keyword">in</span> /home/data1/Ghl/bam/*.bam; <span class="keyword">do</span></span><br><span class="line">  <span class="comment"># 获取不包含路径和扩展名的文件名</span></span><br><span class="line">  BASE_NAME=$(<span class="built_in">basename</span> <span class="string">&quot;<span class="variable">$&#123;BAM_FILE&#125;</span>&quot;</span> .bam)</span><br><span class="line"></span><br><span class="line">  <span class="comment"># 运行StringTie</span></span><br><span class="line">  stringtie -p 8 -G <span class="string">&quot;<span class="variable">$&#123;GTF_PATH&#125;</span>&quot;</span> -o <span class="string">&quot;<span class="variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="variable">$&#123;BASE_NAME&#125;</span>.gtf&quot;</span> -A <span class="string">&quot;<span class="variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="variable">$&#123;BASE_NAME&#125;</span>_gene_abundance.txt&quot;</span> -e -B <span class="string">&quot;<span class="variable">$&#123;BAM_FILE&#125;</span>&quot;</span></span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<p>这个脚本做了以下几件事：</p>
<ol>
<li>定义了GTF文件的路径和输出目录的路径。</li>
<li>创建了输出目录（如果它还不存在的话）。</li>
<li>遍历指定目录下的所有<code>.bam</code>文件。</li>
<li>对每个<code>.bam</code>文件运行<code>StringTie</code>，组装转录本并量化表达量，输出结果到指定的输出目录。</li>
</ol>
<p>在运行这个脚本之前，请确保你已经安装了<code>StringTie</code>，并且替换脚本中的<code>GTF_PATH</code>和<code>OUTPUT_DIR</code>为你自己的路径。你可能还需要根据你的系统环境和具体需求调整脚本中的其他参数。</p>
<p>要运行这个脚本，你需要将其保存到一个文件中，例如<code>run_stringtie.sh</code>，然后在终端中给予执行权限并执行它：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">chmod</span> +x run_stringtie.sh</span><br><span class="line">./run_stringtie.sh</span><br></pre></td></tr></table></figure>

<p>这将自动处理指定目录下的所有<code>.bam</code>文件，并为每个文件生成相应的转录本组装和表达量估计文件。</p>
<p>如果你的目标是将多个<code>.bam</code>文件（每个代表一个样本）的表达量（如FPKM值）合并到一个大的文件中，你可以采取以下步骤来实现：</p>
<h3 id="步骤1-为每个-bam文件运行StringTie"><a href="#步骤1-为每个-bam文件运行StringTie" class="headerlink" title="步骤1: 为每个.bam文件运行StringTie"></a>步骤1: 为每个<code>.bam</code>文件运行<code>StringTie</code></h3><p>首先，你需要为每个<code>.bam</code>文件运行<code>StringTie</code>来生成各自的表达量文件。这一步可以使用之前提到的批处理脚本来自动化完成。</p>
<h3 id="步骤2-使用StringTie的prepDE-py脚本合并表达量"><a href="#步骤2-使用StringTie的prepDE-py脚本合并表达量" class="headerlink" title="步骤2: 使用StringTie的prepDE.py脚本合并表达量"></a>步骤2: 使用<code>StringTie</code>的<code>prepDE.py</code>脚本合并表达量</h3><p><code>StringTie</code>附带了一个Python脚本<code>prepDE.py</code>，专门用于从多个样本的输出中提取并合并表达量数据（如FPKM或TPM），生成一个大的CSV或TSV文件，其中包含所有样本的表达量信息。</p>
<p>你可以按照以下步骤使用<code>prepDE.py</code>：</p>
<ol>
<li><p><strong>确保所有<code>StringTie</code>输出文件都在一个目录中</strong>：确保你之前运行<code>StringTie</code>时指定的输出文件（包括每个样本的<code>gene_abundance.txt</code>文件）都在同一个目录下。</p>
</li>
<li><p><strong>运行<code>prepDE.py</code>脚本</strong>：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python prepDE.py -i &lt;input_directory&gt; -g &lt;gene_count_matrix.csv&gt; -t &lt;transcript_count_matrix.csv&gt;</span><br></pre></td></tr></table></figure>
<ul>
<li><code>&lt;input_directory&gt;</code>是包含所有<code>StringTie</code>输出文件的目录。</li>
<li><code>&lt;gene_count_matrix.csv&gt;</code>是你希望生成的基因级别表达量矩阵的文件名。</li>
<li><code>&lt;transcript_count_matrix.csv&gt;</code>是你希望生成的转录本级别表达量矩阵的文件名。</li>
</ul>
</li>
</ol>
<p>这个脚本会扫描指定目录下的所有<code>StringTie</code>输出文件，并从中提取表达量数据，最终生成包含所有样本的表达量矩阵。这些矩阵可以是基于基因或转录本的，并且可以直接用于后续的表达量分析，如差异表达分析等。</p>
<h3 id="注意"><a href="#注意" class="headerlink" title="注意"></a>注意</h3><ul>
<li>确保<code>prepDE.py</code>脚本的路径正确，或者你在其所在的目录下运行命令。</li>
<li>如果你没有<code>prepDE.py</code>脚本，可以从<code>StringTie</code>的GitHub仓库或安装目录中找到。</li>
</ul>
<p>这种方法可以有效地处理大量样本，并将它们的表达量数据合并为单个文件，便于后续分析。</p>
<p>要将多个<code>.bam</code>文件（每个代表一个样本）的FPKM值合并到一个大文件中，你可以遵循以下步骤：</p>
<h3 id="步骤1-组装和量化每个样本的转录本"><a href="#步骤1-组装和量化每个样本的转录本" class="headerlink" title="步骤1: 组装和量化每个样本的转录本"></a>步骤1: 组装和量化每个样本的转录本</h3><ol>
<li><p>对每个<code>.bam</code>文件运行<code>StringTie</code>以组装转录本并计算FPKM值。这可以通过一个批处理脚本自动化完成，脚本需要遍历所有<code>.bam</code>文件并为每个文件执行<code>StringTie</code>命令。</p>
<p>示例命令：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">stringtie -p 8 -G &lt;path_to_annotation.gtf&gt; -o &lt;output_path&gt;/&lt;sample_name&gt;.gtf -A &lt;output_path&gt;/&lt;sample_name&gt;_gene_abundance.txt -e -B &lt;path_to_bam&gt;/&lt;sample_name&gt;.bam</span><br></pre></td></tr></table></figure>
<ul>
<li>其中<code>&lt;path_to_annotation.gtf&gt;</code>是参考注释文件的路径，<code>&lt;output_path&gt;</code>是输出文件存放的目录，<code>&lt;sample_name&gt;</code>是样本名称，<code>&lt;path_to_bam&gt;</code>是<code>.bam</code>文件所在的目录。</li>
</ul>
</li>
</ol>
<h3 id="步骤2-合并所有样本的FPKM值"><a href="#步骤2-合并所有样本的FPKM值" class="headerlink" title="步骤2: 合并所有样本的FPKM值"></a>步骤2: 合并所有样本的FPKM值</h3><ol>
<li><p>使用<code>StringTie</code>提供的<code>prepDE.py</code>脚本来从步骤1生成的所有样本的表达量文件中提取和合并FPKM值。</p>
<p>运行<code>prepDE.py</code>脚本的命令：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python prepDE.py -i &lt;input_directory&gt; -g &lt;gene_count_matrix.csv&gt; -t &lt;transcript_count_matrix.csv&gt;</span><br></pre></td></tr></table></figure>
<ul>
<li><code>&lt;input_directory&gt;</code>是包含所有<code>StringTie</code>输出文件的目录。</li>
<li><code>&lt;gene_count_matrix.csv&gt;</code>是你希望生成的基因级别表达量矩阵的文件名（包含FPKM值）。</li>
<li><code>&lt;transcript_count_matrix.csv&gt;</code>是你希望生成的转录本级别表达量矩阵的文件名（如果需要）。</li>
</ul>
</li>
</ol>
<p>这个脚本将扫描指定目录下的所有<code>StringTie</code>输出文件，提取表达量数据，并生成包含所有样本的表达量矩阵文件。</p>
<h3 id="注意-1"><a href="#注意-1" class="headerlink" title="注意"></a>注意</h3><ul>
<li>确保在执行这些步骤之前，你已经安装了<code>StringTie</code>和Python。</li>
<li>根据你的系统和具体需求，可能需要调整命令中的参数。</li>
<li>确保<code>prepDE.py</code>脚本的路径正确，或者你在其所在的目录下运行命令。</li>
</ul>
<p>通过这些步骤，你将能够得到一个包含所有样本FPKM值的大文件，适用于进一步的表达量分析，如差异表达分析等。</p>

                
            </div>
            <hr/>

            

    <div class="reprint" id="reprint-statement">
        
            <div class="reprint__author">
                <span class="reprint-meta" style="font-weight: bold;">
                    <i class="fas fa-user">
                        文章作者:
                    </i>
                </span>
                <span class="reprint-info">
                    <a href="/about" rel="external nofollow noreferrer">花粱</a>
                </span>
            </div>
            <div class="reprint__type">
                <span class="reprint-meta" style="font-weight: bold;">
                    <i class="fas fa-link">
                        文章链接:
                    </i>
                </span>
                <span class="reprint-info">
                    <a href="https://wplfge3.github.io/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/">https://wplfge3.github.io/2024/02/06/lncRNA%E6%95%99%E7%A8%8B/</a>
                </span>
            </div>
            <div class="reprint__notice">
                <span class="reprint-meta" style="font-weight: bold;">
                    <i class="fas fa-copyright">
                        版权声明:
                    </i>
                </span>
                <span class="reprint-info">
                    本博客所有文章除特別声明外，均采用
                    <a href="https://creativecommons.org/licenses/by/4.0/deed.zh" rel="external nofollow noreferrer" target="_blank">CC BY 4.0</a>
                    许可协议。转载请注明来源
                    <a href="/about" target="_blank">花粱</a>
                    !
                </span>
            </div>
        
    </div>

    <script async defer>
      document.addEventListener("copy", function (e) {
        let toastHTML = '<span>复制成功，请遵循本文的转载规则</span><button class="btn-flat toast-action" onclick="navToReprintStatement()" style="font-size: smaller">查看</a>';
        M.toast({html: toastHTML})
      });

      function navToReprintStatement() {
        $("html, body").animate({scrollTop: $("#reprint-statement").offset().top - 80}, 800);
      }
    </script>



            <div class="tag_share" style="display: block;">
                <div class="post-meta__tag-list" style="display: inline-block;">
                    
                        <div class="article-tag">
                            <span class="chip bg-color">无标签</span>
                        </div>
                    
                </div>
                <div class="post_share" style="zoom: 80%; width: fit-content; display: inline-block; float: right; margin: -0.15rem 0;">
                    <link rel="stylesheet" type="text/css" href="/libs/share/css/share.min.css">
<div id="article-share">

    
    <div class="social-share" data-sites="twitter,facebook,google,qq,qzone,wechat,weibo,douban,linkedin" data-wechat-qrcode-helper="<p>微信扫一扫即可分享！</p>"></div>
    <script src="/libs/share/js/social-share.min.js"></script>
    

    

</div>

                </div>
            </div>
            
                <div id="reward">
    <a href="#rewardModal" class="reward-link modal-trigger btn-floating btn-medium waves-effect waves-light red">赏</a>

    <!-- Modal Structure -->
    <div id="rewardModal" class="modal">
        <div class="modal-content">
            <a class="close modal-close"><i class="fas fa-times"></i></a>
            <h4 class="reward-title">你的赏识是我前进的动力</h4>
            <div class="reward-content">
                <div class="reward-tabs">
                    <ul class="tabs row">
                        <li class="tab col s6 alipay-tab waves-effect waves-light"><a href="#alipay">支付宝</a></li>
                        <li class="tab col s6 wechat-tab waves-effect waves-light"><a href="#wechat">微 信</a></li>
                    </ul>
                    <div id="alipay">
                        <img src="/medias/reward/alipay.jpg" class="reward-img" alt="支付宝打赏二维码">
                    </div>
                    <div id="wechat">
                        <img src="/medias/reward/wechat.png" class="reward-img" alt="微信打赏二维码">
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>

<script>
    $(function () {
        $('.tabs').tabs();
    });
</script>

            
        </div>
    </div>

    

    

    

    

    

    

    

    

    

<article id="prenext-posts" class="prev-next articles">
    <div class="row article-row">
        
        <div class="article col s12 m6" data-aos="fade-up">
            <div class="article-badge left-badge text-color">
                <i class="fas fa-chevron-left"></i>&nbsp;上一篇</div>
            <div class="card">
                <a href="/2024/02/06/%E6%95%99%E5%B8%88%E8%B5%84%E6%A0%BC%E8%80%83%E8%AF%95/">
                    <div class="card-image">
                        
                        
                        <img src="/medias/featureimages/1.jpg" class="responsive-img" alt="教师资格考试">
                        
                        <span class="card-title">教师资格考试</span>
                    </div>
                </a>
                <div class="card-content article-content">
                    <div class="summary block-with-text">
                        
                            
                        
                    </div>
                    <div class="publish-info">
                        <span class="publish-date">
                            <i class="far fa-clock fa-fw icon-date"></i>2024-02-06
                        </span>
                        <span class="publish-author">
                            
                            <i class="fas fa-user fa-fw"></i>
                            花粱
                            
                        </span>
                    </div>
                </div>
                
            </div>
        </div>
        
        
        <div class="article col s12 m6" data-aos="fade-up">
            <div class="article-badge right-badge text-color">
                下一篇&nbsp;<i class="fas fa-chevron-right"></i>
            </div>
            <div class="card">
                <a href="/2024/02/04/tbtools%E6%8F%92%E4%BB%B6%E7%9A%84%E4%B8%89%E7%A7%8D%E6%96%B9%E6%B3%95/">
                    <div class="card-image">
                        
                        
                        <img src="/medias/featureimages/13.jpg" class="responsive-img" alt="tbtools插件的三种方法">
                        
                        <span class="card-title">tbtools插件的三种方法</span>
                    </div>
                </a>
                <div class="card-content article-content">
                    <div class="summary block-with-text">
                        
                            
                        
                    </div>
                    <div class="publish-info">
                            <span class="publish-date">
                                <i class="far fa-clock fa-fw icon-date"></i>2024-02-04
                            </span>
                        <span class="publish-author">
                            
                            <i class="fas fa-user fa-fw"></i>
                            花粱
                            
                        </span>
                    </div>
                </div>
                
            </div>
        </div>
        
    </div>
</article>

</div>



<!-- 代码块功能依赖 -->
<script type="text/javascript" src="/libs/codeBlock/codeBlockFuction.js"></script>



<!-- 代码语言 -->

<script type="text/javascript" src="/libs/codeBlock/codeLang.js"></script>


<!-- 代码块复制 -->

<script type="text/javascript" src="/libs/codeBlock/codeCopy.js"></script>


<!-- 代码块收缩 -->

<script type="text/javascript" src="/libs/codeBlock/codeShrink.js"></script>



    </div>
    <div id="toc-aside" class="expanded col l3 hide-on-med-and-down">
        <div class="toc-widget card" style="background-color: white;">
            <div class="toc-title"><i class="far fa-list-alt"></i>&nbsp;&nbsp;目录</div>
            <div id="toc-content"></div>
        </div>
    </div>
</div>

<!-- TOC 悬浮按钮. -->

<div id="floating-toc-btn" class="hide-on-med-and-down">
    <a class="btn-floating btn-large bg-color">
        <i class="fas fa-list-ul"></i>
    </a>
</div>


<script src="/libs/tocbot/tocbot.min.js"></script>
<script>
    $(function () {
        tocbot.init({
            tocSelector: '#toc-content',
            contentSelector: '#articleContent',
            headingsOffset: -($(window).height() * 0.4 - 45),
            collapseDepth: Number('0'),
            headingSelector: 'h2, h3, h4'
        });

        // Set scroll toc fixed.
        let tocHeight = parseInt($(window).height() * 0.4 - 64);
        let $tocWidget = $('.toc-widget');
        $(window).scroll(function () {
            let scroll = $(window).scrollTop();
            /* add post toc fixed. */
            if (scroll > tocHeight) {
                $tocWidget.addClass('toc-fixed');
            } else {
                $tocWidget.removeClass('toc-fixed');
            }
        });

        
        /* 修复文章卡片 div 的宽度. */
        let fixPostCardWidth = function (srcId, targetId) {
            let srcDiv = $('#' + srcId);
            if (srcDiv.length === 0) {
                return;
            }

            let w = srcDiv.width();
            if (w >= 450) {
                w = w + 21;
            } else if (w >= 350 && w < 450) {
                w = w + 18;
            } else if (w >= 300 && w < 350) {
                w = w + 16;
            } else {
                w = w + 14;
            }
            $('#' + targetId).width(w);
        };

        // 切换TOC目录展开收缩的相关操作.
        const expandedClass = 'expanded';
        let $tocAside = $('#toc-aside');
        let $mainContent = $('#main-content');
        $('#floating-toc-btn .btn-floating').click(function () {
            if ($tocAside.hasClass(expandedClass)) {
                $tocAside.removeClass(expandedClass).hide();
                $mainContent.removeClass('l9');
            } else {
                $tocAside.addClass(expandedClass).show();
                $mainContent.addClass('l9');
            }
            fixPostCardWidth('artDetail', 'prenext-posts');
        });
        
    });
</script>

    

</main>




    <footer class="page-footer bg-color">
    
        <link rel="stylesheet" href="/libs/aplayer/APlayer.min.css">
<style>
    .aplayer .aplayer-lrc p {
        
        display: none;
        
        font-size: 12px;
        font-weight: 700;
        line-height: 16px !important;
    }

    .aplayer .aplayer-lrc p.aplayer-lrc-current {
        
        display: none;
        
        font-size: 15px;
        color: #42b983;
    }

    
    .aplayer.aplayer-fixed.aplayer-narrow .aplayer-body {
        left: -66px !important;
    }

    .aplayer.aplayer-fixed.aplayer-narrow .aplayer-body:hover {
        left: 0px !important;
    }

    
</style>
<div class="">
    
    <div class="row">
        <meting-js class="col l8 offset-l2 m10 offset-m1 s12"
                   server="tencent"
                   type="playlist"
                   id="8835905093"
                   fixed='true'
                   autoplay='false'
                   theme='#42b983'
                   loop='all'
                   order='random'
                   preload='auto'
                   volume='0.7'
                   list-folded='true'
        >
        </meting-js>
    </div>
</div>

<script src="/libs/aplayer/APlayer.min.js"></script>
<script src="/libs/aplayer/Meting.min.js"></script>

    

    <div class="container row center-align"
         style="margin-bottom: 0px !important;">
        <div class="col s12 m8 l8 copy-right">
            Copyright&nbsp;&copy;
            
                <span id="year">2023-2024</span>
            
            <a href="/about" target="_blank">花粱</a>
            |&nbsp;Powered by&nbsp;<a href="https://hexo.io/" target="_blank">Hexo</a>
            |&nbsp;Theme&nbsp;<a href="https://github.com/blinkfox/hexo-theme-matery" target="_blank">Matery</a>
            
            <br>
            
            
            
                
            
            
                <span id="busuanzi_container_site_pv">
                &nbsp;|&nbsp;<i class="far fa-eye"></i>&nbsp;总访问量:&nbsp;
                    <span id="busuanzi_value_site_pv" class="white-color"></span>
            </span>
            
            
                <span id="busuanzi_container_site_uv">
                &nbsp;|&nbsp;<i class="fas fa-users"></i>&nbsp;总访问人数:&nbsp;
                    <span id="busuanzi_value_site_uv" class="white-color"></span>
            </span>
            
            <br>

            <!-- 运行天数提醒. -->
            
            <br>
            
        </div>
        <div class="col s12 m4 l4 social-link social-statis">
    <a href="https://github.com/blinkfox" class="tooltipped" target="_blank" data-tooltip="访问我的GitHub" data-position="top" data-delay="50">
        <i class="fab fa-github"></i>
    </a>



    <a href="mailto:1181062873@qq.com" class="tooltipped" target="_blank" data-tooltip="邮件联系我" data-position="top" data-delay="50">
        <i class="fas fa-envelope-open"></i>
    </a>







    <a href="tencent://AddContact/?fromId=50&fromSubId=1&subcmd=all&uin=1181062873" class="tooltipped" target="_blank" data-tooltip="QQ联系我: 1181062873" data-position="top" data-delay="50">
        <i class="fab fa-qq"></i>
    </a>







    <a href="/atom.xml" class="tooltipped" target="_blank" data-tooltip="RSS 订阅" data-position="top" data-delay="50">
        <i class="fas fa-rss"></i>
    </a>

</div>
    </div>
</footer>

<div class="progress-bar"></div>


    <!-- 搜索遮罩框 -->
<div id="searchModal" class="modal">
    <div class="modal-content">
        <div class="search-header">
            <span class="title"><i class="fas fa-search"></i>&nbsp;&nbsp;搜索</span>
            <input type="search" id="searchInput" name="s" placeholder="请输入搜索的关键字"
                   class="search-input">
        </div>
        <div id="searchResult"></div>
    </div>
</div>

<script type="text/javascript">
$(function () {
    var searchFunc = function (path, search_id, content_id) {
        'use strict';
        $.ajax({
            url: path,
            dataType: "xml",
            success: function (xmlResponse) {
                // get the contents from search data
                var datas = $("entry", xmlResponse).map(function () {
                    return {
                        title: $("title", this).text(),
                        content: $("content", this).text(),
                        url: $("url", this).text()
                    };
                }).get();
                var $input = document.getElementById(search_id);
                var $resultContent = document.getElementById(content_id);
                $input.addEventListener('input', function () {
                    var str = '<ul class=\"search-result-list\">';
                    var keywords = this.value.trim().toLowerCase().split(/[\s\-]+/);
                    $resultContent.innerHTML = "";
                    if (this.value.trim().length <= 0) {
                        return;
                    }
                    // perform local searching
                    datas.forEach(function (data) {
                        var isMatch = true;
                        var data_title = data.title.trim().toLowerCase();
                        var data_content = data.content.trim().replace(/<[^>]+>/g, "").toLowerCase();
                        var data_url = data.url;
                        data_url = data_url.indexOf('/') === 0 ? data.url : '/' + data_url;
                        var index_title = -1;
                        var index_content = -1;
                        var first_occur = -1;
                        // only match artiles with not empty titles and contents
                        if (data_title !== '' && data_content !== '') {
                            keywords.forEach(function (keyword, i) {
                                index_title = data_title.indexOf(keyword);
                                index_content = data_content.indexOf(keyword);
                                if (index_title < 0 && index_content < 0) {
                                    isMatch = false;
                                } else {
                                    if (index_content < 0) {
                                        index_content = 0;
                                    }
                                    if (i === 0) {
                                        first_occur = index_content;
                                    }
                                }
                            });
                        }
                        // show search results
                        if (isMatch) {
                            str += "<li><a href='" + data_url + "' class='search-result-title'>" + data_title + "</a>";
                            var content = data.content.trim().replace(/<[^>]+>/g, "");
                            if (first_occur >= 0) {
                                // cut out 100 characters
                                var start = first_occur - 20;
                                var end = first_occur + 80;
                                if (start < 0) {
                                    start = 0;
                                }
                                if (start === 0) {
                                    end = 100;
                                }
                                if (end > content.length) {
                                    end = content.length;
                                }
                                var match_content = content.substr(start, end);
                                // highlight all keywords
                                keywords.forEach(function (keyword) {
                                    var regS = new RegExp(keyword, "gi");
                                    match_content = match_content.replace(regS, "<em class=\"search-keyword\">" + keyword + "</em>");
                                });

                                str += "<p class=\"search-result\">" + match_content + "...</p>"
                            }
                            str += "</li>";
                        }
                    });
                    str += "</ul>";
                    $resultContent.innerHTML = str;
                });
            }
        });
    };

    searchFunc('/search.xml', 'searchInput', 'searchResult');
});
</script>

    <!-- 白天和黑夜主题 -->
<div class="stars-con">
    <div id="stars"></div>
    <div id="stars2"></div>
    <div id="stars3"></div>  
</div>

<script>
    function switchNightMode() {
        $('<div class="Cuteen_DarkSky"><div class="Cuteen_DarkPlanet"></div></div>').appendTo($('body')),
        setTimeout(function () {
            $('body').hasClass('DarkMode') 
            ? ($('body').removeClass('DarkMode'), localStorage.setItem('isDark', '0'), $('#sum-moon-icon').removeClass("fa-sun").addClass('fa-moon')) 
            : ($('body').addClass('DarkMode'), localStorage.setItem('isDark', '1'), $('#sum-moon-icon').addClass("fa-sun").removeClass('fa-moon')),
            
            setTimeout(function () {
            $('.Cuteen_DarkSky').fadeOut(1e3, function () {
                $(this).remove()
            })
            }, 2e3)
        })
    }
</script>

    <!-- 回到顶部按钮 -->
<div id="backTop" class="top-scroll">
    <a class="btn-floating btn-large waves-effect waves-light" href="#!">
        <i class="fas fa-arrow-up"></i>
    </a>
</div>


    <script src="/libs/materialize/materialize.min.js"></script>
    <script src="/libs/masonry/masonry.pkgd.min.js"></script>
    <script src="/libs/aos/aos.js"></script>
    <script src="/libs/scrollprogress/scrollProgress.min.js"></script>
    <script src="/libs/lightGallery/js/lightgallery-all.min.js"></script>
    <script src="/js/matery.js"></script>

    

    
    
    
        
        <script type="text/javascript">
            // 只在桌面版网页启用特效
            var windowWidth = $(window).width();
            if (windowWidth > 768) {
                document.write('<script type="text/javascript" src="/libs/others/sakura.js"><\/script>');
            }
        </script>
    

    <!-- 雪花特效 -->
    

    <!-- 鼠标星星特效 -->
    

     
        <script src="https://ssl.captcha.qq.com/TCaptcha.js"></script>
        <script src="/libs/others/TencentCaptcha.js"></script>
        <button id="TencentCaptcha" data-appid="xxxxxxxxxx" data-cbfn="callback" type="button" hidden></button>
    

    <!-- Baidu Analytics -->

    <!-- Baidu Push -->

<script>
    (function () {
        var bp = document.createElement('script');
        var curProtocol = window.location.protocol.split(':')[0];
        if (curProtocol === 'https') {
            bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';
        } else {
            bp.src = 'http://push.zhanzhang.baidu.com/push.js';
        }
        var s = document.getElementsByTagName("script")[0];
        s.parentNode.insertBefore(bp, s);
    })();
</script>

    
    <script src="/libs/others/clicklove.js" async="async"></script>
    
    
    <script async src="/libs/others/busuanzi.pure.mini.js"></script>
    

    

    

    <!--腾讯兔小巢-->
    
    

    

    

    
    <script src="/libs/instantpage/instantpage.js" type="module"></script>
    

</body>

</html>
